{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import libraries\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_df_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True):\n",
    "\n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split(\"source_gene=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split(\"source_transcript=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get CHM gene_ids\n",
    "        df[\"CHM_gene_id\"] = df[\"other\"].str.split(\"gene_id=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"CHM_transcript_id\"] = df[\"other\"].str.split(\"transcript_id=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get transcript names\n",
    "        df[\"transcript_name\"] = df[\"other\"].str.split(\"source_transcript_name=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get gene names\n",
    "        df[\"gene_name\"] = df[\"other\"].str.split(\"source_gene_common_name=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get start codon\n",
    "        df[\"start_codon\"] = df[\"other\"].str.split(\"adj_start=\", expand=True)[1].str.split(\";\", expand=True)[0]\n",
    "        \n",
    "        ## Get stop codon\n",
    "        df[\"stop_codon\"] = df[\"other\"].str.split(\"adj_stop=\", expand=True)[1].str.split(\";\", expand=True)[0]        \n",
    "\n",
    "        ## Only keep relevant\n",
    "        df = df[[\"chr\", \"start\", \"end\", \"strand\", \"type\", \"gene_id\", \"transcript_id\", \"CHM_gene_id\",\n",
    "                 \"CHM_transcript_id\", \"transcript_name\", \"gene_name\", \"start_codon\", \"stop_codon\"]].copy()\n",
    "\n",
    "        ## Drop duplicates\n",
    "        df.drop_duplicates(inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get CHM gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get CHM transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Label novel transcripts\n",
    "        df.loc[df[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "        df.loc[~df[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "        ## Label novel genes\n",
    "        df.loc[df[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "        df.loc[~df[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in df.columns:\n",
    "        df.loc[df[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: merge_annotations\n",
    "\n",
    "purpose: Merge useful/relevant information from both annotations while removing repeated and irrelevant information\n",
    "\n",
    "input: Two different GTF annotations\n",
    "\n",
    "output: One GTF annotation containing all the relevant information\n",
    "'''\n",
    "\n",
    "def merge_annotations(ref_gtf, bambu_gtf):\n",
    "    \n",
    "    ## Merge the two annotations\n",
    "    names_ref_gtf = ref_gtf[[\"transcript_id\", \"gene_id\", \"gene_name\", \"transcript_name\"]].copy()\n",
    "    merged_gtf = pd.merge(bambu_gtf, names_ref_gtf, on=['gene_id', 'transcript_id'], how='left')\n",
    "    merged_gtf.drop_duplicates(inplace=True)\n",
    "\n",
    "    ## Label novel transcripts\n",
    "    merged_gtf.loc[merged_gtf[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "    merged_gtf.loc[~merged_gtf[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "    ## Label novel genes\n",
    "    merged_gtf.loc[merged_gtf[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "    merged_gtf.loc[~merged_gtf[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False\n",
    "\n",
    "    ## Create temporary variable only containing novel transcripts\n",
    "    temp = merged_gtf.loc[merged_gtf[\"is_novel_transcript\"] == True]\n",
    "\n",
    "    ## Annotate novel transcripts\n",
    "    merged_tmp = pd.merge(temp, ref_gtf[[\"gene_id\", \"gene_name\"]], on=['gene_id'], how='left')\n",
    "    merged_tmp.drop_duplicates(inplace=True)\n",
    "    merged_tmp[\"gene_name\"] = merged_tmp[\"gene_name_y\"]\n",
    "    merged_tmp.drop(columns=[\"source\", \"gene_name_y\", \"gene_name_x\"], inplace=True)\n",
    "\n",
    "    ## Return novel transcripts to original annotation\n",
    "    merged_final = pd.merge(merged_gtf, merged_tmp, on=['chr', 'type', 'start', 'end', 'strand', 'transcript_id',\n",
    "                    'transcript_name', 'gene_id', 'is_novel_transcript', 'is_novel_gene', 'exon_number'], how=\"left\")\n",
    "\n",
    "    ## Get gene names for novel transcripts of known genes\n",
    "    merged_final.gene_name_x.fillna(merged_final.gene_name_y, inplace=True)\n",
    "    merged_final[\"gene_name\"] = merged_final[\"gene_name_x\"]\n",
    "    merged_final.drop(columns =[\"gene_name_x\", \"gene_name_y\"], inplace=True)\n",
    "    \n",
    "    ## Get start and stop codons for known transcripts and exons of protein coding genes\n",
    "    ref_gtf = ref_gtf[[\"chr\", \"type\", \"start\", \"end\", \"strand\", \"transcript_id\", \"gene_id\", \"start_codon\", \"stop_codon\"]]\n",
    "    merged_final = pd.merge(merged_final, ref_gtf, on=[\"chr\", \"type\", \"start\", \"end\", \"strand\", \"transcript_id\", \"gene_id\"], how=\"left\")\n",
    "    \n",
    "    ## Only keep relevant columns\n",
    "    merged_final = merged_final[[\"chr\", \"type\", \"start\", \"end\", \"strand\", \"transcript_id\", \"gene_id\", \"gene_name\", \n",
    "                    \"exon_number\", \"transcript_name\", \"start_codon\", \"stop_codon\", \"is_novel_gene\", \"is_novel_transcript\"]]\n",
    "\n",
    "\n",
    "    \n",
    "    return merged_final "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: calculate_cpm\n",
    "\n",
    "purpose: calculate cpm from bambu counts file and return it\n",
    "\n",
    "input: Bambu transcript counts file\n",
    "\n",
    "output: Same file with cpm instead of counts\n",
    "'''\n",
    "\n",
    "def calculate_cpm(counts):\n",
    "    \n",
    "    ## Get the names of the count columns\n",
    "    list_counts_cols = counts.columns[2:]\n",
    "    \n",
    "    ## Loop through counts columns\n",
    "    for col in list_counts_cols:\n",
    "        \n",
    "        ## Get sum of counts column\n",
    "        sum_col = counts[col].sum()\n",
    "        \n",
    "        ## Change name of counts column to CPM\n",
    "        new_col_name = col[0:(len(col)-7)] + \"_CPM\"\n",
    "        \n",
    "        ## Calculate CPM for new CPM column\n",
    "        counts[new_col_name] = ((counts[col]/sum_col) * 1000000)\n",
    "        \n",
    "    ## Drop old Columns\n",
    "    counts.drop(columns=list_counts_cols, inplace=True)\n",
    "    \n",
    "    return counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: make_gene_and_transcript_converter\n",
    "\n",
    "input: The CHM13 CAT/Liftoff gff annotation version 2.0\n",
    "\n",
    "output: A dataframe with [\"gene_id\", \"transcript_id\", \"gene_name\", \"transcript_name\"] formatted in the same way as the \n",
    "bambu reference, so that we can properly assign gene and transcript names.\n",
    "\n",
    "purpose: Creating a list that allows us to assign transcript and gene names to the bambu annotation based on the transcript\n",
    "id and gene ID\n",
    "'''\n",
    "\n",
    "\n",
    "def make_gene_and_transcript_converter(gff):\n",
    "    \n",
    "    ## Change name of duplicate Ensembl IDs to CHM IDs\n",
    "    gff.loc[gff[\"transcript_id\"] == \"N/A\", \"transcript_id\"] = gff[\"CHM_transcript_id\"]\n",
    "    gff_transcripts = gff.loc[gff[\"type\"] == \"transcript\"].copy()\n",
    "    gff_transcripts = gff_transcripts[[\"transcript_id\", \"CHM_transcript_id\"]].drop_duplicates()\n",
    "    gff_transcripts = gff_transcripts[gff_transcripts['transcript_id'].duplicated() == True]\n",
    "    dup_trans = gff_transcripts[\"transcript_id\"].dropna().values.tolist()\n",
    "    gff.loc[gff[\"transcript_id\"].isin(dup_trans), \"transcript_id\"] = gff[\"transcript_id\"] + \"(\" + gff[\"CHM_transcript_id\"] + \")\"\n",
    "\n",
    "    ## Change name of duplicate gene ids to CHM ids\n",
    "    gff.loc[gff[\"gene_id\"] == \"None\", \"gene_id\"] = gff[\"CHM_gene_id\"]\n",
    "    gff_genes = gff.loc[gff[\"type\"] == \"transcript\"].copy()\n",
    "    gff_genes = gff_genes[[\"gene_id\", \"CHM_gene_id\"]].drop_duplicates()\n",
    "    gff_genes = gff_genes[gff_genes['gene_id'].duplicated() == True]\n",
    "    dup_genes = gff_genes[\"gene_id\"].dropna().values.tolist()\n",
    "    gff.loc[gff[\"gene_id\"].isin(dup_genes), \"gene_id\"] = gff[\"gene_id\"] + \"(\" + gff[\"CHM_gene_id\"] + \")\"\n",
    "\n",
    "    \n",
    "    ## Fix gene names for MSTRG Genes\n",
    "    gff_names = gff.loc[gff[\"type\"] == \"transcript\"].copy()\n",
    "    gff_names = gff_names[[\"gene_id\", \"gene_name\", \"CHM_gene_id\"]].copy()\n",
    "    gff_names.loc[gff_names[\"gene_name\"].str.contains(\"MSTRG.\"), \"gene_name\"] = np.NaN\n",
    "    gff_names.dropna(inplace=True)\n",
    "    gff_names[\"gene_name\"] = gff_names[\"gene_name\"] + \"(\" + gff_names[\"CHM_gene_id\"] + \")\"\n",
    "    gff_names.drop(columns=\"CHM_gene_id\", inplace=True)\n",
    "    gff_names.drop_duplicates(inplace=True, subset=[\"gene_id\"])\n",
    "    gff.drop(columns=\"gene_name\", inplace=True)\n",
    "    gff = pd.merge(gff, gff_names, on=\"gene_id\", how=\"left\")\n",
    "    gff = gff[[\"gene_id\", \"transcript_id\", \"gene_name\", \"transcript_name\",\n",
    "              \"start\", \"end\", \"type\", \"start_codon\", \"stop_codon\", \"chr\", \"strand\"]].copy()\n",
    "\n",
    "\n",
    "    return gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: make_merged_exons_table\n",
    "\n",
    "input: dataframe with exons file [\"chr\", \"start\", \"end\", \"gene_id\", \"gene_name\", \"strand\"]\n",
    "\n",
    "output: The same dataframe, but the exons that have overlapping positions are merged\n",
    "\n",
    "purpose: Creating a dataframe with merged exons\n",
    "'''\n",
    "\n",
    "def make_merged_exons_table(df):\n",
    "    \n",
    "    ## Drop duplicates and create a list with all gene ids\n",
    "    df.drop_duplicates(inplace=True)\n",
    "    list_gene_ids = df[\"gene_id\"].unique()\n",
    "    \n",
    "    ## Create empty list to hold final results\n",
    "    final_merged_exons_list = []\n",
    "\n",
    "    ## Loop through gene ids\n",
    "    for id in list_gene_ids:\n",
    "        \n",
    "        ## Create dataframe with only that gene ID and sort it by start position\n",
    "        df_gene = df.loc[df[\"gene_id\"] == id].copy()\n",
    "        df_gene.sort_values(by='start', inplace=True)\n",
    "        \n",
    "        ## Create exon list flattened to 1D\n",
    "        exon_list = df_gene.to_numpy()\n",
    "        exon_list = exon_list.flatten()\n",
    "        exon_list = exon_list.tolist()\n",
    "        \n",
    "        ## Start merged exon list\n",
    "        merged_exon_list = exon_list[0:6]\n",
    "\n",
    "        ## Loop through exon list  starting by the second entry, every entry is 6 long\n",
    "        for j in range(12, (len(exon_list) + 6), 6):\n",
    "            \n",
    "            ## If exons overlap (End >= Start on next exon)\n",
    "            if merged_exon_list[(len(merged_exon_list) - 4)] >= (exon_list[j - 5] - 1):\n",
    "                \n",
    "                ## If merged exon does not fully envelop the next exon, then make change the end position\n",
    "                if merged_exon_list[(len(merged_exon_list) - 4)] < exon_list[j - 4]:\n",
    "                    merged_exon_list[(len(merged_exon_list) - 4)] = exon_list[j - 4]\n",
    "            \n",
    "            ## If exons don't overlap add the exon to the list\n",
    "            else:\n",
    "                merged_exon_list.extend(exon_list[(j-6):j])\n",
    "        \n",
    "        ## Once looped through the whole list merge it to the final merged exons list\n",
    "        final_merged_exons_list.extend(merged_exon_list)\n",
    "\n",
    "    ## Create real final list\n",
    "    real_final_list = []\n",
    "\n",
    "    ## Transform 1D final_merged_exon_list into 2D list\n",
    "    for i in range(6, (len(final_merged_exons_list) - 6), 6):\n",
    "        current_list = final_merged_exons_list[i: i + 6]\n",
    "        real_final_list.append(current_list)\n",
    "\n",
    "\n",
    "\n",
    "    ## Transform real_final_list into a dataframe and return it\n",
    "    df_merged_exons = pd.DataFrame(real_final_list,\n",
    "                                   columns =['chr', \"start\", \"end\", \"gene_id\", \"gene_name\", \"strand\"])\n",
    "    \n",
    "    return df_merged_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: find_novel_exons_and_junctions\n",
    "\n",
    "input: dataframe with exons annotation.\n",
    "\n",
    "output: The same dataframe, but the exons flags indicating whether the exon is novel and whether the junction is novel.\n",
    "\n",
    "purpose: Creating flags that indicate whether exons or exon junctions are novel.\n",
    "'''\n",
    "\n",
    "def find_novel_exons_and_junctions(df_test):\n",
    "    \n",
    "    ## Create unique exon ids\n",
    "    df_test[\"unique_exon_id\"] = df_test[\"start\"].astype(str) + \"-\" + df_test[\"end\"].astype(str) + \":\" + df_test[\"gene_id\"]\n",
    "\n",
    "    ## Create a list with all the unique exons ids for all known transcripts\n",
    "    df_known_exons = df_test.loc[df_test[\"is_novel_transcript\"] == False][\"unique_exon_id\"].unique()\n",
    "\n",
    "    ## If the exon ID is contained in the known exons IDs then the exon is not new, otherwise it is\n",
    "    df_test.loc[df_test[\"unique_exon_id\"].isin(df_known_exons), \"is_novel_exon\"] = False\n",
    "    df_test.loc[~df_test[\"unique_exon_id\"].isin(df_known_exons), \"is_novel_exon\"] = True\n",
    "\n",
    "    ## Drop the unique exon id column and set df_known_exons to None to free up memory\n",
    "    df_test.drop(columns=[\"unique_exon_id\"], inplace=True)\n",
    "    df_known_exons = None\n",
    "\n",
    "    ## Sort test dataframe by transcript_id and start position\n",
    "    df_test.sort_values(by=[\"transcript_id\", \"start\"], inplace=True)\n",
    "\n",
    "    ## Create an empty dataframe we can concatenate things into\n",
    "    df_final = df_test.iloc[0:0].copy()\n",
    "\n",
    "    ## Create numpy arrays for data we will be messing with\n",
    "    transcript_ids = df_test['transcript_id'].copy().to_numpy()\n",
    "    gene_ids = df_test['gene_id'].copy().to_numpy()\n",
    "    start = df_test['start'].copy().astype(str).to_numpy()\n",
    "    end = df_test['end'].copy().astype(str).to_numpy()\n",
    "\n",
    "    unique_junction_id_list = np.empty((len(gene_ids)), dtype=object)\n",
    "\n",
    "    ## loop through data and generate unique junction IDs\n",
    "    for i in range(len(gene_ids)):\n",
    "\n",
    "        if i < (len(gene_ids) - 1):\n",
    "\n",
    "            if (transcript_ids[i] == transcript_ids[i+1]):\n",
    "                unique_transcript_id = end[i] + \"-\" + start[i+1] +':' + gene_ids[i]\n",
    "                unique_junction_id_list[i] = unique_transcript_id\n",
    "\n",
    "            else:\n",
    "                unique_junction_id_list[i] = np.NaN\n",
    "        else:\n",
    "            unique_junction_id_list[i] = np.NaN\n",
    "    \n",
    "    ## Set a column in df_test equal to unique_junction_id_list\n",
    "    df_test[\"unique_junction_id\"] = unique_junction_id_list\n",
    "    \n",
    "    ## Create series with known junction ids\n",
    "    df_known_junctions = df_test.loc[df_test[\"is_novel_transcript\"] == False][\"unique_junction_id\"].unique()\n",
    "\n",
    "    ## If an exon has a known junction it is not new, otherwise it is\n",
    "    df_test.loc[df_test[\"unique_junction_id\"].isin(df_known_junctions), \"is_novel_junction\"] = False\n",
    "    df_test.loc[~df_test[\"unique_junction_id\"].isin(df_known_junctions), \"is_novel_junction\"] = True\n",
    "    df_test.loc[df_test[\"unique_junction_id\"].isna(), \"is_novel_junction\"] = False\n",
    "\n",
    "    ## Drop useless columns and free up memory\n",
    "    df_test.drop(columns=[\"unique_junction_id\"], inplace=True)\n",
    "    df_known_exons = None\n",
    "    \n",
    "    return df_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    \n",
    "    ## Import reference GTF, and Create a copy of the original reference CHM13 gff for late user \n",
    "    ref_gff = pd.read_csv(\"../../cDNA-comparison/article_analysis/annotations/CHM13.v2.0.gff3\", delimiter=\"\\t\", header=1, low_memory=False,\n",
    "                     names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "    ## Only keep transcripts and exons, and parse through dataframe to get gene and transcript names.\n",
    "    ref_gff = ref_gff.loc[ref_gff[\"type\"].isin([\"transcript\", \"exon\"])]\n",
    "    ref_gff = parse_df_columns(ref_gff, is_ref=True)\n",
    "\n",
    "\n",
    "    ## Create gene/transcript converter file base on reference genome\n",
    "    ref_gff = make_gene_and_transcript_converter(ref_gff)\n",
    "\n",
    "    ## Import Bambu GTF including novel genes/transcripts. Parse to data to extract gene_id and transcript_id\n",
    "    bambu_gtf = pd.read_csv(\"../novel_gene_body_raw_data/chm13_bambu/extended_annotations.gtf\", header=None, delimiter=\"\\t\", low_memory=False,\n",
    "                           names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "    bambu_gtf = bambu_gtf.loc[bambu_gtf[\"type\"].isin([\"exon\", \"transcript\"])]\n",
    "    bambu_gtf = parse_df_columns(bambu_gtf, is_ref=False)\n",
    "\n",
    "    ## Create merged annotation\n",
    "    merged_gtf = merge_annotations(ref_gff, bambu_gtf)\n",
    "    \n",
    "    ## Import bambu counts file, COLUMN NAMES ARE HARD CODED< PAY ATTENTION, ORDER COULD CHANGE!\n",
    "    bambu_counts = pd.read_csv(\"../novel_gene_body_raw_data/chm13_bambu/counts_transcript.txt\", delimiter=\"\\t\", low_memory=False, header=0,\n",
    "                          names=[\"transcript_id\", \"gene_id\", 'sample_PAM54902', 'sample_PAM54401',\n",
    "                                                              'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814'])\n",
    "\n",
    "    ## Calculate cpm for bambu counts file\n",
    "    bambu_counts = calculate_cpm(bambu_counts)\n",
    "\n",
    "    ## Create annotation with only transcript, add the bambu cpm to the annotation, and save it\n",
    "    transcripts = merged_gtf.loc[merged_gtf[\"type\"] == \"transcript\"].copy()\n",
    "    transcripts = pd.merge(transcripts, bambu_counts, on=[\"transcript_id\", \"gene_id\"], how=\"left\")\n",
    "    transcripts.drop(columns=[\"exon_number\"], inplace=True)\n",
    "    transcripts.to_csv(\"../novel_gene_body_processed_data/2022-06-08_bah_transcript_annotation_and_cpm_cDNA.csv\", index=False)\n",
    "\n",
    "    ## Create exon only annotation and save it\n",
    "    exons = merged_gtf.loc[merged_gtf[\"type\"] == \"exon\"].copy()\n",
    "    exons_final = find_novel_exons_and_junctions(exons)\n",
    "    exons_final.to_csv(\"../novel_gene_body_processed_data/2022-06-08_bah_exon_annotation_cDNA.csv\", index=False)\n",
    "\n",
    "    ## Create a bed file with only exons to later make the merged exons annotation using bedtools\n",
    "    exons_bedfile = exons_final[['chr', 'start', 'end', 'gene_id', 'gene_name', 'strand']].copy()\n",
    "    exons_merged_file = make_merged_exons_table(exons_bedfile)\n",
    "    exons_merged_file.to_csv(\"../novel_gene_body_processed_data/2022-06-08_bah_merged_exons_cDNA.csv\", index=False)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"../novel_gene_body_processed_data/2022-06-08_bah_exon_annotation_cDNA.csv\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_novel = df.loc[df[\"is_novel_transcript\"] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "621"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_novel[\"gene_id\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>start_codon</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>is_novel_gene</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_exon</th>\n",
       "      <th>is_novel_junction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1446945</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>103302104.0</td>\n",
       "      <td>103302530.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.97</td>\n",
       "      <td>ENSG00000170417.16</td>\n",
       "      <td>TMEM182(CHM13_G0031823)</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446946</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>105314453.0</td>\n",
       "      <td>105314750.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.98</td>\n",
       "      <td>ENSG00000198914.5</td>\n",
       "      <td>POU3F3(CHM13_G0031856)</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446947</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>105387145.0</td>\n",
       "      <td>105388956.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.98</td>\n",
       "      <td>ENSG00000198914.5</td>\n",
       "      <td>POU3F3(CHM13_G0031856)</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446948</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>105330404.0</td>\n",
       "      <td>105331134.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.99</td>\n",
       "      <td>ENSG00000229743.4</td>\n",
       "      <td>LINC01159(CHM13_G0031857)</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446949</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>105333705.0</td>\n",
       "      <td>105333802.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.99</td>\n",
       "      <td>ENSG00000229743.4</td>\n",
       "      <td>LINC01159(CHM13_G0031857)</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  type        start          end strand transcript_id  \\\n",
       "1446945  chr2  exon  103302104.0  103302530.0      +         tx.97   \n",
       "1446946  chr2  exon  105314453.0  105314750.0      +         tx.98   \n",
       "1446947  chr2  exon  105387145.0  105388956.0      +         tx.98   \n",
       "1446948  chr2  exon  105330404.0  105331134.0      -         tx.99   \n",
       "1446949  chr2  exon  105333705.0  105333802.0      -         tx.99   \n",
       "\n",
       "                    gene_id                  gene_name  exon_number  \\\n",
       "1446945  ENSG00000170417.16    TMEM182(CHM13_G0031823)            5   \n",
       "1446946   ENSG00000198914.5     POU3F3(CHM13_G0031856)            1   \n",
       "1446947   ENSG00000198914.5     POU3F3(CHM13_G0031856)            2   \n",
       "1446948   ENSG00000229743.4  LINC01159(CHM13_G0031857)            2   \n",
       "1446949   ENSG00000229743.4  LINC01159(CHM13_G0031857)            1   \n",
       "\n",
       "        transcript_name  start_codon  stop_codon  is_novel_gene  \\\n",
       "1446945             NaN          NaN         NaN          False   \n",
       "1446946             NaN          NaN         NaN          False   \n",
       "1446947             NaN          NaN         NaN          False   \n",
       "1446948             NaN          NaN         NaN          False   \n",
       "1446949             NaN          NaN         NaN          False   \n",
       "\n",
       "         is_novel_transcript  is_novel_exon  is_novel_junction  \n",
       "1446945                 True           True              False  \n",
       "1446946                 True           True               True  \n",
       "1446947                 True           True              False  \n",
       "1446948                 True          False               True  \n",
       "1446949                 True           True              False  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_novel.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_counts = pd.read_csv(\"../novel_gene_body_raw_data/chm13_bambu/counts_transcript.txt\", delimiter=\"\\t\", low_memory=False, header=0,\n",
    "                          names=[\"transcript_id\", \"gene_id\", 'sample_PAM54902', 'sample_PAM54401',\n",
    "                                                              'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_counts_novel = bambu_counts.loc[bambu_counts[\"transcript_id\"].str.startswith(\"tx.\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter novel genes and transcripts, must be present in every sample with 5+ counts.\n",
    "count_cols = ['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']\n",
    "\n",
    "for col in count_cols:\n",
    "    bambu_counts_novel = bambu_counts_novel.loc[bambu_counts_novel[col] > 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_novel = df_novel.loc[df_novel[\"transcript_id\"].isin(bambu_counts_novel[\"transcript_id\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    1515\n",
       "True      668\n",
       "Name: is_novel_junction, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_novel[\"is_novel_junction\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     1185\n",
       "False     998\n",
       "Name: is_novel_exon, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_novel[\"is_novel_exon\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>start_codon</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>is_novel_gene</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_exon</th>\n",
       "      <th>is_novel_junction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1444516</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>9994993.0</td>\n",
       "      <td>9995515.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.1</td>\n",
       "      <td>ENSG00000251503.8</td>\n",
       "      <td>CENPS-CORT(CHM13_G0000323)</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444522</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>113040841.0</td>\n",
       "      <td>113041602.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.101</td>\n",
       "      <td>ENSG00000125611.15</td>\n",
       "      <td>CHCHD5(CHM13_G0032033)</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444527</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>114036198.0</td>\n",
       "      <td>114036977.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.103</td>\n",
       "      <td>ENSG00000240356.6</td>\n",
       "      <td>RPL23AP7(CHM13_G0032075)</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444529</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>114051609.0</td>\n",
       "      <td>114051845.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.103</td>\n",
       "      <td>ENSG00000240356.6</td>\n",
       "      <td>RPL23AP7(CHM13_G0032075)</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444532</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>119715925.0</td>\n",
       "      <td>119716307.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.104</td>\n",
       "      <td>gene.185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446940</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>102082397.0</td>\n",
       "      <td>102082766.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.96</td>\n",
       "      <td>gene.136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446941</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>103195595.0</td>\n",
       "      <td>103195693.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.97</td>\n",
       "      <td>ENSG00000170417.16</td>\n",
       "      <td>TMEM182(CHM13_G0031823)</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446945</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>103302104.0</td>\n",
       "      <td>103302530.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.97</td>\n",
       "      <td>ENSG00000170417.16</td>\n",
       "      <td>TMEM182(CHM13_G0031823)</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446947</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>105387145.0</td>\n",
       "      <td>105388956.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.98</td>\n",
       "      <td>ENSG00000198914.5</td>\n",
       "      <td>POU3F3(CHM13_G0031856)</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1446949</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>105333705.0</td>\n",
       "      <td>105333802.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.99</td>\n",
       "      <td>ENSG00000229743.4</td>\n",
       "      <td>LINC01159(CHM13_G0031857)</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>626 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  type        start          end strand transcript_id  \\\n",
       "1444516  chr1  exon    9994993.0    9995515.0      +          tx.1   \n",
       "1444522  chr2  exon  113040841.0  113041602.0      +        tx.101   \n",
       "1444527  chr2  exon  114036198.0  114036977.0      -        tx.103   \n",
       "1444529  chr2  exon  114051609.0  114051845.0      -        tx.103   \n",
       "1444532  chr2  exon  119715925.0  119716307.0      -        tx.104   \n",
       "...       ...   ...          ...          ...    ...           ...   \n",
       "1446940  chr2  exon  102082397.0  102082766.0      +         tx.96   \n",
       "1446941  chr2  exon  103195595.0  103195693.0      +         tx.97   \n",
       "1446945  chr2  exon  103302104.0  103302530.0      +         tx.97   \n",
       "1446947  chr2  exon  105387145.0  105388956.0      +         tx.98   \n",
       "1446949  chr2  exon  105333705.0  105333802.0      -         tx.99   \n",
       "\n",
       "                    gene_id                   gene_name  exon_number  \\\n",
       "1444516   ENSG00000251503.8  CENPS-CORT(CHM13_G0000323)            6   \n",
       "1444522  ENSG00000125611.15      CHCHD5(CHM13_G0032033)            2   \n",
       "1444527   ENSG00000240356.6    RPL23AP7(CHM13_G0032075)            3   \n",
       "1444529   ENSG00000240356.6    RPL23AP7(CHM13_G0032075)            1   \n",
       "1444532            gene.185                         NaN            1   \n",
       "...                     ...                         ...          ...   \n",
       "1446940            gene.136                         NaN            4   \n",
       "1446941  ENSG00000170417.16     TMEM182(CHM13_G0031823)            1   \n",
       "1446945  ENSG00000170417.16     TMEM182(CHM13_G0031823)            5   \n",
       "1446947   ENSG00000198914.5      POU3F3(CHM13_G0031856)            2   \n",
       "1446949   ENSG00000229743.4   LINC01159(CHM13_G0031857)            1   \n",
       "\n",
       "        transcript_name  start_codon  stop_codon  is_novel_gene  \\\n",
       "1444516             NaN          NaN         NaN          False   \n",
       "1444522             NaN          NaN         NaN          False   \n",
       "1444527             NaN          NaN         NaN          False   \n",
       "1444529             NaN          NaN         NaN          False   \n",
       "1444532             NaN          NaN         NaN           True   \n",
       "...                 ...          ...         ...            ...   \n",
       "1446940             NaN          NaN         NaN           True   \n",
       "1446941             NaN          NaN         NaN          False   \n",
       "1446945             NaN          NaN         NaN          False   \n",
       "1446947             NaN          NaN         NaN          False   \n",
       "1446949             NaN          NaN         NaN          False   \n",
       "\n",
       "         is_novel_transcript  is_novel_exon  is_novel_junction  \n",
       "1444516                 True           True              False  \n",
       "1444522                 True           True              False  \n",
       "1444527                 True           True              False  \n",
       "1444529                 True           True              False  \n",
       "1444532                 True           True              False  \n",
       "...                      ...            ...                ...  \n",
       "1446940                 True           True              False  \n",
       "1446941                 True           True              False  \n",
       "1446945                 True           True              False  \n",
       "1446947                 True           True              False  \n",
       "1446949                 True           True              False  \n",
       "\n",
       "[626 rows x 16 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_novel.loc[((df_novel[\"is_novel_junction\"] == False) & (df_novel[\"is_novel_exon\"] == True))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>start_codon</th>\n",
       "      <th>stop_codon</th>\n",
       "      <th>is_novel_gene</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_exon</th>\n",
       "      <th>is_novel_junction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1444530</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>119714686.0</td>\n",
       "      <td>119715057.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.104</td>\n",
       "      <td>gene.185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444531</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>119715305.0</td>\n",
       "      <td>119715421.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.104</td>\n",
       "      <td>gene.185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444532</th>\n",
       "      <td>chr2</td>\n",
       "      <td>exon</td>\n",
       "      <td>119715925.0</td>\n",
       "      <td>119716307.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.104</td>\n",
       "      <td>gene.185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  type        start          end strand transcript_id   gene_id  \\\n",
       "1444530  chr2  exon  119714686.0  119715057.0      -        tx.104  gene.185   \n",
       "1444531  chr2  exon  119715305.0  119715421.0      -        tx.104  gene.185   \n",
       "1444532  chr2  exon  119715925.0  119716307.0      -        tx.104  gene.185   \n",
       "\n",
       "        gene_name  exon_number transcript_name  start_codon  stop_codon  \\\n",
       "1444530       NaN            3             NaN          NaN         NaN   \n",
       "1444531       NaN            2             NaN          NaN         NaN   \n",
       "1444532       NaN            1             NaN          NaN         NaN   \n",
       "\n",
       "         is_novel_gene  is_novel_transcript  is_novel_exon  is_novel_junction  \n",
       "1444530           True                 True           True               True  \n",
       "1444531           True                 True           True               True  \n",
       "1444532           True                 True           True              False  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df[\"gene_id\"] == \"gene.185\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
