{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cfde5405",
   "metadata": {},
   "source": [
    "# Table of contents\n",
    "\n",
    "### - Import libraries and define functions + Initial setup\n",
    "\n",
    "### - Data pre-processing (Our Data)\n",
    "\n",
    "### - Data pre-processing (GTEX Data)\n",
    "\n",
    "### - Data pre-processing (Merge Data)\n",
    "\n",
    "### - Make figures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9928c02",
   "metadata": {},
   "source": [
    "# Import libraries and define functions + Initial setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e7c2845",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "from wordcloud import WordCloud\n",
    "import csv\n",
    "\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f27e634a",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: relative_transcript_abundance\n",
    "\n",
    "purpose: calculate relative transcript abundance\n",
    "\n",
    "input: a dataframe with a ref_gene_id column identifying the transcript gene of origin and a cov columns with \n",
    "the coverage for the transcripts.\n",
    "\n",
    "output: the same dataframe with a relative abundance column added\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "def relative_transcript_abundance(df):\n",
    "    \n",
    "    dff = df.copy()\n",
    "    \n",
    "    ## Group by gene_id and get total expression for each gene (not counting introns)\n",
    "    dff_sums = dff[[\"gene_id\", \"total_CPM\"]].groupby(\"gene_id\").sum()\n",
    "    dff_sums[\"total_CPM_gene\"] = dff_sums[\"total_CPM\"]\n",
    "    dff_sums.drop(columns=\"total_CPM\", inplace=True)\n",
    "    \n",
    "    ## Merge dataframe with total gene level CPM with regular transcript level CPM dataframe\n",
    "    merged_dff = pd.merge(dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "    \n",
    "    ## Calculater relative percent abundance for each transcript within its gene\n",
    "    merged_dff[\"relative_abundance_percent\"] = ((merged_dff[\"total_CPM\"]/merged_dff[\"total_CPM_gene\"]) * 100)\n",
    "    \n",
    "    ## Rename total_CPM for transcript column\n",
    "    merged_dff[\"total_CPM_transcript\"] = merged_dff[\"total_CPM\"]\n",
    "    merged_dff.drop(columns=\"total_CPM\", inplace=True)\n",
    "\n",
    "\n",
    "    return merged_dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a6446e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: fix_column_names\n",
    "\n",
    "purpose: Fixing the column names, making them smaller, informative, and consistent\n",
    "\n",
    "input: The raw counts dataframe for either genes or transcripts \n",
    "\n",
    "output: Same dataframe with improved column names\n",
    "'''\n",
    "\n",
    "def fix_column_names(df, is_gene=False):\n",
    "    \n",
    "    \n",
    "    dff = df.copy()\n",
    "    \n",
    "    ## Check if this is a gene counts object\n",
    "    if is_gene:\n",
    "        \n",
    "        ## Get count column names and create list of new column names\n",
    "        count_columns = dff.columns.tolist()\n",
    "        list_new_names = [\"gene_id\"]\n",
    "        \n",
    "        ## gene_id comes in as index for gene counts data, make it into the first column instead\n",
    "        dff[\"gene_id\"] = dff.index\n",
    "        cols = list(dff.columns)\n",
    "        cols = [cols[-1]] + cols[:-1]\n",
    "        dff = dff[cols]\n",
    "        dff.reset_index(inplace=True, drop=True)\n",
    "    \n",
    "    ## If it is a transcript dataset\n",
    "    else:\n",
    "        ## Set count columns and create list of new names\n",
    "        count_columns = dff.columns[2:].tolist()\n",
    "        list_new_names = [ \"transcript_id\", \"gene_id\"]\n",
    "    \n",
    "    ## Fix names one by one and add to list of new names\n",
    "    for col in count_columns:\n",
    "        col = col.split(\"_mapped\")[0] + \"_counts\"\n",
    "        list_new_names.append(col)\n",
    "    \n",
    "    ## Rename columns\n",
    "    dff.columns = list_new_names\n",
    "    \n",
    "    return dff "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "143176a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_dff_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True, is_transcript=False, is_prot=False, delete_other=True):\n",
    "\n",
    "    dff = df.copy()\n",
    "    \n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        dff[\"gene_id\"] = dff[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "        \n",
    "        ## Get gene names\n",
    "        dff[\"gene_name\"] = dff[\"other\"].str.split(\"gene_name \\\"\", expand=True)[1].str.split('\\\";', expand=True)[0]\n",
    "        \n",
    "        ## Get get transcript biotype\n",
    "        dff[\"gene_biotype\"] = dff[\"other\"].str.split('gene_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "        \n",
    "        ## If is transcript get transcript id and transcript biotype\n",
    "        if is_transcript:\n",
    "            dff[\"transcript_id\"] = dff[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            dff[\"transcript_biotype\"] = dff[\"other\"].str.split('transcript_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            \n",
    "            ## If is prot get protein_id\n",
    "            if is_prot:\n",
    "                dff[\"protein_id\"] = dff[\"other\"].str.split('protein_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "                dff[\"ccds_id\"] = dff[\"other\"].str.split('ccds_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "                dff[\"exon_number\"] = dff[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        dff.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get gene ids\n",
    "        dff[\"gene_id\"] = dff[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get transcript ids\n",
    "        dff[\"transcript_id\"] = dff[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        dff[\"exon_number\"] = dff[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        if delete_other:\n",
    "            dff.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in dff.columns:\n",
    "        dff.loc[dff[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1117a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: calculate_cpm\n",
    "\n",
    "purpose: Calculate CPM for the each sample given\n",
    "\n",
    "input: Counts dataset\n",
    "\n",
    "output: Counts dataset with CPM columns as well\n",
    "'''\n",
    "\n",
    "def calculate_cpm(df, is_gene=False):\n",
    "\n",
    "    dff = df.copy()\n",
    "    \n",
    "    ## Set count columns if dataframe is gene counts\n",
    "    if is_gene:\n",
    "        count_columns = dff.columns[1:].tolist()\n",
    "    \n",
    "    ## Set count columns if dataframe is transcript counts\n",
    "    else:\n",
    "        count_columns = dff.columns[2:].tolist()\n",
    "\n",
    "    ## Loop through counts columns to calculate CPM and add to the dataframe\n",
    "    for col in count_columns:\n",
    "        \n",
    "        dff[col] = round(dff[col], 2)\n",
    "        cpm_name = col.replace(\"_counts\", \"_CPM\")\n",
    "        dff[cpm_name] = round(((dff[col]/(dff[col].sum())) * 1000000), 2)\n",
    "    \n",
    "    return dff  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4ed11391",
   "metadata": {},
   "outputs": [],
   "source": [
    "## define ggplot colors\n",
    "ggplot2_colors = [\"#F8766D\", \"#CD9600\", \"#7CAE00\", \"#00BE67\", \"#00BFC4\", \"#00A9FF\", \"#C77CFF\", \"#FF61CC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "59f349db",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open original reference\n",
    "original_ref = pd.read_csv(\"../../references/Homo_sapiens.GRCh38.107_ERCC.gtf\", header=None, delimiter=\"\\t\", low_memory=False, \n",
    "                       names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"], comment=\"#\")\n",
    "\n",
    "original_ref = original_ref.loc[~original_ref[\"chr\"].str.startswith(\"ERCC-\")]\n",
    "\n",
    "## Parse through reference to get gene names and ids\n",
    "orig_ref = original_ref.loc[original_ref[\"type\"]==\"gene\"].copy()\n",
    "orig_ref = parse_df_columns(orig_ref, is_ref=True)\n",
    "\n",
    "## Import disease relevant genes\n",
    "disease_relevant_genes = pd.read_csv(\"../../references/medically_relevant_genes_02-04-2023_UPDATED.tsv\", sep=\"\\t\")\n",
    "\n",
    "## Brain disease genes\n",
    "brain_disease_gene_ids = pd.read_csv(\"../../references/brain_disease_genes_only_IDs.tsv\", sep=\"\\t\")\n",
    "brain_disease_annotations = pd.read_csv(\"../../references/brain_disease_genes_with_disease.tsv\", sep=\"\\t\")\n",
    "\n",
    "## Import AD Genes\n",
    "ad_names = pd.read_csv(\"../../references/AD_gwas_genes.tsv\", sep=\"\\t\")\n",
    "\n",
    "## Create disease relevant list including chromosome\n",
    "disease_relevant_genes_annotated = disease_relevant_genes.merge(orig_ref[[\"gene_id\", \"gene_name\", \"chr\"]], \n",
    "                                                               how=\"inner\", on=[\"gene_id\", \"gene_name\"])\n",
    "\n",
    "## Create list of protein coding genes\n",
    "protein_coding_ref = orig_ref.loc[orig_ref[\"gene_biotype\"] == \"protein_coding\"].copy()\n",
    "\n",
    "## Import and parse through extended annotations\n",
    "ref = pd.read_csv(\"../../data/raw/nextflow_pipeline_output/bambu_discovery/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, comment=\"#\", names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "ref = ref.loc[~ref[\"chr\"].str.startswith(\"ERCC-\")]\n",
    "\n",
    "ref = parse_df_columns(ref, is_ref=False)\n",
    "\n",
    "ref_transcripts = ref.loc[ref[\"type\"] == \"transcript\"].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4cb3b2e",
   "metadata": {},
   "source": [
    "# - Data pre-processing (Our Data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "bb5df7d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import data for transcript level counts and fix column names\n",
    "\n",
    "df_ours = pd.read_csv(\"../../data/raw/nextflow_pipeline_output/bambu_discovery/counts_transcript.txt\", \n",
    "                           delimiter=\"\\t\", low_memory=False, header=0)\n",
    "\n",
    "df_ours = fix_column_names(df_ours, is_gene=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2a1c5c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate total counts\n",
    "df_ours[\"total_counts\"] = df_ours[df_ours.filter(regex='count').columns].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5e424d8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate CPM and drop count columns\n",
    "df_ours = calculate_cpm(df_ours, is_gene=False)\n",
    "\n",
    "df_ours = df_ours[df_ours.columns.drop(list(df_ours.filter(regex='counts')))].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c237b1c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Drop ERCCs\n",
    "df_ours = df_ours.loc[~df_ours[\"gene_id\"].str.startswith(\"ERCC\")].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "02cefca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate median CPM\n",
    "df_ours[\"median_CPM\"] = df_ours[df_ours.filter(regex='[0-9]_CPM').columns].median(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "3593fa3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Annotate transcritps in counts matrix\n",
    "df_ours = df_ours.merge(ref_transcripts[[\"gene_id\", \"transcript_id\", \"chr\"]], on=[\"gene_id\", \"transcript_id\"], how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3d942f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add relative transcript abundance to dataframe\n",
    "df_ours = relative_transcript_abundance(df_ours)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "bd4c14f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only keep genes and transcripts with median CPM > 1\n",
    "df_ours = df_ours.loc[df_ours[\"median_CPM\"] > 1].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "30527ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only keep high confidence mitochondrial transcripts\n",
    "df_ours_mito_high = df_ours.loc[df_ours[\"transcript_id\"].isin([\"BambuTx1845\", \"BambuTx1846\", \"BambuTx1847\", \"BambuTx1848\", \n",
    "                                                \"BambuTx1850\"])].copy()\n",
    "\n",
    "df_ours = df_ours.loc[~((df_ours[\"chr\"] == \"MT\") & (df_ours[\"transcript_id\"].str.startswith(\"BambuTx\")))].copy()\n",
    "\n",
    "df_ours = pd.concat([df_ours, df_ours_mito_high])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b5722fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only keep relevant columns\n",
    "df_ours = df_ours[[\"transcript_id\", \"gene_id\", \"chr\", \"median_CPM\",\n",
    "                   \"total_CPM_gene\", \"total_CPM_transcript\", \"relative_abundance_percent\"]].copy()\n",
    "\n",
    "## Rename columns to indicate they are from our data\n",
    "df_ours.columns = [\"transcript_id\", \"gene_id\", \"chr\", \"median_CPM_OURS\",\n",
    "                   \"total_CPM_gene_OURS\", \"total_CPM_transcript_OURS\", \"relative_abundance_percent_OURS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "06cce5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Separate dataframe in three (known, new_from_known, new_from_new)\n",
    "df_ours_known = df_ours.loc[~(df_ours[\"transcript_id\"].str.startswith(\"Bambu\"))].copy()\n",
    "\n",
    "df_ours_new_from_known = df_ours.loc[((df_ours[\"transcript_id\"].str.startswith(\"Bambu\")) & \n",
    "                                ~(df_ours[\"gene_id\"].str.startswith(\"Bambu\")))].copy()\n",
    "\n",
    "df_ours_new_from_new = df_ours.loc[((df_ours[\"transcript_id\"].str.startswith(\"Bambu\")) & \n",
    "                                (df_ours[\"gene_id\"].str.startswith(\"Bambu\")))].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9a63ad3",
   "metadata": {},
   "source": [
    "# - Data pre-processing (GTEX BA9 data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a45971a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import data for transcript level counts and fix column names\n",
    "\n",
    "df_gtex = pd.read_csv(\"../../data/raw/nextflow_GTEX_pipeline_output/bambu_discovery/counts_transcript.txt\", \n",
    "                           delimiter=\"\\t\", low_memory=False, header=0)\n",
    "\n",
    "## Only keep BA9 brain samples\n",
    "df_gtex = df_gtex[[\"TXNAME\", \"GENEID\",\"GTEX-1192X-0011-R10a-SM-4RXXZ\", \"GTEX-13X6J-0011-R10b-SM-5CEKT\", \n",
    "                   \"GTEX-14BIL-0011-R10a-SM-5EQV4\", \"GTEX-15DCD-0011-R10b-SM-5S51M\", \"GTEX-QDT8-0011-R10A-SM-2FKJB\",\n",
    "                   \"GTEX-T5JC-0011-R10A-SM-2TT23\"]]\n",
    "\n",
    "df_gtex = fix_column_names(df_gtex, is_gene=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "34d7ffe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate total counts\n",
    "df_gtex[\"total_counts\"] = df_gtex[df_gtex.filter(regex='count').columns].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a1f628f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate CPM and drop count columns\n",
    "df_gtex = calculate_cpm(df_gtex, is_gene=False)\n",
    "\n",
    "df_gtex = df_gtex[df_gtex.columns.drop(list(df_gtex.filter(regex='counts')))].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c8dd56a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Drop ERCCs\n",
    "df_gtex = df_gtex.loc[~df_gtex[\"gene_id\"].str.startswith(\"ERCC\")].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "56bb084e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate median CPM\n",
    "df_gtex[\"median_CPM\"] = df_gtex[df_gtex.filter(regex='[0-9]_CPM').columns].median(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8a8e4184",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Annotate transcritps in counts matrix\n",
    "df_gtex = df_gtex.merge(ref_transcripts[[\"gene_id\", \"transcript_id\", \"chr\"]], on=[\"gene_id\", \"transcript_id\"], how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "dfe7fcea",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add relative transcript abundance to dataframe\n",
    "df_gtex = relative_transcript_abundance(df_gtex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "105d9684",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only keep relevant columns\n",
    "df_gtex = df_gtex[[\"transcript_id\", \"gene_id\", \"chr\", \"median_CPM\",\n",
    "                   \"total_CPM_gene\", \"total_CPM_transcript\", \"relative_abundance_percent\"]].copy()\n",
    "\n",
    "## Rename columns to indicate they are from our data\n",
    "df_gtex.columns = [\"transcript_id\", \"gene_id\", \"chr\", \"median_CPM_GTEX\",\n",
    "                   \"total_CPM_gene_GTEX\", \"total_CPM_transcript_GTEX\", \"relative_abundance_percent_GTEX\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a49c05",
   "metadata": {},
   "source": [
    "# - Merge ours and GTEX data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e2e943",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_known = df_ours_known.merge(df_gtex, on=[\"transcript_id\", \"gene_id\", \"chr\"], how=\"inner\")\n",
    "\n",
    "df_new_from_known = df_ours_new_from_known.merge(df_gtex, on=[\"transcript_id\", \"gene_id\", \"chr\"], how=\"inner\")\n",
    "\n",
    "df_new_from_new = df_ours_new_from_new.merge(df_gtex, on=[\"transcript_id\", \"gene_id\", \"chr\"], how=\"inner\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f400860a",
   "metadata": {},
   "source": [
    "# - Make figures (Relative Abundance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "338a7437",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_known, x=\"relative_abundance_percent_GTEX\", y=\"relative_abundance_percent_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX relative abundance\")\n",
    "plt.xlabel(\"OURS relative abundance\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_known[\"relative_abundance_percent_GTEX\"].copy(),\n",
    "                                                                     df_known[\"relative_abundance_percent_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_known[\"relative_abundance_percent_GTEX\"].copy(),\n",
    "                                                                     df_known[\"relative_abundance_percent_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f50a8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_new_from_known, x=\"relative_abundance_percent_GTEX\", y=\"relative_abundance_percent_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX relative abundance\")\n",
    "plt.xlabel(\"OURS relative abundance\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_new_from_known[\"relative_abundance_percent_GTEX\"].copy(),\n",
    "                                                                     df_new_from_known[\"relative_abundance_percent_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_new_from_known[\"relative_abundance_percent_GTEX\"].copy(),\n",
    "                                                                     df_new_from_known[\"relative_abundance_percent_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71c24857",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_new_from_new, x=\"relative_abundance_percent_GTEX\", y=\"relative_abundance_percent_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX relative abundance\")\n",
    "plt.xlabel(\"OURS relative abundance\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_new_from_new[\"relative_abundance_percent_GTEX\"].copy(),\n",
    "                                                                     df_new_from_new[\"relative_abundance_percent_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_new_from_new[\"relative_abundance_percent_GTEX\"].copy(),\n",
    "                                                                     df_new_from_new[\"relative_abundance_percent_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1817242a",
   "metadata": {},
   "source": [
    "# - Make figures (median CPM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb56232b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_known, x=\"median_CPM_GTEX\", y=\"median_CPM_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX median CPM\")\n",
    "plt.xlabel(\"OURS median CPM\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_known[\"median_CPM_GTEX\"].copy(),\n",
    "                                                                     df_known[\"median_CPM_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_known[\"median_CPM_GTEX\"].copy(),\n",
    "                                                                     df_known[\"median_CPM_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb262d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_new_from_known, x=\"median_CPM_GTEX\", y=\"median_CPM_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX median CPM\")\n",
    "plt.xlabel(\"OURS median CPM\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_new_from_known[\"median_CPM_GTEX\"].copy(),\n",
    "                                                                     df_new_from_known[\"median_CPM_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_new_from_known[\"median_CPM_GTEX\"].copy(),\n",
    "                                                                     df_new_from_known[\"median_CPM_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd8fd6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_new_from_new, x=\"median_CPM_GTEX\", y=\"median_CPM_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX median CPM\")\n",
    "plt.xlabel(\"OURS median CPM\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_new_from_new[\"median_CPM_GTEX\"].copy(),\n",
    "                                                                     df_new_from_new[\"median_CPM_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_new_from_new[\"median_CPM_GTEX\"].copy(),\n",
    "                                                                     df_new_from_new[\"median_CPM_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d99947c",
   "metadata": {},
   "source": [
    "# - Make figures (total CPM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e270a706",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_known, x=\"total_CPM_transcript_GTEX\", y=\"total_CPM_transcript_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX total CPM\")\n",
    "plt.xlabel(\"OURS total CPM\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_known[\"total_CPM_transcript_GTEX\"].copy(),\n",
    "                                                                     df_known[\"total_CPM_transcript_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_known[\"total_CPM_transcript_GTEX\"].copy(),\n",
    "                                                                     df_known[\"total_CPM_transcript_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e44fb2b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_new_from_known, x=\"total_CPM_transcript_GTEX\", y=\"total_CPM_transcript_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX total CPM\")\n",
    "plt.xlabel(\"OURS total CPM\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_new_from_known[\"total_CPM_transcript_GTEX\"].copy(),\n",
    "                                                                     df_new_from_known[\"total_CPM_transcript_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_new_from_known[\"total_CPM_transcript_GTEX\"].copy(),\n",
    "                                                                     df_new_from_known[\"total_CPM_transcript_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbe7936d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.regplot(data=df_new_from_new, x=\"total_CPM_transcript_GTEX\", y=\"total_CPM_transcript_OURS\", color=ggplot2_colors[0], \n",
    "           scatter_kws={\"alpha\": 0.5, \"s\":5}, line_kws={\"linewidth\": 1})\n",
    "\n",
    "plt.ylabel(\"GTEX total CPM\")\n",
    "plt.xlabel(\"OURS total CPM\")\n",
    "\n",
    "spearman_coeff, spearman_p = stats.spearmanr(df_new_from_new[\"total_CPM_transcript_GTEX\"].copy(),\n",
    "                                                                     df_new_from_new[\"total_CPM_transcript_OURS\"].copy())\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(df_new_from_new[\"total_CPM_transcript_GTEX\"].copy(),\n",
    "                                                                     df_new_from_new[\"total_CPM_transcript_OURS\"].copy())\n",
    "\n",
    "print(\"Spearman coefficient:\", str(round(spearman_coeff, 2)))\n",
    "print(\"Spearman R-squared:\", str(round((spearman_coeff * spearman_coeff), 2)))\n",
    "print(\"Spearman p-value:\", str(spearman_p))\n",
    "\n",
    "print(\"\\nPearson coefficient:\", str(round(r_value, 2)))\n",
    "print(\"Pearson R-squared:\", str(round(r_value*r_value,2)))\n",
    "print(\"Pearson Slope:\", str(round(slope,2)))\n",
    "print(\"Pearson Intercept:\", str(round(intercept,2)))\n",
    "print(\"Pearson P-value:\", str(p_value))\n",
    "print(\"Pearson Standard Error:\", str(round(std_err,4)), \"\\n\")\n",
    "\n",
    "#plt.savefig('../../figures/paper_figures/supplement/ercc_correlation_by_mean_cpm.pdf',\n",
    "#            dpi=600, transparent=True, bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
