{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "59f6e7c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "45b26dc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_df_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True):\n",
    "\n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split(\"source_gene=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split(\"source_transcript=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get CHM gene_ids\n",
    "        df[\"CHM_gene_id\"] = df[\"other\"].str.split(\"gene_id=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"CHM_transcript_id\"] = df[\"other\"].str.split(\"transcript_id=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get transcript names\n",
    "        df[\"transcript_name\"] = df[\"other\"].str.split(\"source_transcript_name=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get gene names\n",
    "        df[\"gene_name\"] = df[\"other\"].str.split(\"source_gene_common_name=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get start codon\n",
    "        df[\"start_codon\"] = df[\"other\"].str.split(\"adj_start=\", expand=True)[1].str.split(\";\", expand=True)[0]\n",
    "        \n",
    "        ## Get stop codon\n",
    "        df[\"stop_codon\"] = df[\"other\"].str.split(\"adj_stop=\", expand=True)[1].str.split(\";\", expand=True)[0]        \n",
    "\n",
    "        ## Only keep relevant\n",
    "        df = df[[\"chr\", \"start\", \"end\", \"strand\", \"type\", \"gene_id\", \"transcript_id\", \"CHM_gene_id\",\n",
    "                 \"CHM_transcript_id\", \"transcript_name\", \"gene_name\", \"start_codon\", \"stop_codon\"]].copy()\n",
    "\n",
    "        ## Drop duplicates\n",
    "        df.drop_duplicates(inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get CHM gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get CHM transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Label novel transcripts\n",
    "        df.loc[df[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "        df.loc[~df[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "        ## Label novel genes\n",
    "        df.loc[df[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "        df.loc[~df[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in df.columns:\n",
    "        df.loc[df[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e17280aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation = pd.read_csv(\"C:/Users/bag22/Desktop/current_files/python_files/ebbert_lab/data/2022-07-14_cDNA_data_ebbert_lab_nextflow_pipeline_output_raw/ONT_only_data/GRCh38-106_good_ONT/bambu_discovery/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b1d1061f",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation = parse_df_columns(bambu_annotation, is_ref=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d8c3fd7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation_mt = bambu_annotation.loc[bambu_annotation[\"chr\"]==\"MT\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "19df5cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation = pd.read_csv(\"C:/Users/bag22/Desktop/2022_ebbert_generated_ONT_only_data/GRCh38-106_good_ONT/bambu_discovery/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9ded8b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation_final = bambu_annotation.merge(bambu_annotation_mt[[\"chr\", \"source\", \"type\", \"start\", \"end\", \"strand\"]], on=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"strand\"], how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "daf20110",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation_final[\"chr\"] = \"chrM\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b6eb0e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "bambu_annotation_final.to_csv(\"~/Desktop/grch38_mt_only.gtf\", sep=\"\\t\", index=False, header=False, quoting=csv.QUOTE_NONE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "260c14b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>13670.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>12057.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr source        type    start      end strand          gene_id  \\\n",
       "0   1  Bambu  transcript  11869.0  14409.0      +  ENSG00000223972   \n",
       "1   1  Bambu        exon  11869.0  12227.0      +  ENSG00000223972   \n",
       "2   1  Bambu  transcript  12010.0  13670.0      +  ENSG00000223972   \n",
       "3   1  Bambu        exon  12010.0  12057.0      +  ENSG00000223972   \n",
       "4   1  Bambu        exon  12179.0  12227.0      +  ENSG00000223972   \n",
       "\n",
       "     transcript_id exon_number is_novel_transcript is_novel_gene  \n",
       "0  ENST00000456328         NaN               False         False  \n",
       "1  ENST00000456328           1               False         False  \n",
       "2  ENST00000450305         NaN               False         False  \n",
       "3  ENST00000450305           1               False         False  \n",
       "4  ENST00000450305           2               False         False  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bambu_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3a1965bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_mt_genes = bambu_annotation.loc[bambu_annotation[\"chr\"]==\"MT\"][\"gene_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3a7f0179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1758967    ENSG00000210049\n",
       "1758968    ENSG00000210049\n",
       "1758969    ENSG00000211459\n",
       "1758970    ENSG00000211459\n",
       "1758971    ENSG00000210077\n",
       "1758972    ENSG00000210077\n",
       "1758973    ENSG00000210082\n",
       "1758974    ENSG00000210082\n",
       "1758975    ENSG00000210107\n",
       "1758976    ENSG00000210107\n",
       "1758977    ENSG00000209082\n",
       "1758978    ENSG00000209082\n",
       "1758979    ENSG00000198888\n",
       "1758980    ENSG00000198888\n",
       "1758981    ENSG00000210100\n",
       "1758982    ENSG00000210100\n",
       "1758983    ENSG00000210107\n",
       "1758984    ENSG00000210107\n",
       "1758985    ENSG00000210107\n",
       "1758986    ENSG00000210112\n",
       "1758987    ENSG00000210112\n",
       "1758988    ENSG00000198763\n",
       "1758989    ENSG00000198763\n",
       "1758990    ENSG00000210117\n",
       "1758991    ENSG00000210117\n",
       "1758992    ENSG00000210127\n",
       "1758993    ENSG00000210127\n",
       "1758994    ENSG00000210135\n",
       "1758995    ENSG00000210135\n",
       "1758996    ENSG00000210140\n",
       "1758997    ENSG00000210140\n",
       "1758998    ENSG00000210144\n",
       "1758999    ENSG00000210144\n",
       "1759000    ENSG00000198804\n",
       "1759001    ENSG00000198804\n",
       "1759002    ENSG00000210151\n",
       "1759003    ENSG00000210151\n",
       "1759004    ENSG00000210151\n",
       "1759005    ENSG00000210151\n",
       "1759006    ENSG00000210151\n",
       "1759007    ENSG00000210154\n",
       "1759008    ENSG00000210154\n",
       "1759009    ENSG00000198712\n",
       "1759010    ENSG00000198712\n",
       "1759011    ENSG00000210156\n",
       "1759012    ENSG00000210156\n",
       "1759013    ENSG00000228253\n",
       "1759014    ENSG00000228253\n",
       "1759015    ENSG00000198899\n",
       "1759016    ENSG00000198899\n",
       "1759017    ENSG00000198938\n",
       "1759018    ENSG00000198938\n",
       "1759019    ENSG00000210164\n",
       "1759020    ENSG00000210164\n",
       "1759021    ENSG00000198840\n",
       "1759022    ENSG00000198840\n",
       "1759023    ENSG00000210174\n",
       "1759024    ENSG00000210174\n",
       "1759025    ENSG00000212907\n",
       "1759026    ENSG00000212907\n",
       "1759027    ENSG00000198886\n",
       "1759028    ENSG00000198886\n",
       "1759029    ENSG00000210176\n",
       "1759030    ENSG00000210176\n",
       "1759031    ENSG00000210184\n",
       "1759032    ENSG00000210184\n",
       "1759033    ENSG00000210191\n",
       "1759034    ENSG00000210191\n",
       "1759035    ENSG00000198786\n",
       "1759036    ENSG00000198786\n",
       "1759037    ENSG00000198695\n",
       "1759038    ENSG00000198695\n",
       "1759039    ENSG00000210194\n",
       "1759040    ENSG00000210194\n",
       "1759041    ENSG00000198727\n",
       "1759042    ENSG00000198727\n",
       "1759043    ENSG00000210195\n",
       "1759044    ENSG00000210195\n",
       "1759045    ENSG00000210196\n",
       "1759046    ENSG00000210196\n",
       "Name: gene_id, dtype: object"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bambu_mt_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a3784f06",
   "metadata": {},
   "outputs": [],
   "source": [
    "grch_counts = pd.read_csv(\"C:/Users/bag22/Desktop/2022_ebbert_generated_ONT_only_data/GRCh38-106_good_ONT/bambu_discovery/counts_transcript.txt\",\n",
    "                         sep=\"\\t\")\n",
    "\n",
    "new_col_names = []\n",
    "for col in grch_counts.columns:\n",
    "        new_col_names.append(col.split(\"_nanopore\")[0])\n",
    "        \n",
    "grch_counts.columns = new_col_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a6dd805f",
   "metadata": {},
   "outputs": [],
   "source": [
    "grch_counts[\"total_counts\"] = grch_counts[[\"sample_PAM54401\", \"sample_PAM54902\", \"sample_PAM54335\", \"sample_PAM54788\"]].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "17277832",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TXNAME</th>\n",
       "      <th>GENEID</th>\n",
       "      <th>sample_PAM54401</th>\n",
       "      <th>sample_PAM54902</th>\n",
       "      <th>sample_PAM54335</th>\n",
       "      <th>sample_PAM54788</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>tx.755</td>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>8.370439e+05</td>\n",
       "      <td>1.193660e+06</td>\n",
       "      <td>6.846400e+05</td>\n",
       "      <td>9.367790e+05</td>\n",
       "      <td>3.652123e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>755</th>\n",
       "      <td>tx.756</td>\n",
       "      <td>ENSG00000210151</td>\n",
       "      <td>4.402299e+04</td>\n",
       "      <td>4.134357e+04</td>\n",
       "      <td>2.580772e+04</td>\n",
       "      <td>5.690252e+04</td>\n",
       "      <td>1.680768e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18712</th>\n",
       "      <td>ENST00000361227</td>\n",
       "      <td>ENSG00000198840</td>\n",
       "      <td>5.254800e+04</td>\n",
       "      <td>1.373710e+05</td>\n",
       "      <td>9.863300e+04</td>\n",
       "      <td>9.180600e+04</td>\n",
       "      <td>3.803580e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18757</th>\n",
       "      <td>ENST00000361335</td>\n",
       "      <td>ENSG00000212907</td>\n",
       "      <td>1.424500e+04</td>\n",
       "      <td>4.107700e+04</td>\n",
       "      <td>9.262000e+03</td>\n",
       "      <td>9.968600e+04</td>\n",
       "      <td>1.642700e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18777</th>\n",
       "      <td>ENST00000361381</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>4.533990e+05</td>\n",
       "      <td>6.680680e+05</td>\n",
       "      <td>3.432920e+05</td>\n",
       "      <td>5.947410e+05</td>\n",
       "      <td>2.059500e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18782</th>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>6.635120e+05</td>\n",
       "      <td>1.336354e+06</td>\n",
       "      <td>4.721360e+05</td>\n",
       "      <td>9.798480e+05</td>\n",
       "      <td>3.451850e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18809</th>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>7.392900e+04</td>\n",
       "      <td>1.297140e+05</td>\n",
       "      <td>1.279610e+05</td>\n",
       "      <td>1.884430e+05</td>\n",
       "      <td>5.200470e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>ENST00000361567</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>1.277860e+05</td>\n",
       "      <td>2.123370e+05</td>\n",
       "      <td>1.172010e+05</td>\n",
       "      <td>2.080980e+05</td>\n",
       "      <td>6.654220e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18890</th>\n",
       "      <td>ENST00000361624</td>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>3.937200e+05</td>\n",
       "      <td>5.720560e+05</td>\n",
       "      <td>2.580670e+05</td>\n",
       "      <td>4.709160e+05</td>\n",
       "      <td>1.694759e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18916</th>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>7.356900e+04</td>\n",
       "      <td>9.795600e+04</td>\n",
       "      <td>5.485600e+04</td>\n",
       "      <td>9.695100e+04</td>\n",
       "      <td>3.233320e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18951</th>\n",
       "      <td>ENST00000361739</td>\n",
       "      <td>ENSG00000198712</td>\n",
       "      <td>3.712930e+05</td>\n",
       "      <td>6.305460e+05</td>\n",
       "      <td>2.569180e+05</td>\n",
       "      <td>7.088300e+05</td>\n",
       "      <td>1.967587e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18971</th>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>2.525380e+05</td>\n",
       "      <td>4.180940e+05</td>\n",
       "      <td>2.666600e+05</td>\n",
       "      <td>4.546700e+05</td>\n",
       "      <td>1.391962e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19002</th>\n",
       "      <td>ENST00000361851</td>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>8.794000e+03</td>\n",
       "      <td>1.463000e+04</td>\n",
       "      <td>8.831000e+03</td>\n",
       "      <td>3.470900e+04</td>\n",
       "      <td>6.696400e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19015</th>\n",
       "      <td>ENST00000361899</td>\n",
       "      <td>ENSG00000198899</td>\n",
       "      <td>2.076130e+05</td>\n",
       "      <td>4.556800e+05</td>\n",
       "      <td>2.243260e+05</td>\n",
       "      <td>3.826350e+05</td>\n",
       "      <td>1.270254e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19092</th>\n",
       "      <td>ENST00000362079</td>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>2.851370e+05</td>\n",
       "      <td>3.399640e+05</td>\n",
       "      <td>2.541080e+05</td>\n",
       "      <td>4.883350e+05</td>\n",
       "      <td>1.367544e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30106</th>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>1.930000e+02</td>\n",
       "      <td>5.570000e+02</td>\n",
       "      <td>1.240000e+02</td>\n",
       "      <td>1.230000e+02</td>\n",
       "      <td>9.970000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30115</th>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>1.734000e+03</td>\n",
       "      <td>1.026800e+04</td>\n",
       "      <td>1.452000e+03</td>\n",
       "      <td>5.396700e+04</td>\n",
       "      <td>6.742100e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30116</th>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>9.900000e+01</td>\n",
       "      <td>1.030000e+02</td>\n",
       "      <td>8.100000e+01</td>\n",
       "      <td>7.500000e+01</td>\n",
       "      <td>3.580000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30117</th>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>3.648791e+06</td>\n",
       "      <td>5.218008e+06</td>\n",
       "      <td>3.245524e+06</td>\n",
       "      <td>5.521725e+06</td>\n",
       "      <td>1.763405e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30118</th>\n",
       "      <td>ENST00000387365</td>\n",
       "      <td>ENSG00000210100</td>\n",
       "      <td>1.550000e+02</td>\n",
       "      <td>4.300000e+02</td>\n",
       "      <td>6.500000e+01</td>\n",
       "      <td>9.800000e+01</td>\n",
       "      <td>7.480000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30119</th>\n",
       "      <td>ENST00000387372</td>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>6.705125e+03</td>\n",
       "      <td>2.424582e+04</td>\n",
       "      <td>8.960039e+02</td>\n",
       "      <td>8.080017e+02</td>\n",
       "      <td>3.265495e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30120</th>\n",
       "      <td>ENST00000387377</td>\n",
       "      <td>ENSG00000210112</td>\n",
       "      <td>9.500000e+01</td>\n",
       "      <td>4.200000e+02</td>\n",
       "      <td>8.600000e+01</td>\n",
       "      <td>9.300000e+01</td>\n",
       "      <td>6.940000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30121</th>\n",
       "      <td>ENST00000387382</td>\n",
       "      <td>ENSG00000210117</td>\n",
       "      <td>1.251500e+04</td>\n",
       "      <td>5.544100e+04</td>\n",
       "      <td>1.530500e+04</td>\n",
       "      <td>1.380400e+04</td>\n",
       "      <td>9.706500e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30122</th>\n",
       "      <td>ENST00000387392</td>\n",
       "      <td>ENSG00000210127</td>\n",
       "      <td>6.040000e+02</td>\n",
       "      <td>4.900000e+01</td>\n",
       "      <td>1.057000e+03</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>1.723000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30123</th>\n",
       "      <td>ENST00000387400</td>\n",
       "      <td>ENSG00000210135</td>\n",
       "      <td>9.600000e+01</td>\n",
       "      <td>9.300000e+01</td>\n",
       "      <td>7.500000e+01</td>\n",
       "      <td>7.700000e+01</td>\n",
       "      <td>3.410000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30124</th>\n",
       "      <td>ENST00000387405</td>\n",
       "      <td>ENSG00000210140</td>\n",
       "      <td>1.900000e+01</td>\n",
       "      <td>1.000000e+01</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>1.100000e+01</td>\n",
       "      <td>4.900000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30125</th>\n",
       "      <td>ENST00000387409</td>\n",
       "      <td>ENSG00000210144</td>\n",
       "      <td>9.400000e+01</td>\n",
       "      <td>1.420000e+02</td>\n",
       "      <td>1.040000e+02</td>\n",
       "      <td>1.390000e+02</td>\n",
       "      <td>4.790000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30126</th>\n",
       "      <td>ENST00000387416</td>\n",
       "      <td>ENSG00000210151</td>\n",
       "      <td>6.126011e+03</td>\n",
       "      <td>1.674431e+03</td>\n",
       "      <td>1.109276e+03</td>\n",
       "      <td>2.788479e+03</td>\n",
       "      <td>1.169820e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30127</th>\n",
       "      <td>ENST00000387419</td>\n",
       "      <td>ENSG00000210154</td>\n",
       "      <td>4.800000e+01</td>\n",
       "      <td>8.000000e+01</td>\n",
       "      <td>2.100000e+01</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>1.580000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30128</th>\n",
       "      <td>ENST00000387421</td>\n",
       "      <td>ENSG00000210156</td>\n",
       "      <td>6.347000e+03</td>\n",
       "      <td>3.657700e+04</td>\n",
       "      <td>8.236000e+03</td>\n",
       "      <td>9.000000e+01</td>\n",
       "      <td>5.125000e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30129</th>\n",
       "      <td>ENST00000387429</td>\n",
       "      <td>ENSG00000210164</td>\n",
       "      <td>2.700000e+01</td>\n",
       "      <td>3.800000e+01</td>\n",
       "      <td>2.400000e+01</td>\n",
       "      <td>2.500000e+01</td>\n",
       "      <td>1.140000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30130</th>\n",
       "      <td>ENST00000387439</td>\n",
       "      <td>ENSG00000210174</td>\n",
       "      <td>4.100000e+01</td>\n",
       "      <td>5.200000e+01</td>\n",
       "      <td>2.400000e+01</td>\n",
       "      <td>2.900000e+01</td>\n",
       "      <td>1.460000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30131</th>\n",
       "      <td>ENST00000387441</td>\n",
       "      <td>ENSG00000210176</td>\n",
       "      <td>1.924000e+03</td>\n",
       "      <td>4.836000e+03</td>\n",
       "      <td>1.932000e+03</td>\n",
       "      <td>4.260000e+03</td>\n",
       "      <td>1.295200e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30133</th>\n",
       "      <td>ENST00000387449</td>\n",
       "      <td>ENSG00000210184</td>\n",
       "      <td>7.000000e+00</td>\n",
       "      <td>1.599400e+04</td>\n",
       "      <td>3.200000e+01</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>1.604600e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30134</th>\n",
       "      <td>ENST00000387456</td>\n",
       "      <td>ENSG00000210191</td>\n",
       "      <td>3.600000e+01</td>\n",
       "      <td>4.900000e+01</td>\n",
       "      <td>4.100000e+01</td>\n",
       "      <td>6.300000e+01</td>\n",
       "      <td>1.890000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30135</th>\n",
       "      <td>ENST00000387459</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>6.310000e+02</td>\n",
       "      <td>9.960000e+02</td>\n",
       "      <td>3.350000e+02</td>\n",
       "      <td>4.060000e+02</td>\n",
       "      <td>2.368000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30136</th>\n",
       "      <td>ENST00000387460</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>1.869000e+03</td>\n",
       "      <td>5.000000e+00</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>1.885000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30137</th>\n",
       "      <td>ENST00000387461</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>1.161200e+04</td>\n",
       "      <td>3.417000e+03</td>\n",
       "      <td>1.448000e+03</td>\n",
       "      <td>2.553200e+04</td>\n",
       "      <td>4.200900e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30350</th>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>6.544570e+05</td>\n",
       "      <td>1.034675e+06</td>\n",
       "      <td>5.977870e+05</td>\n",
       "      <td>1.058385e+06</td>\n",
       "      <td>3.345304e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                TXNAME           GENEID  sample_PAM54401  sample_PAM54902  \\\n",
       "754             tx.755  ENSG00000210107     8.370439e+05     1.193660e+06   \n",
       "755             tx.756  ENSG00000210151     4.402299e+04     4.134357e+04   \n",
       "18712  ENST00000361227  ENSG00000198840     5.254800e+04     1.373710e+05   \n",
       "18757  ENST00000361335  ENSG00000212907     1.424500e+04     4.107700e+04   \n",
       "18777  ENST00000361381  ENSG00000198886     4.533990e+05     6.680680e+05   \n",
       "18782  ENST00000361390  ENSG00000198888     6.635120e+05     1.336354e+06   \n",
       "18809  ENST00000361453  ENSG00000198763     7.392900e+04     1.297140e+05   \n",
       "18866  ENST00000361567  ENSG00000198786     1.277860e+05     2.123370e+05   \n",
       "18890  ENST00000361624  ENSG00000198804     3.937200e+05     5.720560e+05   \n",
       "18916  ENST00000361681  ENSG00000198695     7.356900e+04     9.795600e+04   \n",
       "18951  ENST00000361739  ENSG00000198712     3.712930e+05     6.305460e+05   \n",
       "18971  ENST00000361789  ENSG00000198727     2.525380e+05     4.180940e+05   \n",
       "19002  ENST00000361851  ENSG00000228253     8.794000e+03     1.463000e+04   \n",
       "19015  ENST00000361899  ENSG00000198899     2.076130e+05     4.556800e+05   \n",
       "19092  ENST00000362079  ENSG00000198938     2.851370e+05     3.399640e+05   \n",
       "30106  ENST00000386347  ENSG00000209082     1.930000e+02     5.570000e+02   \n",
       "30115  ENST00000387314  ENSG00000210049     1.734000e+03     1.026800e+04   \n",
       "30116  ENST00000387342  ENSG00000210077     9.900000e+01     1.030000e+02   \n",
       "30117  ENST00000387347  ENSG00000210082     3.648791e+06     5.218008e+06   \n",
       "30118  ENST00000387365  ENSG00000210100     1.550000e+02     4.300000e+02   \n",
       "30119  ENST00000387372  ENSG00000210107     6.705125e+03     2.424582e+04   \n",
       "30120  ENST00000387377  ENSG00000210112     9.500000e+01     4.200000e+02   \n",
       "30121  ENST00000387382  ENSG00000210117     1.251500e+04     5.544100e+04   \n",
       "30122  ENST00000387392  ENSG00000210127     6.040000e+02     4.900000e+01   \n",
       "30123  ENST00000387400  ENSG00000210135     9.600000e+01     9.300000e+01   \n",
       "30124  ENST00000387405  ENSG00000210140     1.900000e+01     1.000000e+01   \n",
       "30125  ENST00000387409  ENSG00000210144     9.400000e+01     1.420000e+02   \n",
       "30126  ENST00000387416  ENSG00000210151     6.126011e+03     1.674431e+03   \n",
       "30127  ENST00000387419  ENSG00000210154     4.800000e+01     8.000000e+01   \n",
       "30128  ENST00000387421  ENSG00000210156     6.347000e+03     3.657700e+04   \n",
       "30129  ENST00000387429  ENSG00000210164     2.700000e+01     3.800000e+01   \n",
       "30130  ENST00000387439  ENSG00000210174     4.100000e+01     5.200000e+01   \n",
       "30131  ENST00000387441  ENSG00000210176     1.924000e+03     4.836000e+03   \n",
       "30133  ENST00000387449  ENSG00000210184     7.000000e+00     1.599400e+04   \n",
       "30134  ENST00000387456  ENSG00000210191     3.600000e+01     4.900000e+01   \n",
       "30135  ENST00000387459  ENSG00000210194     6.310000e+02     9.960000e+02   \n",
       "30136  ENST00000387460  ENSG00000210195     2.000000e+00     1.869000e+03   \n",
       "30137  ENST00000387461  ENSG00000210196     1.161200e+04     3.417000e+03   \n",
       "30350  ENST00000389680  ENSG00000211459     6.544570e+05     1.034675e+06   \n",
       "\n",
       "       sample_PAM54335  sample_PAM54788  total_counts  \n",
       "754       6.846400e+05     9.367790e+05  3.652123e+06  \n",
       "755       2.580772e+04     5.690252e+04  1.680768e+05  \n",
       "18712     9.863300e+04     9.180600e+04  3.803580e+05  \n",
       "18757     9.262000e+03     9.968600e+04  1.642700e+05  \n",
       "18777     3.432920e+05     5.947410e+05  2.059500e+06  \n",
       "18782     4.721360e+05     9.798480e+05  3.451850e+06  \n",
       "18809     1.279610e+05     1.884430e+05  5.200470e+05  \n",
       "18866     1.172010e+05     2.080980e+05  6.654220e+05  \n",
       "18890     2.580670e+05     4.709160e+05  1.694759e+06  \n",
       "18916     5.485600e+04     9.695100e+04  3.233320e+05  \n",
       "18951     2.569180e+05     7.088300e+05  1.967587e+06  \n",
       "18971     2.666600e+05     4.546700e+05  1.391962e+06  \n",
       "19002     8.831000e+03     3.470900e+04  6.696400e+04  \n",
       "19015     2.243260e+05     3.826350e+05  1.270254e+06  \n",
       "19092     2.541080e+05     4.883350e+05  1.367544e+06  \n",
       "30106     1.240000e+02     1.230000e+02  9.970000e+02  \n",
       "30115     1.452000e+03     5.396700e+04  6.742100e+04  \n",
       "30116     8.100000e+01     7.500000e+01  3.580000e+02  \n",
       "30117     3.245524e+06     5.521725e+06  1.763405e+07  \n",
       "30118     6.500000e+01     9.800000e+01  7.480000e+02  \n",
       "30119     8.960039e+02     8.080017e+02  3.265495e+04  \n",
       "30120     8.600000e+01     9.300000e+01  6.940000e+02  \n",
       "30121     1.530500e+04     1.380400e+04  9.706500e+04  \n",
       "30122     1.057000e+03     1.300000e+01  1.723000e+03  \n",
       "30123     7.500000e+01     7.700000e+01  3.410000e+02  \n",
       "30124     9.000000e+00     1.100000e+01  4.900000e+01  \n",
       "30125     1.040000e+02     1.390000e+02  4.790000e+02  \n",
       "30126     1.109276e+03     2.788479e+03  1.169820e+04  \n",
       "30127     2.100000e+01     9.000000e+00  1.580000e+02  \n",
       "30128     8.236000e+03     9.000000e+01  5.125000e+04  \n",
       "30129     2.400000e+01     2.500000e+01  1.140000e+02  \n",
       "30130     2.400000e+01     2.900000e+01  1.460000e+02  \n",
       "30131     1.932000e+03     4.260000e+03  1.295200e+04  \n",
       "30133     3.200000e+01     1.300000e+01  1.604600e+04  \n",
       "30134     4.100000e+01     6.300000e+01  1.890000e+02  \n",
       "30135     3.350000e+02     4.060000e+02  2.368000e+03  \n",
       "30136     5.000000e+00     9.000000e+00  1.885000e+03  \n",
       "30137     1.448000e+03     2.553200e+04  4.200900e+04  \n",
       "30350     5.977870e+05     1.058385e+06  3.345304e+06  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch_counts.loc[grch_counts[\"GENEID\"].isin(bambu_mt_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "720cb446",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "15093350",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation_grch = pd.read_csv(\"C:/Users/bag22/Desktop/current_files/python_files/ebbert_lab/data/2022-07-14_cDNA_data_ebbert_lab_nextflow_pipeline_output_raw/ONT_only_data/GRCh38-106_good_ONT/bambu_discovery/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "bambu_annotation_grch = parse_df_columns(bambu_annotation_grch, is_ref=False)\n",
    "\n",
    "bambu_annotation_grch_mt = bambu_annotation_grch.loc[bambu_annotation_grch[\"chr\"]==\"MT\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "04c64f8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bambu_annotation_chm13 = pd.read_csv(\"C:/Users/bag22/Desktop/current_files/python_files/ebbert_lab/data/2022-07-14_cDNA_data_ebbert_lab_nextflow_pipeline_output_raw/ONT_only_data/CHM13_good_ONT/bambu_discovery/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "bambu_annotation_chm13 = parse_df_columns(bambu_annotation_chm13, is_ref=False)\n",
    "\n",
    "bambu_annotation_chm13_mt = bambu_annotation_chm13.loc[bambu_annotation_chm13[\"chr\"]==\"chrM\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ec23cece",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1758975</th>\n",
       "      <td>MT</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1689.0</td>\n",
       "      <td>5449.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>tx.755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1758976</th>\n",
       "      <td>MT</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>1689.0</td>\n",
       "      <td>3502.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>tx.755</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1758985</th>\n",
       "      <td>MT</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>4380.0</td>\n",
       "      <td>5449.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>tx.755</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr source        type   start     end strand          gene_id  \\\n",
       "1758975  MT  Bambu  transcript  1689.0  5449.0      -  ENSG00000210107   \n",
       "1758976  MT  Bambu        exon  1689.0  3502.0      -  ENSG00000210107   \n",
       "1758985  MT  Bambu        exon  4380.0  5449.0      -  ENSG00000210107   \n",
       "\n",
       "        transcript_id exon_number is_novel_transcript is_novel_gene  \n",
       "1758975        tx.755         NaN                True         False  \n",
       "1758976        tx.755           2                True         False  \n",
       "1758985        tx.755           1                True         False  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bambu_annotation_grch_mt.loc[bambu_annotation_grch_mt['transcript_id']==\"tx.755\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "414a7baf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1623754</th>\n",
       "      <td>chrM</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1113.0</td>\n",
       "      <td>4872.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210107.1</td>\n",
       "      <td>tx.783</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1623755</th>\n",
       "      <td>chrM</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>1113.0</td>\n",
       "      <td>2925.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210107.1</td>\n",
       "      <td>tx.783</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1623765</th>\n",
       "      <td>chrM</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>3803.0</td>\n",
       "      <td>4872.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210107.1</td>\n",
       "      <td>tx.783</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr source        type   start     end strand            gene_id  \\\n",
       "1623754  chrM  Bambu  transcript  1113.0  4872.0      -  ENSG00000210107.1   \n",
       "1623755  chrM  Bambu        exon  1113.0  2925.0      -  ENSG00000210107.1   \n",
       "1623765  chrM  Bambu        exon  3803.0  4872.0      -  ENSG00000210107.1   \n",
       "\n",
       "        transcript_id exon_number is_novel_transcript is_novel_gene  \n",
       "1623754        tx.783         NaN                True         False  \n",
       "1623755        tx.783           2                True         False  \n",
       "1623765        tx.783           1                True         False  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bambu_annotation_chm13_mt.loc[bambu_annotation_chm13_mt['transcript_id']==\"tx.783\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
