{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_df_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True):\n",
    "\n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split(\"source_gene=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split(\"source_transcript=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get CHM gene_ids\n",
    "        df[\"CHM_gene_id\"] = df[\"other\"].str.split(\"gene_id=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"CHM_transcript_id\"] = df[\"other\"].str.split(\"transcript_id=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get transcript names\n",
    "        df[\"transcript_name\"] = df[\"other\"].str.split(\"source_transcript_name=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get gene names\n",
    "        df[\"gene_name\"] = df[\"other\"].str.split(\"source_gene_common_name=\", expand=True)[1].str.split(';', expand=True)[0]\n",
    "        \n",
    "        ## Get start codon\n",
    "        df[\"start_codon\"] = df[\"other\"].str.split(\"adj_start=\", expand=True)[1].str.split(\";\", expand=True)[0]\n",
    "        \n",
    "        ## Get stop codon\n",
    "        df[\"stop_codon\"] = df[\"other\"].str.split(\"adj_stop=\", expand=True)[1].str.split(\";\", expand=True)[0]        \n",
    "\n",
    "        ## Only keep relevant\n",
    "        df = df[[\"chr\", \"start\", \"end\", \"strand\", \"type\", \"gene_id\", \"transcript_id\", \"CHM_gene_id\",\n",
    "                 \"CHM_transcript_id\", \"transcript_name\", \"gene_name\", \"start_codon\", \"stop_codon\"]].copy()\n",
    "\n",
    "        ## Drop duplicates\n",
    "        df.drop_duplicates(inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get CHM gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get CHM transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Label novel transcripts\n",
    "        df.loc[df[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "        df.loc[~df[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "        ## Label novel genes\n",
    "        df.loc[df[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "        df.loc[~df[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in df.columns:\n",
    "        df.loc[df[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: merge_annotations\n",
    "\n",
    "purpose: Merge useful/relevant information from both annotations while removing repeated and irrelevant information\n",
    "\n",
    "input: Two different GTF annotations\n",
    "\n",
    "output: One GTF annotation containing all the relevant information\n",
    "'''\n",
    "\n",
    "def merge_annotations(ref_gtf, bambu_gtf):\n",
    "    \n",
    "    ## Merge the two annotations\n",
    "    names_ref_gtf = ref_gtf[[\"transcript_id\", \"gene_id\", \"gene_name\", \"transcript_name\"]].copy()\n",
    "    merged_gtf = pd.merge(bambu_gtf, names_ref_gtf, on=['gene_id', 'transcript_id'], how='left')\n",
    "    merged_gtf.drop_duplicates(inplace=True)\n",
    "\n",
    "    ## Label novel transcripts\n",
    "    merged_gtf.loc[merged_gtf[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "    merged_gtf.loc[~merged_gtf[\"transcript_id\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "    ## Label novel genes\n",
    "    merged_gtf.loc[merged_gtf[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "    merged_gtf.loc[~merged_gtf[\"gene_id\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False\n",
    "\n",
    "    ## Create temporary variable only containing novel transcripts\n",
    "    temp = merged_gtf.loc[merged_gtf[\"is_novel_transcript\"] == True]\n",
    "\n",
    "    ## Annotate novel transcripts\n",
    "    merged_tmp = pd.merge(temp, ref_gtf[[\"gene_id\", \"gene_name\"]], on=['gene_id'], how='left')\n",
    "    merged_tmp.drop_duplicates(inplace=True)\n",
    "    merged_tmp[\"gene_name\"] = merged_tmp[\"gene_name_y\"]\n",
    "    merged_tmp.drop(columns=[\"source\", \"gene_name_y\", \"gene_name_x\"], inplace=True)\n",
    "\n",
    "    ## Return novel transcripts to original annotation\n",
    "    merged_final = pd.merge(merged_gtf, merged_tmp, on=['chr', 'type', 'start', 'end', 'strand', 'transcript_id',\n",
    "                    'transcript_name', 'gene_id', 'is_novel_transcript', 'is_novel_gene', 'exon_number'], how=\"left\")\n",
    "\n",
    "    ## Get gene names for novel transcripts of known genes\n",
    "    merged_final.gene_name_x.fillna(merged_final.gene_name_y, inplace=True)\n",
    "    merged_final[\"gene_name\"] = merged_final[\"gene_name_x\"]\n",
    "    merged_final.drop(columns =[\"gene_name_x\", \"gene_name_y\"], inplace=True)\n",
    "    \n",
    "    ## Get start and stop codons for known transcripts and exons of protein coding genes\n",
    "    ref_gtf = ref_gtf[[\"chr\", \"type\", \"start\", \"end\", \"strand\", \"transcript_id\", \"gene_id\", \"start_codon\", \"stop_codon\"]]\n",
    "    merged_final = pd.merge(merged_final, ref_gtf, on=[\"chr\", \"type\", \"start\", \"end\", \"strand\", \"transcript_id\", \"gene_id\"], how=\"left\")\n",
    "    \n",
    "    \n",
    "    merged_final = merged_final[[\"chr\", \"type\", \"start\", \"end\", \"strand\", \"transcript_id\", \"gene_id\", \"gene_name\", \n",
    "                    \"exon_number\", \"transcript_name\", \"start_codon\", \"stop_codon\", \"is_novel_gene\", \"is_novel_transcript\"]]\n",
    "\n",
    "\n",
    "    \n",
    "    return merged_final "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: make_gene_and_transcript_converter\n",
    "\n",
    "input: The CHM13 CAT/Liftoff gff annotation version 2.0\n",
    "\n",
    "output: A dataframe with [\"gene_id\", \"transcript_id\", \"gene_name\", \"transcript_name\"] formatted in the same way as the \n",
    "bambu reference, so that we can properly assign gene and transcript names.\n",
    "\n",
    "purpose: Creating a list that allows us to assign transcript and gene names to the bambu annotation based on the transcript\n",
    "id and gene ID\n",
    "'''\n",
    "\n",
    "\n",
    "def make_gene_and_transcript_converter(gff):\n",
    "    \n",
    "    ## Change name of duplicate Ensembl IDs to CHM IDs\n",
    "    gff.loc[gff[\"transcript_id\"] == \"N/A\", \"transcript_id\"] = gff[\"CHM_transcript_id\"]\n",
    "    gff_transcripts = gff.loc[gff[\"type\"] == \"transcript\"].copy()\n",
    "    gff_transcripts = gff_transcripts[[\"transcript_id\", \"CHM_transcript_id\"]].drop_duplicates()\n",
    "    gff_transcripts = gff_transcripts[gff_transcripts['transcript_id'].duplicated() == True]\n",
    "    dup_trans = gff_transcripts[\"transcript_id\"].dropna().values.tolist()\n",
    "    gff.loc[gff[\"transcript_id\"].isin(dup_trans), \"transcript_id\"] = gff[\"transcript_id\"] + \"(\" + gff[\"CHM_transcript_id\"] + \")\"\n",
    "\n",
    "    ## Change name of duplicate gene ids to CHM ids\n",
    "    gff.loc[gff[\"gene_id\"] == \"None\", \"gene_id\"] = gff[\"CHM_gene_id\"]\n",
    "    gff_genes = gff.loc[gff[\"type\"] == \"transcript\"].copy()\n",
    "    gff_genes = gff_genes[[\"gene_id\", \"CHM_gene_id\"]].drop_duplicates()\n",
    "    gff_genes = gff_genes[gff_genes['gene_id'].duplicated() == True]\n",
    "    dup_genes = gff_genes[\"gene_id\"].dropna().values.tolist()\n",
    "    gff.loc[gff[\"gene_id\"].isin(dup_genes), \"gene_id\"] = gff[\"gene_id\"] + \"(\" + gff[\"CHM_gene_id\"] + \")\"\n",
    "\n",
    "    \n",
    "    ## Fix gene names for MSTRG Genes\n",
    "    gff_names = gff.loc[gff[\"type\"] == \"transcript\"].copy()\n",
    "    gff_names = gff_names[[\"gene_id\", \"gene_name\"]].copy()\n",
    "    gff_names.loc[gff_names[\"gene_name\"].str.contains(\"MSTRG.\"), \"gene_name\"] = np.NaN\n",
    "    gff_names.dropna(inplace=True)\n",
    "    gff_names.drop_duplicates(inplace=True, subset=[\"gene_id\"])\n",
    "    gff.drop(columns=\"gene_name\", inplace=True)\n",
    "    gff = pd.merge(gff, gff_names, on=\"gene_id\", how=\"left\")\n",
    "    gff = gff[[\"gene_id\", \"transcript_id\", \"gene_name\", \"transcript_name\",\n",
    "              \"start\", \"end\", \"type\", \"start_codon\", \"stop_codon\", \"chr\", \"strand\"]].copy()\n",
    "\n",
    "\n",
    "    return gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_counts = pd.read_csv(\"../../data/novel_genes/chm13_bambu/counts_transcript.txt\", delimiter=\"\\t\", low_memory=False, \n",
    "                           header=0,)\n",
    "grch_counts = pd.read_csv(\"../../data/novel_genes/GRCh38_bambu/counts_transcript.txt\", delimiter=\"\\t\", low_memory=False, \n",
    "                           header=0,)\n",
    "\n",
    "list_counts = [chm13_counts, grch_counts]\n",
    "\n",
    "for df in list_counts:\n",
    "    \n",
    "    new_col_names = []\n",
    "    \n",
    "    for col in df.columns:\n",
    "        new_col_names.append(col.split(\"_nanopore\")[0])\n",
    "        \n",
    "    df.columns = new_col_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TXNAME', 'GENEID', 'sample_PAM54902', 'sample_PAM54401',\n",
       "       'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_counts.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHM13 counts per sample:\n",
      "\n",
      "sample_PAM54902    3.194615e+07\n",
      "sample_PAM54401    3.739172e+07\n",
      "sample_PAM54335    2.695662e+07\n",
      "sample_PAM54788    4.773581e+07\n",
      "sample_PAK20814    3.806709e+07\n",
      "dtype: float64\n",
      "\n",
      " GRCH38 counts per sample:\n",
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"None of [Index(['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335',\\n       'sample_PAM54788', 'sample_PAK20814'],\\n      dtype='object')] are in the [columns]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-13-727c745062c3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"\\n GRCH38 counts per sample:\\n\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgrch_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'sample_PAM54902'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'sample_PAM54401'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'sample_PAM54335'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'sample_PAM54788'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'sample_PAK20814'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msum\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"\\nDifference in Counts per Sample:\\n\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   2906\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2907\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2908\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2909\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2910\u001b[0m         \u001b[1;31m# take() does not accept boolean indexers\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[1;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[0;32m   1252\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0max\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1253\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1254\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_validate_read_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mraise_missing\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1255\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mkeyarr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1256\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[1;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[0;32m   1296\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mmissing\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1297\u001b[0m                 \u001b[0maxis_name\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_axis_name\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1298\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"None of [{key}] are in the [{axis_name}]\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1299\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1300\u001b[0m             \u001b[1;31m# We (temporarily) allow for some missing keys with .loc, except in\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: \"None of [Index(['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335',\\n       'sample_PAM54788', 'sample_PAK20814'],\\n      dtype='object')] are in the [columns]\""
     ]
    }
   ],
   "source": [
    "print(\"CHM13 counts per sample:\\n\")\n",
    "print(chm13_counts[['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']].sum())\n",
    "\n",
    "print(\"\\n GRCH38 counts per sample:\\n\")\n",
    "print(grch_counts[['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']].sum())\n",
    "\n",
    "print(\"\\nDifference in Counts per Sample:\\n\")\n",
    "difference = chm13_counts[['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']].sum() - grch_counts[['sample_421', 'sample_20484', 'sample_27527', 'sample_356', 'sample_20814']].sum()\n",
    "print(difference)\n",
    "\n",
    "print(\"\\nPercent difference in Counte per Sample:\\n\")\n",
    "percent_difference = difference/(chm13_counts[['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']].sum())\n",
    "print(percent_difference * 100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Find novel transcripts in both annotations\n",
    "chm13_counts.loc[chm13_counts[\"TXNAME\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "chm13_counts.loc[~chm13_counts[\"TXNAME\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "grch_counts.loc[grch_counts[\"TXNAME\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = True\n",
    "grch_counts.loc[~grch_counts[\"TXNAME\"].str.startswith(\"tx.\"), \"is_novel_transcript\"] = False\n",
    "\n",
    "## Find novel genes in both annotations\n",
    "chm13_counts.loc[chm13_counts[\"GENEID\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "chm13_counts.loc[~chm13_counts[\"GENEID\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False\n",
    "\n",
    "grch_counts.loc[grch_counts[\"GENEID\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = True\n",
    "grch_counts.loc[~grch_counts[\"GENEID\"].str.startswith(\"gene.\"), \"is_novel_gene\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclusive_chm13 = pd.read_csv(\"../../data/novel_genes/CHM13_novel_genes_isoseq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENEID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHM13_G0002360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHM13_G0002906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHM13_G0003812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHM13_G0006285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHM13_G0007830</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENEID\n",
       "0  CHM13_G0002360\n",
       "1  CHM13_G0002906\n",
       "2  CHM13_G0003812\n",
       "3  CHM13_G0006285\n",
       "4  CHM13_G0007830"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exclusive_chm13.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_exclusive = chm13_counts.loc[chm13_counts[\"GENEID\"].isin(exclusive_chm13[\"GENEID\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(57, 9)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_exclusive.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter novel genes and transcripts, must be present in every sample with 5+ counts.\n",
    "count_cols = ['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']\n",
    "\n",
    "for col in count_cols:\n",
    "    chm13_exclusive = chm13_exclusive.loc[chm13_exclusive[col] > 5].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13, 9)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_exclusive.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create dataframes with either only novel genes or only novel transcripts that are not from novel genes\n",
    "chm13_novel_genes = chm13_counts.loc[chm13_counts[\"is_novel_gene\"]]\n",
    "grch_novel_genes = grch_counts.loc[grch_counts[\"is_novel_gene\"]]\n",
    "\n",
    "chm13_novel_transcripts = chm13_counts.loc[((chm13_counts[\"is_novel_transcript\"]) & (~chm13_counts[\"is_novel_gene\"]))]\n",
    "grch_novel_transcripts = grch_counts.loc[((grch_counts[\"is_novel_transcript\"]) & (~grch_counts[\"is_novel_gene\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We found 212 novel genes in CHM13\n",
      "We found 160 novel genes in GRCh38\n"
     ]
    }
   ],
   "source": [
    "print(\"We found\", chm13_novel_genes.shape[0], \"novel genes in CHM13\")\n",
    "\n",
    "print(\"We found\", grch_novel_genes.shape[0], \"novel genes in GRCh38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter novel genes and transcripts, must be present in every sample with 5+ counts.\n",
    "count_cols = ['sample_PAM54902', 'sample_PAM54401', 'sample_PAM54335', 'sample_PAM54788', 'sample_PAK20814']\n",
    "\n",
    "for col in count_cols:\n",
    "    chm13_novel_genes = chm13_novel_genes.loc[chm13_novel_genes[col] > 5]\n",
    "    #grch_novel_genes = grch_novel_genes.loc[grch_novel_genes[col] > 5]\n",
    "    chm13_novel_transcripts = chm13_novel_transcripts.loc[chm13_novel_transcripts[col] > 5]\n",
    "    #grch_novel_transcripts = grch_novel_transcripts.loc[grch_novel_transcripts[col] > 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We found 146 novel genes in CHM13\n",
      "We found 160 novel genes in GRCh38\n",
      "We found 428 novel transcripts of known genes in CHM13\n",
      "We found 270 novel transcripts of known genes in GRCh38\n"
     ]
    }
   ],
   "source": [
    "## Count number of novel genes and novel transcripts in each dataframe\n",
    "print(\"We found\", chm13_novel_genes.shape[0], \"novel genes in CHM13\")\n",
    "print(\"We found\", grch_novel_genes.shape[0], \"novel genes in GRCh38\")\n",
    "\n",
    "print(\"We found\", chm13_novel_transcripts.shape[0], \"novel transcripts of known genes in CHM13\")\n",
    "print(\"We found\", grch_novel_transcripts.shape[0], \"novel transcripts of known genes in GRCh38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene.989     4\n",
       "gene.746     2\n",
       "gene.379     2\n",
       "gene.892     2\n",
       "gene.222     2\n",
       "gene.292     2\n",
       "gene.72      2\n",
       "gene.358     2\n",
       "gene.676     2\n",
       "gene.718     2\n",
       "gene.42      2\n",
       "gene.178     2\n",
       "gene.570     2\n",
       "gene.553     2\n",
       "gene.197     2\n",
       "gene.536     2\n",
       "gene.635     1\n",
       "gene.1015    1\n",
       "gene.879     1\n",
       "gene.237     1\n",
       "gene.13      1\n",
       "gene.651     1\n",
       "gene.955     1\n",
       "gene.468     1\n",
       "gene.755     1\n",
       "gene.520     1\n",
       "gene.351     1\n",
       "gene.195     1\n",
       "gene.806     1\n",
       "gene.855     1\n",
       "gene.1013    1\n",
       "gene.514     1\n",
       "gene.752     1\n",
       "gene.136     1\n",
       "gene.941     1\n",
       "gene.683     1\n",
       "gene.550     1\n",
       "gene.959     1\n",
       "gene.306     1\n",
       "gene.244     1\n",
       "gene.480     1\n",
       "gene.997     1\n",
       "gene.895     1\n",
       "gene.668     1\n",
       "gene.603     1\n",
       "gene.22      1\n",
       "gene.453     1\n",
       "gene.185     1\n",
       "gene.332     1\n",
       "gene.788     1\n",
       "gene.367     1\n",
       "gene.748     1\n",
       "gene.679     1\n",
       "gene.967     1\n",
       "gene.316     1\n",
       "gene.428     1\n",
       "gene.818     1\n",
       "gene.501     1\n",
       "gene.470     1\n",
       "gene.802     1\n",
       "gene.987     1\n",
       "gene.915     1\n",
       "gene.51      1\n",
       "gene.17      1\n",
       "gene.392     1\n",
       "gene.647     1\n",
       "gene.950     1\n",
       "gene.5       1\n",
       "gene.345     1\n",
       "gene.562     1\n",
       "gene.832     1\n",
       "gene.778     1\n",
       "gene.193     1\n",
       "gene.510     1\n",
       "gene.865     1\n",
       "gene.188     1\n",
       "gene.272     1\n",
       "gene.324     1\n",
       "gene.94      1\n",
       "gene.456     1\n",
       "gene.37      1\n",
       "gene.318     1\n",
       "gene.786     1\n",
       "gene.585     1\n",
       "gene.304     1\n",
       "gene.829     1\n",
       "gene.870     1\n",
       "gene.66      1\n",
       "gene.771     1\n",
       "gene.279     1\n",
       "gene.277     1\n",
       "gene.434     1\n",
       "gene.81      1\n",
       "gene.356     1\n",
       "gene.313     1\n",
       "gene.792     1\n",
       "gene.994     1\n",
       "gene.842     1\n",
       "gene.767     1\n",
       "gene.125     1\n",
       "gene.618     1\n",
       "gene.911     1\n",
       "gene.105     1\n",
       "gene.590     1\n",
       "gene.1029    1\n",
       "gene.923     1\n",
       "gene.969     1\n",
       "gene.31      1\n",
       "gene.443     1\n",
       "gene.999     1\n",
       "gene.981     1\n",
       "gene.79      1\n",
       "gene.518     1\n",
       "gene.384     1\n",
       "gene.64      1\n",
       "gene.710     1\n",
       "gene.102     1\n",
       "gene.363     1\n",
       "gene.643     1\n",
       "gene.900     1\n",
       "gene.578     1\n",
       "gene.759     1\n",
       "gene.611     1\n",
       "gene.252     1\n",
       "gene.119     1\n",
       "gene.948     1\n",
       "gene.153     1\n",
       "gene.848     1\n",
       "Name: GENEID, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_novel_genes[\"GENEID\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TXNAME</th>\n",
       "      <th>GENEID</th>\n",
       "      <th>sample_PAM54902</th>\n",
       "      <th>sample_PAM54401</th>\n",
       "      <th>sample_PAM54335</th>\n",
       "      <th>sample_PAM54788</th>\n",
       "      <th>sample_PAK20814</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>tx.652</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>653.065333</td>\n",
       "      <td>2424.468517</td>\n",
       "      <td>1571.293188</td>\n",
       "      <td>1356.549767</td>\n",
       "      <td>3210.254654</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>tx.653</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>444.249361</td>\n",
       "      <td>1717.037009</td>\n",
       "      <td>830.732977</td>\n",
       "      <td>949.355434</td>\n",
       "      <td>2446.676183</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>tx.654</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>250.205703</td>\n",
       "      <td>791.827812</td>\n",
       "      <td>391.937493</td>\n",
       "      <td>537.436985</td>\n",
       "      <td>1210.517097</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>tx.655</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>208.479603</td>\n",
       "      <td>1758.666662</td>\n",
       "      <td>726.036342</td>\n",
       "      <td>967.657814</td>\n",
       "      <td>2644.552066</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TXNAME    GENEID  sample_PAM54902  sample_PAM54401  sample_PAM54335  \\\n",
       "651  tx.652  gene.989       653.065333      2424.468517      1571.293188   \n",
       "652  tx.653  gene.989       444.249361      1717.037009       830.732977   \n",
       "653  tx.654  gene.989       250.205703       791.827812       391.937493   \n",
       "654  tx.655  gene.989       208.479603      1758.666662       726.036342   \n",
       "\n",
       "     sample_PAM54788  sample_PAK20814 is_novel_transcript is_novel_gene  \n",
       "651      1356.549767      3210.254654                True          True  \n",
       "652       949.355434      2446.676183                True          True  \n",
       "653       537.436985      1210.517097                True          True  \n",
       "654       967.657814      2644.552066                True          True  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_novel_genes.loc[chm13_novel_genes[\"GENEID\"] == \"gene.989\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open extended annotations and parse through them\n",
    "chm13_ref = pd.read_csv(\"../../data/novel_genes//chm13_bambu/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "grch_ref = pd.read_csv(\"../../data/novel_genes/GRCh38_bambu/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                       low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "chm13_ref = parse_df_columns(chm13_ref, is_ref=False)\n",
    "grch_ref = parse_df_columns(grch_ref, is_ref=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fix up extended annotations to merge with counts matrices\n",
    "chm13_ref.columns = ['chr', 'source', 'type', 'start', 'end', 'strand', 'GENEID', \n",
    "                     'TXNAME', 'exon_number', 'is_novel_transcript', 'is_novel_gene']\n",
    "chm13_ref = chm13_ref.loc[chm13_ref[\"type\"] == \"transcript\"].copy()\n",
    "chm13_ref = chm13_ref[[\"chr\", \"start\", \"end\", \"GENEID\", \"TXNAME\"]]\n",
    "\n",
    "grch_ref.columns = ['chr', 'source', 'type', 'start', 'end', 'strand', 'GENEID', \n",
    "                     'TXNAME', 'exon_number', 'is_novel_transcript', 'is_novel_gene']\n",
    "grch_ref = grch_ref.loc[grch_ref[\"type\"] == \"transcript\"].copy()\n",
    "grch_ref = grch_ref[[\"chr\", \"start\", \"end\", \"GENEID\", \"TXNAME\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create gene location extractor for novel genes in both dataframes\n",
    "chm13_extractor = pd.merge(chm13_novel_genes[[\"GENEID\", \"TXNAME\"]], chm13_ref, how=\"left\", on=[\"GENEID\", \"TXNAME\"])\n",
    "grch_extractor = pd.merge(grch_novel_genes[[\"GENEID\", \"TXNAME\"]], grch_ref, how=\"left\", on=[\"GENEID\", \"TXNAME\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create extraction column and drop others\n",
    "chm13_extractor[[\"start\", \"end\"]] = chm13_extractor[[\"start\", \"end\"]].astype(\"int\")\n",
    "grch_extractor[[\"start\", \"end\"]] = grch_extractor[[\"start\", \"end\"]].astype(\"int\")\n",
    "\n",
    "\n",
    "chm13_extractor[[\"chr\", \"start\", \"end\", \"GENEID\", \"TXNAME\"]] = chm13_extractor[[\"chr\", \"start\", \"end\", \"GENEID\", \"TXNAME\"]].astype(\"str\")\n",
    "grch_extractor[[\"chr\", \"start\", \"end\", \"GENEID\", \"TXNAME\"]] = grch_extractor[[\"chr\", \"start\", \"end\", \"GENEID\", \"TXNAME\"]].astype(\"str\")\n",
    "\n",
    "\n",
    "chm13_extractor[\"extraction\"] = chm13_extractor[\"chr\"] + \":\" + chm13_extractor[\"start\"] + \"-\" + chm13_extractor[\"end\"]\n",
    "grch_extractor[\"extraction\"] = grch_extractor[\"chr\"] + \":\" + grch_extractor[\"start\"] + \"-\" + grch_extractor[\"end\"]\n",
    "\n",
    "chm13_extractor = chm13_extractor[[\"GENEID\", \"TXNAME\", \"extraction\"]].copy()\n",
    "grch_extractor = grch_extractor[[\"GENEID\", \"TXNAME\", \"extraction\"]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save novel gene sequence extractors for samtools\n",
    "chm13_extractor.to_csv(\"chm13_novel_gene_extractor.txt\", sep=\"\\t\", index=False, header=False)\n",
    "grch_extractor.to_csv(\"grch_novel_gene_extractor.txt\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Genes with Novel Transcripts')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATAAAAD8CAYAAADwpviIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAwmUlEQVR4nO2dd3gc1dm372d3VW3JliX3bmxsuk0NHQKEhN6SACEJUVg+AgEUSAIkeYnpnSxvCgSFfWkpdJJAICSAaTbgijHGNi5ykYtsy7J6We35/jgjvCy7a8mW5sxI576uuaSdOTvnN+23p81zRCmFxWKx+JGAaQEWi8Wyq1gDs1gsvsUamMVi8S3WwCwWi2+xBmaxWHyLNTCLxeJbPGtgInK4iPxNRNaJSKuI1IrIbBG5RUSGm9bnJiLyqIhUJHweJyLTRWRCirQVIvLkLuZzsYgoEakRkaKkbSFn2/Rd2Xd3ISLHOTqOS7N9nLN9Z8sMV4X3EM59sEtjoUSkTETO6W5NbuJJAxORa4H3gMHAr4ATgfOBfwOXAlFz6oxwC3B2wudxwK+BLxlYNzEAuK6H9t3TbAAOT1oAHk1ad7kJcT3An9hxjF2lDPC1gYVMC0hGRI4H7gEeUEr9JGnzv0TkDuCb7iszh1JqhctZvgZcKSIRpdRGl/PeLZRSLcD7ietEBKBSKfV+yi/pNFlATPlkZLeI5CilWpRS64B1pvWYwoslsOuALaQpASilGpRSjyauE5F8EblLRFY51c1VIvJLEQkkpOmoepwhIr8TkS0isllEnhSRgUn7C4nIDSKyRERaRGS9iNwnIrlJaW4RkRUi0uzs710ROSrdgYnIT0WkQUSyE9Y95+g6MWFdWERiIlLofP68CulUnd50kv4noUp0XFJe54vIp05+czLpSsGtzt9f7iyhiBwqIv8VkXonr9dF5NCE7T93rklxiu8uFpEXEz7v9Dp2BwnVzMtF5G4RWQ+0AANFZLCI/FFElolIo4isFZG/iMjIpH1Md/YxSURedo5/tYjcmHTf9ReR34rIGude2uScrykJaUIicp1zPpqd+/LVjjQJ9+45IlIuIpuBTYk6krQpEbnNOXfrRKRJRN4WkakJaSqAscB3Eu6hR51te4rICyJS5ehZIyLPiIjnCjwopTyzoEuETcCfu/idd4Ct6CLxCegHrxm4LyHdcYACVgG/Bb4GXOnk91jSPv8GNAA3oquvVwI1wHMJaX4J1ANXA8cCpwM3AWdk0Hqgo+EY57OgzboRuD0h3V+BDxI+PwpUOP8Xoqs/ytH1FWcpdLZXAKuB2cB5wGnAfEf/wJ2cy4ud/U4E7kI/1GMTzrMCpiek3985f3OdvM518m0CDnDSjADagcuT8jrI2d+5u3gdj+vCPaKAWxM+j3PWVQIvOufoTCAPmAw84BzLMeimi9nOec1N2Md0Zx+LgGud++QBZ90PEtKVo83mh87+zgbuBb6SkOZZIOas/zpwFnA/cHzSMVeiq4xfB85K1JHieNeim2HOAr4NLHXO7SAnzTR0dftVdtxDezjblgEfOufgWOBC4Ekg27RHfOnamhaQdOKHOif/jhTbQolLwvrvkmAKCet/CbQCQ5JugmSz+p3zkIjz+Wgn3feS0n3HWT/V+fwS8HwXjy8AVAO/dj5PBeLAb4BZCek2AHcmfH4Ux8CSjuXEFHlUANuAooR1BzvpL9yJvovZYWCD0KYXTTj/yQb2LEnGiDbY6sRzA/wn8ficdREnXc4uXsfjunDe0xnYvI7rnuG7QWC0k/7shPXTSTIrZ/3HwGsJnxcB92fY/1ed/VyVIU3HMb+QYtt0UhvYFqBf0jG3Abck3StPJn23xPl+2h9iLy1eq0JKypUiw9An//MloTj7dXSJY6ZTFA85214DstC/LIm8nPT5YyAHbZ4d+2sFnkuxP9C/oqB/lU9xiupHJVYL06GUigNvo29anL8LgaeBg0WkQET2BoYBb+xsfxmYpZTalvD5Y+fvmM7uQClVDdwHfE9EJqdJdgzwklKqJuF7tcA/0L/cHTwBfEVEJoGuMqFLNk8r3WYFXb+O3cGLynlqExGRH4nIRyJSjy4ZrXE2pToPyffTIr54nmcDF4vIL0TkYBEJJqX/Gtowyjuh94VOpOngX0qpho4PSqkKdNvgzhr8twIrgTudpoxJXcjTdbxmYFvQpaHkB20LcIizJF/oIei6fFvS8qGzPbntpTrpc8cD1NG+NQTIRlcPE/dXlbS/29E9gWfgVH1E5P9EpGQnx/gG+mHOA45Ht2fNRh/30c66NnTxf1f5wjEmmERuirSZ+I2zr5vTbB+ELi0msxFIHIbxHLpKfpHz+WvoH4wnEtJ09Tp2B1/SLiJXAn8A/ovuoTuUHeaZ6vylup8S010J/BEoRV/nKhH5jYjkO9uLgWqlVNOu6M3ApjTrRqZY/zmOoZ8EzAHuAJaJyEoR+VEX8nYNTzXKKaViIvI2cJKIZCulWjvWo08oInJa0te2otu1vpVmtxVdlLGVHWaSivWOpjZ0O9FdTgnxNHS7RT66zSEdb6IN8hhnedg57nfQJbLxwIeJv56mUErVi+71vQ/dM5xMNbq0mMwwEh5spVSDiLyArob/Gm1kK5VSiSbd3dexM6TqcTwfeF0pdW3HChEZv8sZKFUP3ADcICJj0W2Fd6JL+R0dVoNEJK8TJpZKbzqGpllX2QnNK9ElbwEOAH4M/EFEKpRSr3RBQ4/jtRIYwN3oevhdnUz/KrqNol4pNSfFsqWL+b+K/gUdkGZ/65O/oJTaqJT6E/pXe9+d7H8RsBn4GdAPXaUEXTI7AV312ln1saNEldepI9o9/oC+6W9Nse0t4FQRKehY4fx/urMtkSeAPUTkZHSD+RNJ27v7Ou4q+eiSXyI/6I4dK6VWK6XuQ1fpO+6T19BNJ5d0Rx4JnCIi/To+iMg4dElyVkKaFjLcQ0qzALjGWbWze9t1PFUCA1BKvS4i16Pr4PsDj6N/mXOBPdG/kA3s+DX6M/oGe11E7gM+Qpdw9kBX785SSjV2If8ZIvJX4FkRuR9dhYmjG0FPAa5TSi0Tkb87ec1DN5pPQ7fj/HEn++8YBf5NYLZSaruz6U12lHLe3InMZei2mVIRqUbfiEuVUnWdPc7OopRqEZGbgYdTbL4FXfJ8XUTuQl+T69AmkFzt/C+69PqIsz35bYFuvY67wavAdSLyC/S1/yq61LRLiMgsdJvgx+hmiWPRpZrHAJRSb4rIc8D9IjIa/eOVhS6dv6yUmrGLWTcBr4nIPeg23puAWnSzQAeLgaOdWs1GdGmwEN2b+hSwHN2JcTH6ftuddtmewXQvQroFOBLduF2JLm7XotsQbgKGJ6XNRffGLEE/zNVO2uk4PZak6bljR8/buIR1AfTwiI/Q1cntzv93o0tmoLvO30dXfZrQ3dTTgaxOHNuPnDzvTMqz2skvNyn9oyT0Qjrr/h+6sTVGQq8cKXqW1I6eqek70dVxLiYmrQ+hTfNL+wAOQ5tTPfqH5XXg0DT7v8fZx8w027tyHY/rwr2UrhfykhRp84AH0aXkOnRv8/jkY2dHL2Qo6ftfuFbomsR85x5qQBvZVUnfCaF7W5eh7/XNwL+AyZnu3UQdKY73NuAX6EGuzeh22qlJ6aY46xud7zyKbot8zNHS6FyDt4CTTXtCqqVj6IDFYuklOANbb1NK/cq0lp7Gi21gFovF0imsgVksFt9iq5AWi8W32BKYxWLxLdbALBaLb7EGZrFYfIs1MIvF4lusgVksFt9iDcxisfgWa2AWi8W3WAOzWCy+xRqYxWLxLdbALBaLb7EGZrFYfIs1MIvF4lusgVksFt9iDcxisfgWa2AWi8W3WAOzWCy+xRqYxWLxLdbALBaLb7EGZrFYfIs1MIvF4lusgVksFt8SMi3AkplwhAKgJGEZhJ7+vT/Qz1n6A9noGZg7liZnSV7XjJ7lfDVQWV5G3MXD6XZUqQSAEcBY9HnJc5bchCXVulZ2zCbe4Pxfi56Jegt6xvUtElW1Lh6OpYvYadU8QjiCoKe8nwJMdv6OA3J6MNs2YA1QAaxK+LuhvAxP3RiqVAQYjj4n451lHDAGyOrBrFvR52VJwlIhUfvgeAFrYIYIRxiONqmOZRK6lOAFWtDGtgpYCswqL2ODmwJUqQwDDmeHkY+lZ828KzQDy0gwNYkqV8+PRWMNzCXCEQqBY4Cj0A/lALOKukwFMBOYBXzS3SU0p4S1N3CEs4zrzv27QC3wKfAe8JaterqDNbAeJByhH9qwvgocBATNKuo2tgMfoA1tdnkZjbuyE1Uq+cAhaMM6FBjYXQIN0w7MA94A3pGoajCsp9diDaybCUfIBY4Ejkc/lD3ZPuMFYsBHaDN7o7yMmkyJVakMAE5AVw+n0vs7ktqA2WgzmylR1WRYT6/CGlg3EI6QBXwFXdI6HO+01bhNG/AW8GJ5GZ8kblClsjdwJtrYe7upp6MFeB9tZrMkqtoM6/E91sB2g3CEHOAM4NtAsWE5XuOz/oGW5+77ZCjStv1cdCeFZQdbgaeBf0hUNZsW41esge0C4Qj5wFnAN+k97TbdRj9pbTyjcMn6Q/PWDc9SzQTWz9gYXPbYMGmr7WdamwfZDjwDvCBRtUttiX0Za2BdIByhP3CusxQYluM58qW18dsDPl4/LXfDmICQ/YWNKt4qm2auCS3+/Qhpq883JNHL1AHPA89JVNWZFuMXrIF1gnCEAejS1tmAffiSyKK99ezCxWuOzF8zMigq81i2eKw5sO61yuDSR0ZLvDU7Y9q+SSPwIvCMRFWNWSnexxpYBhzjuhDdzpVrWI7nCBJvP7n/ZxUn9V8xJEviXSuRtrfWBSqe3xxc8bexotp7y/CS7qQF+AfwF2tk6bEGlgLntZ5vAJdhq4op2T9n47rvDlyQnxeIDdqtHcUaq0MfRxoCVbNGd5O03kYd8DDwsn196ctYA0siHGEscA2wv2ktXqQg0NIQLppTNSF72/ju3K/ULKkIzb91sLRutw39qVkE3CdRVWFaiJewBubgjOX6HnA+vX9w5S5xYr/lK08rWDosJKpn2gHjscbg8j9vDKx6dryA9Ege/iaGHnrxmERVq2kxXsAaGBCOMAn4Bf57/84VhgTrt10x6IPG4lDTSFcybKqqzJp7Y540VO5e9bT3shq4XaJqmWkhpunTBhaOEAQuchZb6krBMfmrVp1buHh4UJS7nRjx9ubg0uiG4Jp/dGtVtRfRDjwJPCFR1W5ajCn6rIGFI4xBl7omm9biRbIl1vqjog8rJ+VUGzUQ2fbJqtDcX4+U9hY75CI1nwG3SVStNi3EBH3SwMIRDgNuxI7pSsmYrJqqHw96P5gfiHnj9ai2hurQ3Btjge3LhpiW4lEagVskqt43LcRt+pyBhSOcC1yOnQ8gJSf1W77y9IIlowPisReuVTwWXP7n1cGVT+9hWopHUcCDElXPmBbiJn3GwJz2rivREREsX0KpHw6cu3Ja3kZPG4Rsmrk8tOCOPWwvZVr+CTzQV9rF+oSBOYEFpwMHG5biSbIl1npt8XsbRmbVjTWtpTNIXcXq0Ac/HW7bxdIyF5guUVVvWkhP0+sNzIk9fwc6proliYGBptrrSt5pKgi2DjWtpUu01GzKmvWTPGnZUmhaikdZA1zf22P192oDC0fYF7gV/8Wfd4XRoZrNPymZmZMtcX+aQHtLXejD65sCtctt435qtgM3SlQtNC2kp+i1BhaOcBLwM/pu9M+MjMvatqmseGZBj42qd4t4rDH04fV1ge1L/VWCdI8YcI9E1WumhfQEvdLAwhFOQZuXJQUTsqo3XlU8q9D35tVBPNYUmv3L7YGaxcNMS/Ew90pUvWxaRHfT64YShCMcDVxrWodXmZi9dcPVxbMG9hrzAgiE8mKH3DYwXrRfr27v2U2uUaVyjGkR3U2vKoGFI0wD7sJWG1OyZ/aWDVcM+qDI9deC3CLe3hya++vqQPVHI0xL8Sht6Ib9eaaFdBe9xsDCESYD92NH16dkTFZN1bXF7xX2WvPqIN7eEvrwuhrbJpaWJuAaiaolpoV0B72iChmOMBpd8rLmlYJBwcbtZcUz83q9eQEEgjmxQ27LV3lDa0xL8Sh5wF2qVMaYFtId+N7AwhGGAPdih0qkJE/amq4reactu6shn/1MMKeg7fDfxFRWfzvLT2oKgXtVqfh++ImvDcyJWX8P4PsL0RMEibdfV/J2db9AW4lpLa6TVVDSdvgD25SEYqaleJTBaBPz9Q+/bw0sHCEPuBPoFUXh7kepsuKZa0rcCkLoRfKGjIwddvc6pV90tnyZ0ejqpG+bXnxrYMANwBTTIrzKeYWfrBifXdPngwGqAZPGte912QrTOjzMZHRcPF/iSwMLRzgNONq0Dq+yV3ZV5bH5FRNM6/AK8dGnTIiXHFxpWoeHOVKVyhmmRewKvjMwJ5Lqj03r8CqFgeb68KA5hSL+u7Y9hkggNvWGQpUzyM54nZ7LVan4LuCBr27ycIQQ8Csgx7QWb6LUNcUza/pUj2NnCWYXtB16V61C4qaleJQc4FeqVHw1N4SvDAy4BJhkWoRXuWjARytLQo2jTOvwLPnDRrbv95NVpmV4mInApaZFdAXfGFg4wkHAt0zr8Cp7ZVdVHpa3zrZ77YT48OMmxEsOsu1h6TlPlYpvAn/6wsDCEQrRvY42jHAKsmhv/WHRvHwRe352iojEDvh5vgrm2IlhUyPADapUBpoW0hl8YWDAzwFvzJDjQUqL5q7NDcSKTOvwDaH8otgB1601LcPDDMIn4ag8b2DhCKcDR5rW4VUmZW/ZsG9Ola06dhFVcvCE+KD9bfid9ByhSsXzE+B42sDCEQYDV5jW4VWCxNsvKZobslXHXUBEYlOvz1IS7BOz9+wiP/L6+5KeNjB0j4gdMpGGcwoXr+oXaBtsWodvySooad/rUtsrmZ4cPN4r6VkDC0fYCzjBtA6vUhBoaTgqf7UdMrGbxEeePFplD+z104/tBieoUtnHtIh0eNbA0KPtbdUoDRcNWLCxT8T36mkCwZzYftdUmZbhca5QpeLJZ9GTBhaOcAKwt2kdXmV4qHbr3jmbx5nW0VtQxVPHxfuP22Jah4fxbG3IcwbmvC70Q9M6vMwPBs6vFyFoWkevQSQQO+BnNvhhZi7x4mtGnjMw4BRguGkRXmXvnKrKEVl1vnvp1vP0HzMmPviQdaZleJihwKmmRSTjKQMLR8gGvmtah5e5YMBCT7ZF9AZie19hS7WZ+a4qlWzTIhLxlIEBZwB9L/xxJ5mcvXlDUbDZThnWU+QWD48XT7PvSaanGPDU4FbPGFg4Qg5woWkdXubcwsX2/b0eJjblEjuwNTMXqlLxTO+3ZwwMOAmw7/OlYUSodsvwUJ2N/9/T9Bs9Ol4w3vZIpmcg+ln1BF4yMM81EHqJ8wo/qbWvDLmAiLRPudRGbs2MZ55VTxhYOMIE7AQdaRkYaKqdlL3V9jy6hCrae4zKHVxrWoeHmaxKZQ/TIsAjBoaHHN2LnFqwrMqO+3IRCQTbJ15oR+dnxhPPrHEDC0fIwkN1au+h1LTc9bZn1mXiQ48abOeTzMiJXhhSYdzAgGMAOwlFGvbNqarMDbQPNK2jzxHKHaAGH2qHVKSnAP3sGsULBuaJoqhXOan/cjt0whDt489rM63B4xh/do0aWDjCcGCqSQ1eJlfamsdnbbMhcwyhBk4eqYJ5zaZ1eJgDVKmMNCnAdAnsVGzInLQcnb96fUAw3s7QZ5FAdnzs6etNy/Awgn532RjGDCwcIQB83VT+fuDw/LVZpjX0ddpHnmh/QDJzsioVYz3kJktgB2BnGkpLjsRaBgcb7HuPpskbOtxWIzNSjMFmIJMGdqDBvD3PtNwNG+3YLw8ggWB82NGbTMvwONNMZWzSwIwdtB84NG+dfanYI8RHHB83rcHj9C0DC0fIBSabyNsfKDU+e5unp7PqS6iBew6xg1ozMlmVSp6JjE2VwPYDPBee1ivskVW9KUvi/U3rsDgEsvupov02mpbhYYLA/iYyNmVgtvqYgYPz1ttpvjxGfPixDaY1eJypJjK1BuZBJmZvtV33HiNetLedYDkzRp5p1w0sHCEf2NPtfP3E4FCDfXnba+QPL7btYBmZpEqln9uZmiiB7W8oX18wNFRXHRKVb1qHJYlAKF/1H7vNtAwPE0CP7XQ9U7eZaiBP37B3zuYa0xosqVElB243rcHjTHU7QxMGZtu/MjAle7OtpniUePE0Ox4sM1PdztBVAwtHEGCCm3n6jdFZ2+3wCY+iCsbba5OZ8apUXPUUt0tgJdjxXxnpF2gbaFqDJQ3ZBUW2IT8jIVye19VtAxvmcn6+ojDQXBcUZbvrvYoEs8kpsTMWZcbVZ9wamIcYGaq1M+F4HFUwzhpYZqyB9VVGZdW2mNZgyYwqnGBD62TGGlhfZVRWre3l8jjxgvG2DSwz1sD6KkND9TYCq9fpN8peo8xYA+ur9Au02ncgPY7KLrSdLJnpnQbmjAGzMa4ykCsx+3B4nWCuvUaZGezmWDA3S2B2DNhOyJL2XNMaLDshkG2vUWZcHQvmpoENdTEvH6JUEGUfDq8jwVyF2M6WzLhWjXTTwApdzMt35Etbs4iN0uF5RISsAjuUIjOuPetuVulcrT5u37ws583HL9yruWFLtkiAiQdftP6gb9y8bsnMhwd/9Pqd4+uqK/JPufz1uSMmHV8HEGtrljcfP39y9YZFBSLCYWfc+9nY/c6scUtv/0BrM2Akrvju8K37bpvy9uKPi4v6FbR++r/lswFOu/3GvSuqNuUD1DU3hQpy82KLIn+c88bHCwqueuTByQAKJT8/85urvn/8SVtM6t8lsgubaas1EvJo3lbyLpjBPh2fKxvJu2YfVm1tIeu1SkoCghqUQ9ufj+XTCQW0mtCIi8+6m6biavdzIJClDjvzvuXDJx5b39JYHXz+nqkHj9nntOriUQc2nPCDZz9+56nwFyYVWfjG3SMAvv2r5bPrt63NeuWhkw4Yvc9pcwIBd2Y2yw+0mbrZdovvH3fShiu/cea6Sx96YK+OdS/94ubFHf+HH4zsUZifHwM4ZOKeDfPu/f3c7FBIVVRtyv7KDVcfcsHRx2/NDoV8NbZKZRW2mppO/sBimpaeyxyAWByG/ZUjLpzA5sF5xH5/OKsAps9n5A1zGPfU8SwzJNM1X3GzyuLqHIcFxeNbh088th4gJ39Qe0HxhIa66tU5g8cc3Fgy6sCm5PQ1mz7NHzHpq9sA+heNbsvKKWzbsPzNArf0Bon76iHu4NSDDt0+eMCAWKptcaV4df6cId8/9sQqgIK8/HiHWTW2tgQEUzawmwS8YbjPr6ZoVD+apwykpTiHz6fha4gRFLOn1pbAupNtmz7N3b7p04IRk76a9l3DQSMOqF/zyT9Lphx5WVXt5s9yajZ9WlBXXZELuPLuW1CUJx6K7uSVebMHDOpf0Lr/uAmf/2C8vnB+wRXlv99rY822nAdKL/vUb6UvACTgCc1/W8nQc8bx+aS7P5rJ+BdWM6x/FrEZ32CBQWm2BNZdtDTVBP8bPXvfg0+9/bPcfsVpJ4vd7/hrNuYPGNHy3F37HvTes1dMLB55QG3AxV9aofcZ2F/fnTH0zEMPr0pcd8L+0+qW/PZPH74+/c65kZdeGNvQ0uy/jgt3Q16lpDmGvLmB4osn8vn5ffAIVm28gFlnj2XTXR8z0qA81551N6+E6zNNt8da5N9//MY+4/c/d+Pkr/wwY2NxMJitjr3wseXf+sWyOaf9eMai1ubaUNGwvRvd0qowXOjvZlpjMXl94fzB3zvuxKpU2w/aY1JjXk5O++zly1yfCGK3UeZHUfxtFcVTBlI/pj9tydt+OIlNr6xjsAldDq49624aWMp2kp5CqTivP/rNKYWDJzUefOpt63aWvrW5NtDatD0AsHLBM0WBQEgNHnOoawbWrnqXgb344cyisYOHNk4cNuLzCBuL163JbY3FBGBp5bqc1Zs35e85fKT/hiSouPFr9dQqhpyXUH38qHpHD/ZTqygZ3x/X7t0UuGZgbraBuWpgaz99ZcCaT/45tLBkj4anb9/zYIADT56+sj3WHJj90vWTWhtrsv/7f+fuP3DI5LozymYtbNhemf3qQ18/QERUXsHQluO/+5fFO8ujO2knYPyh2BXOuOPXe89Z8dnA7Y0NWaMuvejwa04/p+Ka08/Z8PR7bw85+7AjNiWmfePjBQPOuevmsaFgMB4Q4Y7vlC4bMaj4SyUIzxOPGb1Wta0EZlUx6IljWNqx7toPmVBRT74AI/Jpjh61Y5sBXLumolxqeglHOBaY7kpmPmRoqG7b/wx+q8i0DsvOCb334+pA/epBpnV4mJskqma4kZGbVUj//dK6SF17jn1J2CdIa429VplxrbblpoHVu5iX72hUWXlK2QkjPI9SitZa370x4TINbmXkpoFt2nmSvoxIO/KlAbYWj6HaWwRlfhyFt9noVkZuXojNGBhK4SfaVNB/PXJ9jXir/ZHJTBxIOXSmJ3DNwMrLiKNNzJKGFhX05fuQfYr2ZnuNMrNFoqpXjgMDF4uWfqQ+nm1nJfI40lZnr1FmXH3GrYF5iKpYf9tT63UaKm0JLDPWwPoqlbFCXw5m7UsE6irsNcqMNbC+yrq2QhtS2uNI7Qo7Biwz1sD6KpVtha7FH7PsGlJXYa9RZqyB9VVq4rkFcWUsDLBlZ6h4K81Vdm6HzPRqA7NjwTIi0hDPrjGtwpKGtvrtgl/DyLqCq2PAwGUDc8aCrXYzT79RGSt0JQKspetIfYW9NplZ7eYYMHC/BAYYDXXreZa0lNhfeI8S2LLAtASvs8DtDE0Y2HwDefqGxS1DbBuLR5Etc+21yYzrz7YJA/sIbNSFdKyPFRS3K/tSt+eItzdL3cpi0zI8jEI/267iuoGVl1EHLHc7X/8gsqU933+TvfZ2mjZtsQ34GVkpUZV21q+ewlRYkAWG8vUFK1oH2fftPEag5lMbKSQzRpqGTBmYbQfLwLymEf6bqaeXE9jwlg1imJk+ZWAfoceMWFKwtLVkaEyJyVllLInEY42ydcFw0zI8jAIWmsjYiIGVl9EILDORtx9QSGBN20AbwdYjSO3yTTYKa0Y+k6gyEjLe5EVZYDBvzzO7aaRtMPYIgQ0zTEvwOsaahEwa2DyDeXueOU0jhyplX7syjlLxwIa3hpqW4XH6pIHNB2oM5u9pmlRWXnV7nq1GmqZ58wZpq883LcPDbAfmmsrcmIGVlxEDXjOVvx/4oGmUHdBqmMD6N+3wicy8JlHl2jyQyZhumHzZcP6eZkbD+OFx5d4koZYkVDwWXP2i7X3MjNFn2KiBlZexBlhkUoOXaVTZ+WvbBqwzraOvIrUrK231MSOfSFQZjS5jugQGthSWkf827BE0raGvEqh4zvYEZ8b4s+sFA5sB2EGbaVjQPHxUqwrYOFRu095SF9j43ijTMjxMI/CmaRHGDay8jGbgDdM6vIpCZGHzMNsb6TKBzR/awauZeVOiyngHh1cukPGiqJd5qW5yiVL21SvXUCoe/OwJGzonM554Zj1hYOVlLAFWmNbhVba09xu4um3gGtM6+gpSu3yNNG4oMq3Dw6yUqPrUtAjwiIE5/Mu0AC/zbO0+NhqCSwSXPGzPdWY886x6ycBeBWxjdRoq2oqGVsX6rTWto9fTuKEyULPEvjqUnnrg36ZFdOAZA3MiVDxlWoeXebF2L9ut38OElj1qw51n5mlTkSdS4RkDc3gO+35kWha2DBtV155teyR7itbtVYFNM+3QifRsB541LSIRTxmYM6TiL6Z1eJlnave14aZ7iOCSPxkfFuBx/ipR5an3cz1lYA5/B7aaFuFV5jWPGLM5lm9fL+puGjdWBjfMGGNahofZCrxoWkQynjOw8jJagahpHV7m0Zpp2UrZqem6DaVU6OP7QqZleJxHJao8V/r3nIE5vIKdei0tq9uKhqxsK1plWkdvQWqWVNiex4ysxENDJxLxpIGVl6GA35vW4WUeq5lWHFe0mdbhe1S8LbTw3kGmZXic30lUefJNEE8aGEB5GQuAd03r8CrV7fkD5jSNNBrKpDcQ2PjOammuGmBah4eZKVHl2WkQPWtgDg+CDeiXjr9u339Mczy0zbQO3xJr3BZc9L+24T49MfQz6Fk8bWDlZawHHjGtw6u0Ecx+vGZqg2kdfiX0caRe4q3ZpnV4mEclqjzd4+1pA3N4CjuTd1oWtgwbtayleKVpHX5Dti1eGaiaNdq0Dg+zAB+MyfS8gTkN+rcDtaa1eJWHtx08olUF7PnpLO0tdaF5N9tY9+mpA26XqPL8UB3PGxhAeRlbgHtM6/AqzSor98maqTWmdfiF4KL/3SaxBhtxIj33SlRtNi2iM/jCwADKy3gXeMm0Dq8yr3nEmI+bh9iYajtBNs9ZEdz4tm24T8+/JKreNi2is/jGwBx+B9jAfmn407aDx21vz9loWodnaaneGJp/6zjTMjzMOuC3pkV0BV8ZWHkZLcCt2KEVKWknELxv65F5MSV2kpRk4rHGrA+uyxXVbmd5Sk0MuMULce67gq8MDKC8jM+AP5nW4VWq2/MHPF4zbYt9VzIBpVRw0QObpWnjQNNSPMwjElXLTIvoKr4zMIengbmmRXiVec0jxsxpHmGHVjgENr27IrhhxljTOjzMfHwaTNSXBuYMrbgZsK/SpOHxmmnjN8b62/bChnVrgwvvHW9ahodZA0z3w5CJVPjSwADKy6gFfgrYCKUpUEjg7i1HDavtyxFcW2o2Zc36yRBRcdvulZoq4KcSVb4dQ+hbA4PPx4f9FBuGOiWtKpR955Zj+jXHg33vfclYU03WrKvzpb05x7QUj7IdbV6+GO+VDl8bGEB5GeuAn6OnOrckURvP7X/v1qPifapnMh5rDH3ws3ZpqS4wLcWjNALXSVT5fpYr3xsYfN4z+Uuw8bFSsTFWUPxg9aE1fSJ+mIrHQvNu3haoX21n1k5NG/AriaqlpoV0B73CwODz+GE3A54MvGaapa2DR/xfzYEbe7WJqXgsuPCe9YGt80ealuJR4uixXr0mOEKvMTD4/HWje03r8Crzm0eMfmTbQRvjilbTWrodFW8NLbhzQ3Dju/Y1ofTcJ1H1jmkR3UmvMjCA8jJeAR4yrcOrfNQyfPTD2w6p6lUmpuKtofm3VtnwOBl5WKLKk3HtdwdR/hz+sVPCEc4FrgDsbNYp2Cu7qvKyQbNLgqL83Uun2ltD827ZHNgy11YbU6OAByWqnjEtpCfotQYGEI7wFeB/gHzTWrzIntlbNlw+6IMBIVH+PD/xWGNo3k01ga0LRpiW4lGa0G1es0wL6Sl6tYEBhCNMAO4AhpjW4kWGhuqqf1r8HnmBmL9m5ok1Vme9fy3SsM5fut2jCrhBoqpXv1LW6w0MIByhCLgN2Mu0Fi+SL62N15W8s6041OSPalhTVWXWrLIiaavzZ8mx51kC/FKiqtq0kJ6mTxgYQDhCNnA9cLxpLV4kSLz9x4PeXz0pp3qCaS2ZkG2LV4Zm/2KsDYuTlhnAnV6cRbsn6DMGBhCOIMDFwPcMS/Es5xUuWn5sfsUEEY/1UCulAmv/tSL06UMTTUvxME8CUb++mL0r9CkD6yAc4UT060dZprV4kb2yqyrDg+YWZEt7oWktALS31IYW3FkX2DLHH1Vc94kB90hUvWZaiNv0SQMDCEeYgn79aJRpLV6kn7Q2Xl08a/OIrDqjcbSkfvWa0Ic3lNj2rrRUArdJVH1qWogJ+qyBAYQj5ACXAmdjx4ul5PSCJStO6rd8bEAIuZqxircHVj2/KvTZY7bKmBoF/B14qK+0d6WiTxtYB+EI09AN/HaoRQrGZm2r+lHRbPoHW905P621m0Pzbo4Hti8d6kp+/qMKuEuiap5pIaaxBuYQjpAPXAachi2NfYkA8fazCz9ddWz+qjEBIbtHMlHx1sCal9YElz4y3gYhTIkCXkaXuhpMi/EC1sCSCEfYB7gG8PRwAlOUBBtqLhs0u3ZYqL57X5puqFybNe+m/tK4oahb99t7WAXcL1G1yLQQL2ENLAXhCEHgm+ghF/5+V7CHOCq/ouLcwsXFWRLfvaCB7a31waWPbA6u/ZeNW5+aFuBx4GmJKjudYBLWwDIQjjAM+D5wEmCrNElkS6z1nILFaw/PXzsyKCq3S1+OtzcHKv9TGVz6yCgb9jkl7cDrwKMSVRtMi/Eq1sA6QTjCcOBC4Ovgcm+cD+gnrY3nD1i44YDcjWMCspOxdSoek6r3V4c++d1wOzQiJTHgVeAv1rh2jjWwLhCOMBg4H93Q3zMN2T6mKNBYe9HAj7bsmb113JdG8isVl22LKkIfR4qluWqAIYlephXdQP83iaoq02L8gjWwXSAcYRDwLeBMoGtVpz5AcbCx5tzCT7bsm7NpTIB2ZPPctaEl5cV2ZuyUNAP/AJ7qCy9fdzfWwHaDcIQB6Mb+s7Exx5LZOCRY/9wty6Yoaag8FxhuWpDHaAReAJ6RqNpuWoxfsQbWDYQj9AeOA74KHEAvDNXdSRQwG3gReN+ZQR1VKgIcBpwFHErfHWengAXAG8AMiap6s3L8jzWwbsaJPXYs2sz2pW88rJ8Bs4B/l5exPlNCVSrDgZOBI4BJLmgzjQI+QZvWW7aa2L1YA+tBnEb/49BmNsWsmm6lFZgHzARmOTOkdxlVKiXA4Wgzm0bvGnO3BHgTXdKyjfI9hDUwl3CGYhwPHIkuefgtlM9WdClrFjC3vIxufYFYlUoOcBDa0A4H/DYxbRuwHHgPeMMOgXAHa2AGCEcIAXugS2WT0aGux+Kd6qZCvzBcgS5JzAKWdbRp9Xjmus1sEtrIpgDjgKF46/ysRp+bjmWFHSnvPtbAPEI4Qh76od2LHcbmRs/dZrRRVaDft1sFrC4vo8mFvDuNKpU8tMmPRxtax9/BLmS/kS+a1TKJKk+dn76KNTAPE46QC5Sgq1Mdf4uBAqB/wtIP3X7UhB5X1LE0pVlXiy5BVJSX4euoBqpU+qGNbCxQCOShx+Z1LHlp1rUC9QlLA1AHbEFXlz//K1HV7NoBWbqENTCLxeJb+up4JYvF0guwBmaxWHyLNTCLxeJbrIFZLBbfYg3MYrH4FmtgFovFt1gDs1gsvsUamMVi8S3WwCyWHkBEHhWRW03r6O1YA+siInKhiMwRkXoR2SAir4jIUSIyXUSeTJFeichE5//pzuerktKUOeunO5+zReRZEalw1h+XIv1KEakVkfUi8hsRsZONdBEROV9EPhCRBhGpcv6/XDSPikirc52rReQ/IjIl4bsiIleJyCLn++tE5BkR2a+Teb8pIpuda/iRiJyZtP1KEVnlbJ8jIkd19/H3BqyBdQERuQaIALejoyOMAf6Ajo3fWZahp2pL5HvO+kTeBS5Cv0iczD+BA5VSheigiQcAV6VIZ0mDiFwLPADcAwxDX8/L0OGOOiZsuVsp1R8YCVQCjyTs4gHgavR5HwTsiY5Ee2onJVwNDHeu4aXAkyIy3NF2GHAncB4wwMn3BRGxU/slYQ2sk4jIAOBm4Aql1PNKqQalVJtS6p9KqZ91YVezgXwR2cfZ7z7ol4tndyRQSrUqpSJKqXfR8wN+AaXUCqVUTYc0IA5M3KUD64MkXMvLlVLPKqXqlGa+Uuo7SqkvxDpTSjUBTwNTne9PAq4ALlBKvaGUalFKNSql/qyUujPhq0Ui8rKI1Dmluz0S9rlQqc/D7yh0fLjRzudxwCdKqblKv6z8OPpl/iHdfCp8jzWwznM4OpLBC92wryfQpS7QpbHHu7oDpypbi46acADwx27Q1Vc4HB294++dSSwi/YAL0AELAU4A1imlPtzJVy8AbgKKnO/elrTfl0SkGfgAmAHMcTa9AgRF5DCn1FWKjqWfqjTep7EG1nmKgS0Jv5qp+JaI1CQuadI9CVwgIlnoeSa/1Ha2M5RSf3GqH3sCDwGburqPPkwJSddSRGY616xJRI5xVv/UuYZ1wFHAd531xUBnIq4+r5T60MnnzzgluA6UUqehQyOdAvxbKRV3NtUBz6GbEVqAXwOXKhs65ktYA+s8W4GSnTSWP62UGpi4pEqklFqD/kW+HfhMKbV2V0UppT5DTxrxh13dRx/kS9dSKXWEc722suO5uNdZNw4dR21ywvc7E2wyscTUiI7d9gWcZohXgJNF5Axn9SXoUtc+6Pa4i4CXRGREZw6uL2ENrPPMQgcEPKub9vc4cC27UH1MQUeIakvnmIUu2XSq88X5wbkaeEBE8oDXgVEicnA3akq8hgcA/1RKLVNKxZVSr6JLfEd0Y369AmtgnUQptR24Efi9iJwlIvkikiUi3xCRu3dhl08BX0M3Dn8JEckRkY5Zv7NFJFdExNl2iYgMcf7fG7gB/VBZOoHTAXIT8AcROU9E+otIQESmoqPbpvrOf4D16KrcZ+gS719F5Dhn2EuuMyzj+p3lLyJTnPsmz7mHLgKOAd5ykswGThWRCc5wjZPQTQWLdvPQex127FAXUErdLyKbgF+h2zTqgLnoxtmvdXFfTcB/MyRZig6TDPBv5+94dOz6I4HbRKQ/Oqb9M8D/dCX/vo5S6m4RqQR+ji4FNwArgevQ08VdnOJr9wD3i8hD6OETVwG/R1+Xbeg2q5s7kb0A04G90b3MnwHfVkrNc7Y/ji6NzUB3AKwD/p9SakkXD7PXY0NKWywW32KrkBaLxbdYA7NYLL7FGpjFYvEt1sAsFotvsQZmsVh8izUwi8XiW6yBWSwW32INzGKx+BZrYBaLxbf8f5PDh462Uee0AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Make venn diagram for genes expressing novel isoforms.\n",
    "\n",
    "chm13_novel_transcripts[\"GENEID\"] = chm13_novel_transcripts[\"GENEID\"].str.split(\".\", expand=True)[0]\n",
    "\n",
    "chm13_set = chm13_novel_transcripts[\"GENEID\"]\n",
    "grch_set = grch_novel_transcripts[\"GENEID\"]\n",
    "\n",
    "venn2_unweighted(subsets = [set(chm13_set), set(grch_set)], set_labels = ('CHM13', 'GRCh38'), \n",
    "                 set_colors=('#648FFF', '#FE6100'), alpha = 0.75);\n",
    "\n",
    "plt.title(\"Genes with Novel Transcripts\", size=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save novel transcript in known genes for GRCh38 so that we can compare them to known trancripts in ENSEMBL tracks\n",
    "chm13_ref = pd.read_csv(\"../../data/novel_genes/chm13_bambu/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])\n",
    "\n",
    "\n",
    "chm13_ref = parse_df_columns(chm13_ref, is_ref=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11136.0</td>\n",
       "      <td>12457.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000239945.1(LOFF_G0000001)</td>\n",
       "      <td>ENST00000495576.1(LOFF_T0000001)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11136.0</td>\n",
       "      <td>11635.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000239945.1(LOFF_G0000001)</td>\n",
       "      <td>ENST00000495576.1(LOFF_T0000001)</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11630.0</td>\n",
       "      <td>13433.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000286448.1(LOFF_G0000002)</td>\n",
       "      <td>ENST00000669836.1(LOFF_T0000002)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11630.0</td>\n",
       "      <td>11831.0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000286448.1(LOFF_G0000002)</td>\n",
       "      <td>ENST00000669836.1(LOFF_T0000002)</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11639.0</td>\n",
       "      <td>12457.0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000239945.1(LOFF_G0000001)</td>\n",
       "      <td>ENST00000495576.1(LOFF_T0000001)</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr source        type    start      end strand  \\\n",
       "0  chr1  Bambu  transcript  11136.0  12457.0      -   \n",
       "1  chr1  Bambu        exon  11136.0  11635.0      -   \n",
       "2  chr1  Bambu  transcript  11630.0  13433.0      +   \n",
       "3  chr1  Bambu        exon  11630.0  11831.0      +   \n",
       "4  chr1  Bambu        exon  11639.0  12457.0      -   \n",
       "\n",
       "                            gene_id                     transcript_id  \\\n",
       "0  ENSG00000239945.1(LOFF_G0000001)  ENST00000495576.1(LOFF_T0000001)   \n",
       "1  ENSG00000239945.1(LOFF_G0000001)  ENST00000495576.1(LOFF_T0000001)   \n",
       "2  ENSG00000286448.1(LOFF_G0000002)  ENST00000669836.1(LOFF_T0000002)   \n",
       "3  ENSG00000286448.1(LOFF_G0000002)  ENST00000669836.1(LOFF_T0000002)   \n",
       "4  ENSG00000239945.1(LOFF_G0000001)  ENST00000495576.1(LOFF_T0000001)   \n",
       "\n",
       "  exon_number is_novel_transcript is_novel_gene  \n",
       "0         NaN               False         False  \n",
       "1           2               False         False  \n",
       "2         NaN               False         False  \n",
       "3           1               False         False  \n",
       "4           1               False         False  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_ref[\"gene_id\"] = chm13_ref[\"gene_id\"].str.split(\".\", expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_ref = chm13_ref.loc[chm13_ref[\"type\"] == \"transcript\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_novel_transcripts_expressed = chm13_ref.loc[chm13_ref[\"transcript_id\"].isin(chm13_novel_transcripts[\"TXNAME\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_relevant_genes = pd.read_csv(\"../../../cDNA-comparison/article_analysis/annotations/disease_relevant_gene_names.csv\", names=[\"gene_id\", \"gene_name\"],\n",
    "                                    header=0)\n",
    "\n",
    "## Create list of novel transcripts in disease relevant genes\n",
    "novel_relevant = pd.merge(disease_relevant_genes, chm13_novel_transcripts_expressed, how='inner', on=\"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000027847</td>\n",
       "      <td>B4GALT7</td>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>178142421.0</td>\n",
       "      <td>178147501.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000054282</td>\n",
       "      <td>SDCCAG8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>242667072.0</td>\n",
       "      <td>242717514.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000067225</td>\n",
       "      <td>PKM</td>\n",
       "      <td>chr15</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>70015757.0</td>\n",
       "      <td>70047905.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.497</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000070501</td>\n",
       "      <td>POLB</td>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>42608550.0</td>\n",
       "      <td>42642104.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000091592</td>\n",
       "      <td>NLRP1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5393165.0</td>\n",
       "      <td>5394285.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name    chr source        type        start  \\\n",
       "0  ENSG00000027847   B4GALT7   chr5  Bambu  transcript  178142421.0   \n",
       "1  ENSG00000054282   SDCCAG8   chr1  Bambu  transcript  242667072.0   \n",
       "2  ENSG00000067225       PKM  chr15  Bambu  transcript   70015757.0   \n",
       "3  ENSG00000070501      POLB   chr8  Bambu  transcript   42608550.0   \n",
       "4  ENSG00000091592     NLRP1  chr17  Bambu  transcript    5393165.0   \n",
       "\n",
       "           end strand transcript_id exon_number is_novel_transcript  \\\n",
       "0  178147501.0      +        tx.230         NaN                True   \n",
       "1  242717514.0      +         tx.63         NaN                True   \n",
       "2   70047905.0      -        tx.497         NaN                True   \n",
       "3   42642104.0      +        tx.316         NaN                True   \n",
       "4    5394285.0      -        tx.531         NaN                True   \n",
       "\n",
       "  is_novel_gene  \n",
       "0         False  \n",
       "1         False  \n",
       "2         False  \n",
       "3         False  \n",
       "4         False  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_relevant.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "neuro_disease = pd.read_csv(\"../../data/novel_genes/psych_neuro_relevant_genes_list_bah_2021-09-07.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "neuro_disease = neuro_disease[\"Gene Name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'neuro_disease_expressed' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-61-518fe8de796f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mneuro_disease_expressed\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'neuro_disease_expressed' is not defined"
     ]
    }
   ],
   "source": [
    "neuro_disease_expressed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "neuro_disease_expressed = novel_relevant.loc[novel_relevant[\"gene_name\"].isin(neuro_disease)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000091592</td>\n",
       "      <td>NLRP1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5393165.0</td>\n",
       "      <td>5394285.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000114956</td>\n",
       "      <td>DGUOK</td>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>73935343.0</td>\n",
       "      <td>73967416.0</td>\n",
       "      <td>+</td>\n",
       "      <td>tx.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000165092</td>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>85051589.0</td>\n",
       "      <td>85231348.0</td>\n",
       "      <td>-</td>\n",
       "      <td>tx.349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id gene_name    chr source        type       start  \\\n",
       "4   ENSG00000091592     NLRP1  chr17  Bambu  transcript   5393165.0   \n",
       "8   ENSG00000114956     DGUOK   chr2  Bambu  transcript  73935343.0   \n",
       "28  ENSG00000165092   ALDH1A1   chr9  Bambu  transcript  85051589.0   \n",
       "\n",
       "           end strand transcript_id exon_number is_novel_transcript  \\\n",
       "4    5394285.0      -        tx.531         NaN                True   \n",
       "8   73967416.0      +         tx.91         NaN                True   \n",
       "28  85231348.0      -        tx.349         NaN                True   \n",
       "\n",
       "   is_novel_gene  \n",
       "4          False  \n",
       "8          False  \n",
       "28         False  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neuro_disease_expressed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_relevant.drop(columns=[\"gene_id\", \"gene_name\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(44, 10)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_relevant.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_relevant.to_csv(\"../../data/novel_genes/novel_med_relevant_transcript_annotation_chm13.gtf\", \n",
    "                                        sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_ref = pd.read_csv(\"../novel_gene_body_raw_data/chm13_bambu/extended_annotations.gtf\", header=None, delimiter=\"\\t\",\n",
    "                        low_memory=False, names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>dot_1</th>\n",
       "      <th>strand</th>\n",
       "      <th>dot_2</th>\n",
       "      <th>other</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11136</td>\n",
       "      <td>12457</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000239945.1(LOFF_G0000001)\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11136</td>\n",
       "      <td>11635</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000239945.1(LOFF_G0000001)\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11630</td>\n",
       "      <td>13433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000286448.1(LOFF_G0000002)\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11630</td>\n",
       "      <td>11831</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000286448.1(LOFF_G0000002)\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>exon</td>\n",
       "      <td>11639</td>\n",
       "      <td>12457</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000239945.1(LOFF_G0000001)\"; tr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr source        type  start    end dot_1 strand dot_2  \\\n",
       "0  chr1  Bambu  transcript  11136  12457     .      -     .   \n",
       "1  chr1  Bambu        exon  11136  11635     .      -     .   \n",
       "2  chr1  Bambu  transcript  11630  13433     .      +     .   \n",
       "3  chr1  Bambu        exon  11630  11831     .      +     .   \n",
       "4  chr1  Bambu        exon  11639  12457     .      -     .   \n",
       "\n",
       "                                               other  \n",
       "0  gene_id \"ENSG00000239945.1(LOFF_G0000001)\"; tr...  \n",
       "1  gene_id \"ENSG00000239945.1(LOFF_G0000001)\"; tr...  \n",
       "2  gene_id \"ENSG00000286448.1(LOFF_G0000002)\"; tr...  \n",
       "3  gene_id \"ENSG00000286448.1(LOFF_G0000002)\"; tr...  \n",
       "4  gene_id \"ENSG00000239945.1(LOFF_G0000001)\"; tr...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_ref.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13_ref = chm13_ref.loc[chm13_ref[\"type\"] == \"transcript\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>is_novel_transcript</th>\n",
       "      <th>is_novel_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17202</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16245735.0</td>\n",
       "      <td>16247595.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.51</td>\n",
       "      <td>tx.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26029</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26268774.0</td>\n",
       "      <td>26270140.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.64</td>\n",
       "      <td>tx.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27377</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>27231889.0</td>\n",
       "      <td>27233743.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.66</td>\n",
       "      <td>tx.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29806</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29948153.0</td>\n",
       "      <td>29953421.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.68</td>\n",
       "      <td>tx.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30273</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>31007199.0</td>\n",
       "      <td>31007976.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.5</td>\n",
       "      <td>tx.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36935</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>38167408.0</td>\n",
       "      <td>38300289.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.72</td>\n",
       "      <td>tx.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36936</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>38167408.0</td>\n",
       "      <td>38300320.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.72</td>\n",
       "      <td>tx.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61708</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>68243853.0</td>\n",
       "      <td>68261649.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.76</td>\n",
       "      <td>tx.22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64998</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>77862296.0</td>\n",
       "      <td>77863997.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.13</td>\n",
       "      <td>tx.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85157</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120334901.0</td>\n",
       "      <td>120336836.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.17</td>\n",
       "      <td>tx.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91413</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148917243.0</td>\n",
       "      <td>148922629.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.19</td>\n",
       "      <td>tx.37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92695</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>149753837.0</td>\n",
       "      <td>149759779.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.22</td>\n",
       "      <td>tx.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98730</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>154007468.0</td>\n",
       "      <td>154010460.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.31</td>\n",
       "      <td>tx.41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111178</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>161794715.0</td>\n",
       "      <td>161832275.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.79</td>\n",
       "      <td>tx.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117401</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>175875912.0</td>\n",
       "      <td>175907502.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.81</td>\n",
       "      <td>tx.47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125445</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>200352725.0</td>\n",
       "      <td>200355594.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.35</td>\n",
       "      <td>tx.51</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131400</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>204819539.0</td>\n",
       "      <td>204833822.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.94</td>\n",
       "      <td>tx.53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132960</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>206395084.0</td>\n",
       "      <td>206587072.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.98</td>\n",
       "      <td>tx.54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135005</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>210503378.0</td>\n",
       "      <td>210505309.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.102</td>\n",
       "      <td>tx.56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140388</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>223623753.0</td>\n",
       "      <td>223710822.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.37</td>\n",
       "      <td>tx.58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144983</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>230181846.0</td>\n",
       "      <td>230187587.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.105</td>\n",
       "      <td>tx.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146378</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>232706207.0</td>\n",
       "      <td>232713755.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.107</td>\n",
       "      <td>tx.61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153290</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>246706575.0</td>\n",
       "      <td>246719608.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.42</td>\n",
       "      <td>tx.67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153292</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>246706576.0</td>\n",
       "      <td>246719608.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.42</td>\n",
       "      <td>tx.66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155249</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2056961.0</td>\n",
       "      <td>2057421.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.601</td>\n",
       "      <td>tx.372</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155700</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3271431.0</td>\n",
       "      <td>3278712.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.603</td>\n",
       "      <td>tx.374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155702</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3278199.0</td>\n",
       "      <td>3283710.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.570</td>\n",
       "      <td>tx.375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155705</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3278380.0</td>\n",
       "      <td>3294924.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.570</td>\n",
       "      <td>tx.376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171270</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>44529886.0</td>\n",
       "      <td>44530712.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.578</td>\n",
       "      <td>tx.379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189091</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>78505098.0</td>\n",
       "      <td>78505743.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.585</td>\n",
       "      <td>tx.388</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193716</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>92435427.0</td>\n",
       "      <td>92441761.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.590</td>\n",
       "      <td>tx.394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194879</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>94625049.0</td>\n",
       "      <td>94637789.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.611</td>\n",
       "      <td>tx.395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208300</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>117366422.0</td>\n",
       "      <td>117367340.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.616</td>\n",
       "      <td>tx.396</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215889</th>\n",
       "      <td>chr10</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>133672119.0</td>\n",
       "      <td>133686919.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.618</td>\n",
       "      <td>tx.400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226187</th>\n",
       "      <td>chr11</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6443663.0</td>\n",
       "      <td>6448780.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.643</td>\n",
       "      <td>tx.405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269596</th>\n",
       "      <td>chr11</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>66284625.0</td>\n",
       "      <td>66285418.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.647</td>\n",
       "      <td>tx.415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273498</th>\n",
       "      <td>chr11</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>68008808.0</td>\n",
       "      <td>68010284.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.633</td>\n",
       "      <td>tx.416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279345</th>\n",
       "      <td>chr11</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72722034.0</td>\n",
       "      <td>72738805.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.635</td>\n",
       "      <td>tx.417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308884</th>\n",
       "      <td>chr11</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>125265652.0</td>\n",
       "      <td>125268543.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.651</td>\n",
       "      <td>tx.424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313300</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>58097.0</td>\n",
       "      <td>58757.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.679</td>\n",
       "      <td>tx.425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350302</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>51969337.0</td>\n",
       "      <td>51970360.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.683</td>\n",
       "      <td>tx.428</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357784</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>56071418.0</td>\n",
       "      <td>56071983.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.687</td>\n",
       "      <td>tx.429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370756</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>71649099.0</td>\n",
       "      <td>71663543.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.689</td>\n",
       "      <td>tx.433</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374900</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>84966212.0</td>\n",
       "      <td>84972081.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.665</td>\n",
       "      <td>tx.434</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388906</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>110427575.0</td>\n",
       "      <td>110433747.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.668</td>\n",
       "      <td>tx.439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389681</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>111010612.0</td>\n",
       "      <td>111014762.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.695</td>\n",
       "      <td>tx.440</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404578</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>130046464.0</td>\n",
       "      <td>130066361.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.674</td>\n",
       "      <td>tx.444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407118</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>133013744.0</td>\n",
       "      <td>133041046.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.676</td>\n",
       "      <td>tx.446</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407120</th>\n",
       "      <td>chr12</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>133013753.0</td>\n",
       "      <td>133035037.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.676</td>\n",
       "      <td>tx.447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417849</th>\n",
       "      <td>chr13</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>40766018.0</td>\n",
       "      <td>40768303.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.716</td>\n",
       "      <td>tx.455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421677</th>\n",
       "      <td>chr13</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48425802.0</td>\n",
       "      <td>48469158.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.718</td>\n",
       "      <td>tx.457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421693</th>\n",
       "      <td>chr13</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48456818.0</td>\n",
       "      <td>48469157.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.718</td>\n",
       "      <td>tx.458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426221</th>\n",
       "      <td>chr13</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>66423827.0</td>\n",
       "      <td>66433257.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.706</td>\n",
       "      <td>tx.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430190</th>\n",
       "      <td>chr13</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>94505657.0</td>\n",
       "      <td>94515582.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.710</td>\n",
       "      <td>tx.464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447602</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>18265046.0</td>\n",
       "      <td>18270516.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.752</td>\n",
       "      <td>tx.472</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456102</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>32395914.0</td>\n",
       "      <td>32397985.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.755</td>\n",
       "      <td>tx.477</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484306</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>86904499.0</td>\n",
       "      <td>86919989.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.759</td>\n",
       "      <td>tx.482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487013</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90175238.0</td>\n",
       "      <td>90180287.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.744</td>\n",
       "      <td>tx.483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494251</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>98898949.0</td>\n",
       "      <td>98899910.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.746</td>\n",
       "      <td>tx.484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494253</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>98898967.0</td>\n",
       "      <td>98899911.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.746</td>\n",
       "      <td>tx.485</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495833</th>\n",
       "      <td>chr14</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>99548453.0</td>\n",
       "      <td>99554926.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.748</td>\n",
       "      <td>tx.486</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512179</th>\n",
       "      <td>chr15</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>37841702.0</td>\n",
       "      <td>37844407.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.767</td>\n",
       "      <td>tx.490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514644</th>\n",
       "      <td>chr15</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>39299216.0</td>\n",
       "      <td>39300931.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.783</td>\n",
       "      <td>tx.493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521539</th>\n",
       "      <td>chr15</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>42975296.0</td>\n",
       "      <td>42996825.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.771</td>\n",
       "      <td>tx.494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555183</th>\n",
       "      <td>chr15</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>87252266.0</td>\n",
       "      <td>87253187.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.778</td>\n",
       "      <td>tx.502</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568741</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1963459.0</td>\n",
       "      <td>1964607.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.786</td>\n",
       "      <td>tx.505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572961</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2788183.0</td>\n",
       "      <td>2789466.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.802</td>\n",
       "      <td>tx.506</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578543</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>7463970.0</td>\n",
       "      <td>7473639.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.804</td>\n",
       "      <td>tx.509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582226</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14524554.0</td>\n",
       "      <td>14526611.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.788</td>\n",
       "      <td>tx.510</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610001</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>62456161.0</td>\n",
       "      <td>62458251.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.806</td>\n",
       "      <td>tx.516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612575</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>63530738.0</td>\n",
       "      <td>63531850.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.808</td>\n",
       "      <td>tx.517</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614745</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>67824318.0</td>\n",
       "      <td>67825229.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.792</td>\n",
       "      <td>tx.518</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629577</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>83064170.0</td>\n",
       "      <td>83164948.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.810</td>\n",
       "      <td>tx.520</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634074</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>92127704.0</td>\n",
       "      <td>92128126.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.798</td>\n",
       "      <td>tx.523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634544</th>\n",
       "      <td>chr16</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>92829633.0</td>\n",
       "      <td>92832182.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.818</td>\n",
       "      <td>tx.524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649181</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6138050.0</td>\n",
       "      <td>6141095.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.829</td>\n",
       "      <td>tx.532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662261</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17699000.0</td>\n",
       "      <td>17701276.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.832</td>\n",
       "      <td>tx.534</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673585</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>31608319.0</td>\n",
       "      <td>31632258.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.848</td>\n",
       "      <td>tx.535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683404</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>40984434.0</td>\n",
       "      <td>40985750.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.855</td>\n",
       "      <td>tx.540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691105</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>44866785.0</td>\n",
       "      <td>44872362.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.838</td>\n",
       "      <td>tx.544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706520</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>59381670.0</td>\n",
       "      <td>59384925.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.840</td>\n",
       "      <td>tx.549</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715546</th>\n",
       "      <td>chr17</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>69085396.0</td>\n",
       "      <td>69089945.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.842</td>\n",
       "      <td>tx.555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742119</th>\n",
       "      <td>chr18</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10610363.0</td>\n",
       "      <td>10617136.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.879</td>\n",
       "      <td>tx.562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748338</th>\n",
       "      <td>chr18</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26741726.0</td>\n",
       "      <td>26749456.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.865</td>\n",
       "      <td>tx.566</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752121</th>\n",
       "      <td>chr18</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>37759890.0</td>\n",
       "      <td>37760308.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.870</td>\n",
       "      <td>tx.568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771656</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1413060.0</td>\n",
       "      <td>1417349.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.920</td>\n",
       "      <td>tx.570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774109</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2467933.0</td>\n",
       "      <td>2471009.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.923</td>\n",
       "      <td>tx.572</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774111</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2469533.0</td>\n",
       "      <td>2471009.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.923</td>\n",
       "      <td>tx.573</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777360</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4568029.0</td>\n",
       "      <td>4569755.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.892</td>\n",
       "      <td>tx.574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777362</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4568032.0</td>\n",
       "      <td>4569755.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.892</td>\n",
       "      <td>tx.575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783220</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>7920226.0</td>\n",
       "      <td>7921212.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.928</td>\n",
       "      <td>tx.576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793397</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12845394.0</td>\n",
       "      <td>12847086.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.895</td>\n",
       "      <td>tx.581</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804559</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>18256654.0</td>\n",
       "      <td>18257704.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.930</td>\n",
       "      <td>tx.586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813325</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>36042689.0</td>\n",
       "      <td>36046532.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.900</td>\n",
       "      <td>tx.592</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847914</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>54680415.0</td>\n",
       "      <td>54683045.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.941</td>\n",
       "      <td>tx.607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848477</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>55095524.0</td>\n",
       "      <td>55109767.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.911</td>\n",
       "      <td>tx.608</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854521</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>58325760.0</td>\n",
       "      <td>58326123.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.915</td>\n",
       "      <td>tx.609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864944</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10101554.0</td>\n",
       "      <td>10103412.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.113</td>\n",
       "      <td>tx.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891191</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>61799577.0</td>\n",
       "      <td>61804796.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.119</td>\n",
       "      <td>tx.84</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899117</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>73297500.0</td>\n",
       "      <td>73299117.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.123</td>\n",
       "      <td>tx.88</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907354</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91752692.0</td>\n",
       "      <td>91756493.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.125</td>\n",
       "      <td>tx.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911514</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>98630427.0</td>\n",
       "      <td>98634337.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.178</td>\n",
       "      <td>tx.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911515</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>98630427.0</td>\n",
       "      <td>98634503.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.178</td>\n",
       "      <td>tx.95</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913734</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>102075198.0</td>\n",
       "      <td>102082766.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.136</td>\n",
       "      <td>tx.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921797</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>119714686.0</td>\n",
       "      <td>119716307.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.185</td>\n",
       "      <td>tx.104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>922150</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120157246.0</td>\n",
       "      <td>120193844.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.188</td>\n",
       "      <td>tx.105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926396</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>130955720.0</td>\n",
       "      <td>130957449.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.191</td>\n",
       "      <td>tx.106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935270</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>153100945.0</td>\n",
       "      <td>153102389.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.193</td>\n",
       "      <td>tx.110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939774</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>165751335.0</td>\n",
       "      <td>165753268.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.195</td>\n",
       "      <td>tx.111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946553</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>175288147.0</td>\n",
       "      <td>175336308.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.153</td>\n",
       "      <td>tx.113</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961734</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>203013644.0</td>\n",
       "      <td>203037175.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.197</td>\n",
       "      <td>tx.120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961753</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>203034011.0</td>\n",
       "      <td>203037177.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.197</td>\n",
       "      <td>tx.121</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974730</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>225594940.0</td>\n",
       "      <td>225595646.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.161</td>\n",
       "      <td>tx.122</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>980978</th>\n",
       "      <td>chr2</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>235870026.0</td>\n",
       "      <td>235880218.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.210</td>\n",
       "      <td>tx.125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990124</th>\n",
       "      <td>chr20</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2673938.0</td>\n",
       "      <td>2682913.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.959</td>\n",
       "      <td>tx.618</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998062</th>\n",
       "      <td>chr20</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21575784.0</td>\n",
       "      <td>21579289.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.961</td>\n",
       "      <td>tx.620</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998923</th>\n",
       "      <td>chr20</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>23673408.0</td>\n",
       "      <td>23673932.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.963</td>\n",
       "      <td>tx.621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013489</th>\n",
       "      <td>chr20</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>46678603.0</td>\n",
       "      <td>46679406.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.948</td>\n",
       "      <td>tx.626</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1016063</th>\n",
       "      <td>chr20</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48142813.0</td>\n",
       "      <td>48180597.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.950</td>\n",
       "      <td>tx.627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1023504</th>\n",
       "      <td>chr20</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>63753479.0</td>\n",
       "      <td>63756724.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.955</td>\n",
       "      <td>tx.631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029111</th>\n",
       "      <td>chr21</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12726210.0</td>\n",
       "      <td>12736277.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.967</td>\n",
       "      <td>tx.633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1031905</th>\n",
       "      <td>chr21</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>22480907.0</td>\n",
       "      <td>22511040.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.969</td>\n",
       "      <td>tx.637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034200</th>\n",
       "      <td>chr21</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30240182.0</td>\n",
       "      <td>30241210.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.975</td>\n",
       "      <td>tx.638</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1045610</th>\n",
       "      <td>chr21</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>43713514.0</td>\n",
       "      <td>43719491.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.973</td>\n",
       "      <td>tx.643</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057067</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24441306.0</td>\n",
       "      <td>24449445.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>tx.652</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057068</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24441306.0</td>\n",
       "      <td>24449438.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>tx.653</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057069</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24441306.0</td>\n",
       "      <td>24449440.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>tx.654</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057070</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24441306.0</td>\n",
       "      <td>24449457.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.989</td>\n",
       "      <td>tx.655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059717</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>27094628.0</td>\n",
       "      <td>27109928.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.994</td>\n",
       "      <td>tx.656</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060073</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>28134962.0</td>\n",
       "      <td>28138620.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.981</td>\n",
       "      <td>tx.657</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1061738</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29952946.0</td>\n",
       "      <td>29954661.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.997</td>\n",
       "      <td>tx.659</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1067845</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>33531107.0</td>\n",
       "      <td>33538458.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.999</td>\n",
       "      <td>tx.661</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081056</th>\n",
       "      <td>chr22</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50134537.0</td>\n",
       "      <td>50135382.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.987</td>\n",
       "      <td>tx.662</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084613</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2091485.0</td>\n",
       "      <td>2092562.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.244</td>\n",
       "      <td>tx.126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091894</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11715305.0</td>\n",
       "      <td>11725053.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.214</td>\n",
       "      <td>tx.129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100230</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>33277040.0</td>\n",
       "      <td>33279143.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.250</td>\n",
       "      <td>tx.133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101471</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>35639358.0</td>\n",
       "      <td>35640647.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.252</td>\n",
       "      <td>tx.134</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110109</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>46011738.0</td>\n",
       "      <td>46017127.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.222</td>\n",
       "      <td>tx.142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110111</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>46011751.0</td>\n",
       "      <td>46017127.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.222</td>\n",
       "      <td>tx.143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1132806</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>71621703.0</td>\n",
       "      <td>71622859.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.229</td>\n",
       "      <td>tx.149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1145454</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>118681174.0</td>\n",
       "      <td>118698136.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.233</td>\n",
       "      <td>tx.154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1153520</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>131928955.0</td>\n",
       "      <td>131941197.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.237</td>\n",
       "      <td>tx.160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162885</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>147500963.0</td>\n",
       "      <td>147510850.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.254</td>\n",
       "      <td>tx.163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1186901</th>\n",
       "      <td>chr3</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>197080131.0</td>\n",
       "      <td>197091651.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.242</td>\n",
       "      <td>tx.168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201674</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10063511.0</td>\n",
       "      <td>10072577.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.258</td>\n",
       "      <td>tx.174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204051</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17464779.0</td>\n",
       "      <td>17467542.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.277</td>\n",
       "      <td>tx.175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205196</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>22708168.0</td>\n",
       "      <td>22708552.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.279</td>\n",
       "      <td>tx.176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205800</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25171560.0</td>\n",
       "      <td>25215801.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.281</td>\n",
       "      <td>tx.177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1218292</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>74241667.0</td>\n",
       "      <td>74244666.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.287</td>\n",
       "      <td>tx.180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1221372</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>80495548.0</td>\n",
       "      <td>80497541.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.292</td>\n",
       "      <td>tx.183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1221373</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>80495548.0</td>\n",
       "      <td>80496517.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.292</td>\n",
       "      <td>tx.184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251415</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>154493284.0</td>\n",
       "      <td>154497983.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.268</td>\n",
       "      <td>tx.194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255614</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>162346996.0</td>\n",
       "      <td>162399093.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.298</td>\n",
       "      <td>tx.195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1257794</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>172729385.0</td>\n",
       "      <td>172749149.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.300</td>\n",
       "      <td>tx.196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1263598</th>\n",
       "      <td>chr4</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>189834257.0</td>\n",
       "      <td>189837658.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.272</td>\n",
       "      <td>tx.197</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1271493</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24750070.0</td>\n",
       "      <td>24750707.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.304</td>\n",
       "      <td>tx.200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278222</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>43427116.0</td>\n",
       "      <td>43444990.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.328</td>\n",
       "      <td>tx.201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1287731</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72679720.0</td>\n",
       "      <td>72687282.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.332</td>\n",
       "      <td>tx.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1288326</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>73727808.0</td>\n",
       "      <td>73756224.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.335</td>\n",
       "      <td>tx.206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302651</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>113062662.0</td>\n",
       "      <td>113064001.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.306</td>\n",
       "      <td>tx.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312281</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>135401779.0</td>\n",
       "      <td>135402332.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.341</td>\n",
       "      <td>tx.218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1328350</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>158094251.0</td>\n",
       "      <td>158095637.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.313</td>\n",
       "      <td>tx.220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329641</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>162337156.0</td>\n",
       "      <td>162379506.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.345</td>\n",
       "      <td>tx.222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330428</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>164101947.0</td>\n",
       "      <td>164266752.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.316</td>\n",
       "      <td>tx.225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1333368</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>174504523.0</td>\n",
       "      <td>174505895.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.318</td>\n",
       "      <td>tx.226</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335939</th>\n",
       "      <td>chr5</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>178002662.0</td>\n",
       "      <td>178003592.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.324</td>\n",
       "      <td>tx.228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1342763</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3324738.0</td>\n",
       "      <td>3338104.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.351</td>\n",
       "      <td>tx.237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1346946</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>13201864.0</td>\n",
       "      <td>13203102.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.354</td>\n",
       "      <td>tx.238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1347642</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17204392.0</td>\n",
       "      <td>17231356.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.356</td>\n",
       "      <td>tx.239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1348823</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21008097.0</td>\n",
       "      <td>21013056.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.379</td>\n",
       "      <td>tx.241</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1348825</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21010120.0</td>\n",
       "      <td>21013113.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.379</td>\n",
       "      <td>tx.240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362371</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>32716063.0</td>\n",
       "      <td>32725090.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.358</td>\n",
       "      <td>tx.254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362373</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>32716076.0</td>\n",
       "      <td>32725092.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.358</td>\n",
       "      <td>tx.253</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1365077</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>34355448.0</td>\n",
       "      <td>34357262.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.384</td>\n",
       "      <td>tx.255</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1379257</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>69898039.0</td>\n",
       "      <td>69939015.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.388</td>\n",
       "      <td>tx.260</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387623</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90930513.0</td>\n",
       "      <td>90931188.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.363</td>\n",
       "      <td>tx.262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390318</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>106355801.0</td>\n",
       "      <td>106356344.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.365</td>\n",
       "      <td>tx.264</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393683</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>111594139.0</td>\n",
       "      <td>111599693.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.367</td>\n",
       "      <td>tx.265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394528</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>112869310.0</td>\n",
       "      <td>112870158.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.369</td>\n",
       "      <td>tx.267</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397103</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>123969269.0</td>\n",
       "      <td>124021626.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.392</td>\n",
       "      <td>tx.268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407155</th>\n",
       "      <td>chr6</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>147628147.0</td>\n",
       "      <td>147637503.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.394</td>\n",
       "      <td>tx.274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1426025</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17473088.0</td>\n",
       "      <td>17475894.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.432</td>\n",
       "      <td>tx.279</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431399</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30298592.0</td>\n",
       "      <td>30300450.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.434</td>\n",
       "      <td>tx.283</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451600</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>77590337.0</td>\n",
       "      <td>77591649.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.443</td>\n",
       "      <td>tx.291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1454634</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>84217259.0</td>\n",
       "      <td>84385650.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.424</td>\n",
       "      <td>tx.292</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1480451</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>129774130.0</td>\n",
       "      <td>129789047.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.451</td>\n",
       "      <td>tx.299</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1486155</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>140348628.0</td>\n",
       "      <td>140349858.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.428</td>\n",
       "      <td>tx.300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1491854</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>150921784.0</td>\n",
       "      <td>150924213.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.453</td>\n",
       "      <td>tx.304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1495273</th>\n",
       "      <td>chr7</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>153916931.0</td>\n",
       "      <td>153932929.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.456</td>\n",
       "      <td>tx.308</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503145</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12861724.0</td>\n",
       "      <td>12868233.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.497</td>\n",
       "      <td>tx.313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514484</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>37971115.0</td>\n",
       "      <td>37971975.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.499</td>\n",
       "      <td>tx.314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518232</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>42454297.0</td>\n",
       "      <td>42467949.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.468</td>\n",
       "      <td>tx.315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1523192</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>54242425.0</td>\n",
       "      <td>54257253.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.501</td>\n",
       "      <td>tx.317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539471</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>102044592.0</td>\n",
       "      <td>102045602.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.470</td>\n",
       "      <td>tx.319</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1544278</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>120984563.0</td>\n",
       "      <td>120997822.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.472</td>\n",
       "      <td>tx.322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551113</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>133798933.0</td>\n",
       "      <td>133799364.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.478</td>\n",
       "      <td>tx.325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1553225</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>140822598.0</td>\n",
       "      <td>140827244.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.480</td>\n",
       "      <td>tx.326</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1555020</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>143722621.0</td>\n",
       "      <td>143724487.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.484</td>\n",
       "      <td>tx.329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1558308</th>\n",
       "      <td>chr8</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>145242092.0</td>\n",
       "      <td>145243382.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.510</td>\n",
       "      <td>tx.331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1571275</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>34399400.0</td>\n",
       "      <td>34411946.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.514</td>\n",
       "      <td>tx.336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578509</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>78039622.0</td>\n",
       "      <td>78054932.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.518</td>\n",
       "      <td>tx.341</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1579174</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>80416469.0</td>\n",
       "      <td>80423125.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.520</td>\n",
       "      <td>tx.344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1579597</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>81130550.0</td>\n",
       "      <td>81149406.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.550</td>\n",
       "      <td>tx.345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1581379</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>84674326.0</td>\n",
       "      <td>84678574.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.553</td>\n",
       "      <td>tx.347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1581380</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>84674326.0</td>\n",
       "      <td>84678619.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.553</td>\n",
       "      <td>tx.348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1602651</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>136194008.0</td>\n",
       "      <td>136197159.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.562</td>\n",
       "      <td>tx.360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1617463</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148347192.0</td>\n",
       "      <td>148351240.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.536</td>\n",
       "      <td>tx.365</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1617470</th>\n",
       "      <td>chr9</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148348403.0</td>\n",
       "      <td>148351240.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.536</td>\n",
       "      <td>tx.366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1639213</th>\n",
       "      <td>chrX</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48246162.0</td>\n",
       "      <td>48247878.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.1013</td>\n",
       "      <td>tx.668</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648700</th>\n",
       "      <td>chrX</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>70162333.0</td>\n",
       "      <td>70186304.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.1015</td>\n",
       "      <td>tx.669</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648701</th>\n",
       "      <td>chrX</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>70162333.0</td>\n",
       "      <td>70186304.0</td>\n",
       "      <td>-</td>\n",
       "      <td>gene.1029</td>\n",
       "      <td>tx.670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1675176</th>\n",
       "      <td>chrX</td>\n",
       "      <td>Bambu</td>\n",
       "      <td>transcript</td>\n",
       "      <td>154250789.0</td>\n",
       "      <td>154252403.0</td>\n",
       "      <td>+</td>\n",
       "      <td>gene.1019</td>\n",
       "      <td>tx.675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr source        type        start          end strand    gene_id  \\\n",
       "17202     chr1  Bambu  transcript   16245735.0   16247595.0      -    gene.51   \n",
       "26029     chr1  Bambu  transcript   26268774.0   26270140.0      -    gene.64   \n",
       "27377     chr1  Bambu  transcript   27231889.0   27233743.0      -    gene.66   \n",
       "29806     chr1  Bambu  transcript   29948153.0   29953421.0      -    gene.68   \n",
       "30273     chr1  Bambu  transcript   31007199.0   31007976.0      +     gene.5   \n",
       "36935     chr1  Bambu  transcript   38167408.0   38300289.0      -    gene.72   \n",
       "36936     chr1  Bambu  transcript   38167408.0   38300320.0      -    gene.72   \n",
       "61708     chr1  Bambu  transcript   68243853.0   68261649.0      -    gene.76   \n",
       "64998     chr1  Bambu  transcript   77862296.0   77863997.0      +    gene.13   \n",
       "85157     chr1  Bambu  transcript  120334901.0  120336836.0      +    gene.17   \n",
       "91413     chr1  Bambu  transcript  148917243.0  148922629.0      +    gene.19   \n",
       "92695     chr1  Bambu  transcript  149753837.0  149759779.0      +    gene.22   \n",
       "98730     chr1  Bambu  transcript  154007468.0  154010460.0      +    gene.31   \n",
       "111178    chr1  Bambu  transcript  161794715.0  161832275.0      -    gene.79   \n",
       "117401    chr1  Bambu  transcript  175875912.0  175907502.0      -    gene.81   \n",
       "125445    chr1  Bambu  transcript  200352725.0  200355594.0      +    gene.35   \n",
       "131400    chr1  Bambu  transcript  204819539.0  204833822.0      -    gene.94   \n",
       "132960    chr1  Bambu  transcript  206395084.0  206587072.0      -    gene.98   \n",
       "135005    chr1  Bambu  transcript  210503378.0  210505309.0      -   gene.102   \n",
       "140388    chr1  Bambu  transcript  223623753.0  223710822.0      +    gene.37   \n",
       "144983    chr1  Bambu  transcript  230181846.0  230187587.0      -   gene.105   \n",
       "146378    chr1  Bambu  transcript  232706207.0  232713755.0      -   gene.107   \n",
       "153290    chr1  Bambu  transcript  246706575.0  246719608.0      +    gene.42   \n",
       "153292    chr1  Bambu  transcript  246706576.0  246719608.0      +    gene.42   \n",
       "155249   chr10  Bambu  transcript    2056961.0    2057421.0      -   gene.601   \n",
       "155700   chr10  Bambu  transcript    3271431.0    3278712.0      -   gene.603   \n",
       "155702   chr10  Bambu  transcript    3278199.0    3283710.0      +   gene.570   \n",
       "155705   chr10  Bambu  transcript    3278380.0    3294924.0      +   gene.570   \n",
       "171270   chr10  Bambu  transcript   44529886.0   44530712.0      +   gene.578   \n",
       "189091   chr10  Bambu  transcript   78505098.0   78505743.0      +   gene.585   \n",
       "193716   chr10  Bambu  transcript   92435427.0   92441761.0      +   gene.590   \n",
       "194879   chr10  Bambu  transcript   94625049.0   94637789.0      -   gene.611   \n",
       "208300   chr10  Bambu  transcript  117366422.0  117367340.0      -   gene.616   \n",
       "215889   chr10  Bambu  transcript  133672119.0  133686919.0      -   gene.618   \n",
       "226187   chr11  Bambu  transcript    6443663.0    6448780.0      -   gene.643   \n",
       "269596   chr11  Bambu  transcript   66284625.0   66285418.0      -   gene.647   \n",
       "273498   chr11  Bambu  transcript   68008808.0   68010284.0      +   gene.633   \n",
       "279345   chr11  Bambu  transcript   72722034.0   72738805.0      +   gene.635   \n",
       "308884   chr11  Bambu  transcript  125265652.0  125268543.0      -   gene.651   \n",
       "313300   chr12  Bambu  transcript      58097.0      58757.0      -   gene.679   \n",
       "350302   chr12  Bambu  transcript   51969337.0   51970360.0      -   gene.683   \n",
       "357784   chr12  Bambu  transcript   56071418.0   56071983.0      -   gene.687   \n",
       "370756   chr12  Bambu  transcript   71649099.0   71663543.0      -   gene.689   \n",
       "374900   chr12  Bambu  transcript   84966212.0   84972081.0      +   gene.665   \n",
       "388906   chr12  Bambu  transcript  110427575.0  110433747.0      +   gene.668   \n",
       "389681   chr12  Bambu  transcript  111010612.0  111014762.0      -   gene.695   \n",
       "404578   chr12  Bambu  transcript  130046464.0  130066361.0      +   gene.674   \n",
       "407118   chr12  Bambu  transcript  133013744.0  133041046.0      +   gene.676   \n",
       "407120   chr12  Bambu  transcript  133013753.0  133035037.0      +   gene.676   \n",
       "417849   chr13  Bambu  transcript   40766018.0   40768303.0      -   gene.716   \n",
       "421677   chr13  Bambu  transcript   48425802.0   48469158.0      -   gene.718   \n",
       "421693   chr13  Bambu  transcript   48456818.0   48469157.0      -   gene.718   \n",
       "426221   chr13  Bambu  transcript   66423827.0   66433257.0      +   gene.706   \n",
       "430190   chr13  Bambu  transcript   94505657.0   94515582.0      +   gene.710   \n",
       "447602   chr14  Bambu  transcript   18265046.0   18270516.0      -   gene.752   \n",
       "456102   chr14  Bambu  transcript   32395914.0   32397985.0      -   gene.755   \n",
       "484306   chr14  Bambu  transcript   86904499.0   86919989.0      -   gene.759   \n",
       "487013   chr14  Bambu  transcript   90175238.0   90180287.0      +   gene.744   \n",
       "494251   chr14  Bambu  transcript   98898949.0   98899910.0      +   gene.746   \n",
       "494253   chr14  Bambu  transcript   98898967.0   98899911.0      +   gene.746   \n",
       "495833   chr14  Bambu  transcript   99548453.0   99554926.0      +   gene.748   \n",
       "512179   chr15  Bambu  transcript   37841702.0   37844407.0      +   gene.767   \n",
       "514644   chr15  Bambu  transcript   39299216.0   39300931.0      -   gene.783   \n",
       "521539   chr15  Bambu  transcript   42975296.0   42996825.0      +   gene.771   \n",
       "555183   chr15  Bambu  transcript   87252266.0   87253187.0      +   gene.778   \n",
       "568741   chr16  Bambu  transcript    1963459.0    1964607.0      +   gene.786   \n",
       "572961   chr16  Bambu  transcript    2788183.0    2789466.0      -   gene.802   \n",
       "578543   chr16  Bambu  transcript    7463970.0    7473639.0      -   gene.804   \n",
       "582226   chr16  Bambu  transcript   14524554.0   14526611.0      +   gene.788   \n",
       "610001   chr16  Bambu  transcript   62456161.0   62458251.0      -   gene.806   \n",
       "612575   chr16  Bambu  transcript   63530738.0   63531850.0      -   gene.808   \n",
       "614745   chr16  Bambu  transcript   67824318.0   67825229.0      +   gene.792   \n",
       "629577   chr16  Bambu  transcript   83064170.0   83164948.0      -   gene.810   \n",
       "634074   chr16  Bambu  transcript   92127704.0   92128126.0      +   gene.798   \n",
       "634544   chr16  Bambu  transcript   92829633.0   92832182.0      -   gene.818   \n",
       "649181   chr17  Bambu  transcript    6138050.0    6141095.0      +   gene.829   \n",
       "662261   chr17  Bambu  transcript   17699000.0   17701276.0      +   gene.832   \n",
       "673585   chr17  Bambu  transcript   31608319.0   31632258.0      -   gene.848   \n",
       "683404   chr17  Bambu  transcript   40984434.0   40985750.0      -   gene.855   \n",
       "691105   chr17  Bambu  transcript   44866785.0   44872362.0      +   gene.838   \n",
       "706520   chr17  Bambu  transcript   59381670.0   59384925.0      +   gene.840   \n",
       "715546   chr17  Bambu  transcript   69085396.0   69089945.0      +   gene.842   \n",
       "742119   chr18  Bambu  transcript   10610363.0   10617136.0      -   gene.879   \n",
       "748338   chr18  Bambu  transcript   26741726.0   26749456.0      +   gene.865   \n",
       "752121   chr18  Bambu  transcript   37759890.0   37760308.0      +   gene.870   \n",
       "771656   chr19  Bambu  transcript    1413060.0    1417349.0      -   gene.920   \n",
       "774109   chr19  Bambu  transcript    2467933.0    2471009.0      -   gene.923   \n",
       "774111   chr19  Bambu  transcript    2469533.0    2471009.0      -   gene.923   \n",
       "777360   chr19  Bambu  transcript    4568029.0    4569755.0      +   gene.892   \n",
       "777362   chr19  Bambu  transcript    4568032.0    4569755.0      +   gene.892   \n",
       "783220   chr19  Bambu  transcript    7920226.0    7921212.0      -   gene.928   \n",
       "793397   chr19  Bambu  transcript   12845394.0   12847086.0      +   gene.895   \n",
       "804559   chr19  Bambu  transcript   18256654.0   18257704.0      -   gene.930   \n",
       "813325   chr19  Bambu  transcript   36042689.0   36046532.0      +   gene.900   \n",
       "847914   chr19  Bambu  transcript   54680415.0   54683045.0      -   gene.941   \n",
       "848477   chr19  Bambu  transcript   55095524.0   55109767.0      +   gene.911   \n",
       "854521   chr19  Bambu  transcript   58325760.0   58326123.0      +   gene.915   \n",
       "864944    chr2  Bambu  transcript   10101554.0   10103412.0      +   gene.113   \n",
       "891191    chr2  Bambu  transcript   61799577.0   61804796.0      +   gene.119   \n",
       "899117    chr2  Bambu  transcript   73297500.0   73299117.0      +   gene.123   \n",
       "907354    chr2  Bambu  transcript   91752692.0   91756493.0      +   gene.125   \n",
       "911514    chr2  Bambu  transcript   98630427.0   98634337.0      -   gene.178   \n",
       "911515    chr2  Bambu  transcript   98630427.0   98634503.0      -   gene.178   \n",
       "913734    chr2  Bambu  transcript  102075198.0  102082766.0      +   gene.136   \n",
       "921797    chr2  Bambu  transcript  119714686.0  119716307.0      -   gene.185   \n",
       "922150    chr2  Bambu  transcript  120157246.0  120193844.0      -   gene.188   \n",
       "926396    chr2  Bambu  transcript  130955720.0  130957449.0      -   gene.191   \n",
       "935270    chr2  Bambu  transcript  153100945.0  153102389.0      -   gene.193   \n",
       "939774    chr2  Bambu  transcript  165751335.0  165753268.0      -   gene.195   \n",
       "946553    chr2  Bambu  transcript  175288147.0  175336308.0      +   gene.153   \n",
       "961734    chr2  Bambu  transcript  203013644.0  203037175.0      -   gene.197   \n",
       "961753    chr2  Bambu  transcript  203034011.0  203037177.0      -   gene.197   \n",
       "974730    chr2  Bambu  transcript  225594940.0  225595646.0      +   gene.161   \n",
       "980978    chr2  Bambu  transcript  235870026.0  235880218.0      -   gene.210   \n",
       "990124   chr20  Bambu  transcript    2673938.0    2682913.0      -   gene.959   \n",
       "998062   chr20  Bambu  transcript   21575784.0   21579289.0      -   gene.961   \n",
       "998923   chr20  Bambu  transcript   23673408.0   23673932.0      -   gene.963   \n",
       "1013489  chr20  Bambu  transcript   46678603.0   46679406.0      +   gene.948   \n",
       "1016063  chr20  Bambu  transcript   48142813.0   48180597.0      +   gene.950   \n",
       "1023504  chr20  Bambu  transcript   63753479.0   63756724.0      +   gene.955   \n",
       "1029111  chr21  Bambu  transcript   12726210.0   12736277.0      +   gene.967   \n",
       "1031905  chr21  Bambu  transcript   22480907.0   22511040.0      +   gene.969   \n",
       "1034200  chr21  Bambu  transcript   30240182.0   30241210.0      -   gene.975   \n",
       "1045610  chr21  Bambu  transcript   43713514.0   43719491.0      +   gene.973   \n",
       "1057067  chr22  Bambu  transcript   24441306.0   24449445.0      -   gene.989   \n",
       "1057068  chr22  Bambu  transcript   24441306.0   24449438.0      -   gene.989   \n",
       "1057069  chr22  Bambu  transcript   24441306.0   24449440.0      -   gene.989   \n",
       "1057070  chr22  Bambu  transcript   24441306.0   24449457.0      -   gene.989   \n",
       "1059717  chr22  Bambu  transcript   27094628.0   27109928.0      -   gene.994   \n",
       "1060073  chr22  Bambu  transcript   28134962.0   28138620.0      +   gene.981   \n",
       "1061738  chr22  Bambu  transcript   29952946.0   29954661.0      -   gene.997   \n",
       "1067845  chr22  Bambu  transcript   33531107.0   33538458.0      -   gene.999   \n",
       "1081056  chr22  Bambu  transcript   50134537.0   50135382.0      +   gene.987   \n",
       "1084613   chr3  Bambu  transcript    2091485.0    2092562.0      -   gene.244   \n",
       "1091894   chr3  Bambu  transcript   11715305.0   11725053.0      +   gene.214   \n",
       "1100230   chr3  Bambu  transcript   33277040.0   33279143.0      -   gene.250   \n",
       "1101471   chr3  Bambu  transcript   35639358.0   35640647.0      -   gene.252   \n",
       "1110109   chr3  Bambu  transcript   46011738.0   46017127.0      +   gene.222   \n",
       "1110111   chr3  Bambu  transcript   46011751.0   46017127.0      +   gene.222   \n",
       "1132806   chr3  Bambu  transcript   71621703.0   71622859.0      +   gene.229   \n",
       "1145454   chr3  Bambu  transcript  118681174.0  118698136.0      +   gene.233   \n",
       "1153520   chr3  Bambu  transcript  131928955.0  131941197.0      +   gene.237   \n",
       "1162885   chr3  Bambu  transcript  147500963.0  147510850.0      -   gene.254   \n",
       "1186901   chr3  Bambu  transcript  197080131.0  197091651.0      +   gene.242   \n",
       "1201674   chr4  Bambu  transcript   10063511.0   10072577.0      +   gene.258   \n",
       "1204051   chr4  Bambu  transcript   17464779.0   17467542.0      -   gene.277   \n",
       "1205196   chr4  Bambu  transcript   22708168.0   22708552.0      -   gene.279   \n",
       "1205800   chr4  Bambu  transcript   25171560.0   25215801.0      -   gene.281   \n",
       "1218292   chr4  Bambu  transcript   74241667.0   74244666.0      -   gene.287   \n",
       "1221372   chr4  Bambu  transcript   80495548.0   80497541.0      -   gene.292   \n",
       "1221373   chr4  Bambu  transcript   80495548.0   80496517.0      -   gene.292   \n",
       "1251415   chr4  Bambu  transcript  154493284.0  154497983.0      +   gene.268   \n",
       "1255614   chr4  Bambu  transcript  162346996.0  162399093.0      -   gene.298   \n",
       "1257794   chr4  Bambu  transcript  172729385.0  172749149.0      -   gene.300   \n",
       "1263598   chr4  Bambu  transcript  189834257.0  189837658.0      +   gene.272   \n",
       "1271493   chr5  Bambu  transcript   24750070.0   24750707.0      +   gene.304   \n",
       "1278222   chr5  Bambu  transcript   43427116.0   43444990.0      -   gene.328   \n",
       "1287731   chr5  Bambu  transcript   72679720.0   72687282.0      -   gene.332   \n",
       "1288326   chr5  Bambu  transcript   73727808.0   73756224.0      -   gene.335   \n",
       "1302651   chr5  Bambu  transcript  113062662.0  113064001.0      +   gene.306   \n",
       "1312281   chr5  Bambu  transcript  135401779.0  135402332.0      -   gene.341   \n",
       "1328350   chr5  Bambu  transcript  158094251.0  158095637.0      +   gene.313   \n",
       "1329641   chr5  Bambu  transcript  162337156.0  162379506.0      -   gene.345   \n",
       "1330428   chr5  Bambu  transcript  164101947.0  164266752.0      +   gene.316   \n",
       "1333368   chr5  Bambu  transcript  174504523.0  174505895.0      +   gene.318   \n",
       "1335939   chr5  Bambu  transcript  178002662.0  178003592.0      +   gene.324   \n",
       "1342763   chr6  Bambu  transcript    3324738.0    3338104.0      +   gene.351   \n",
       "1346946   chr6  Bambu  transcript   13201864.0   13203102.0      +   gene.354   \n",
       "1347642   chr6  Bambu  transcript   17204392.0   17231356.0      +   gene.356   \n",
       "1348823   chr6  Bambu  transcript   21008097.0   21013056.0      -   gene.379   \n",
       "1348825   chr6  Bambu  transcript   21010120.0   21013113.0      -   gene.379   \n",
       "1362371   chr6  Bambu  transcript   32716063.0   32725090.0      +   gene.358   \n",
       "1362373   chr6  Bambu  transcript   32716076.0   32725092.0      +   gene.358   \n",
       "1365077   chr6  Bambu  transcript   34355448.0   34357262.0      -   gene.384   \n",
       "1379257   chr6  Bambu  transcript   69898039.0   69939015.0      -   gene.388   \n",
       "1387623   chr6  Bambu  transcript   90930513.0   90931188.0      +   gene.363   \n",
       "1390318   chr6  Bambu  transcript  106355801.0  106356344.0      +   gene.365   \n",
       "1393683   chr6  Bambu  transcript  111594139.0  111599693.0      +   gene.367   \n",
       "1394528   chr6  Bambu  transcript  112869310.0  112870158.0      +   gene.369   \n",
       "1397103   chr6  Bambu  transcript  123969269.0  124021626.0      -   gene.392   \n",
       "1407155   chr6  Bambu  transcript  147628147.0  147637503.0      -   gene.394   \n",
       "1426025   chr7  Bambu  transcript   17473088.0   17475894.0      -   gene.432   \n",
       "1431399   chr7  Bambu  transcript   30298592.0   30300450.0      -   gene.434   \n",
       "1451600   chr7  Bambu  transcript   77590337.0   77591649.0      -   gene.443   \n",
       "1454634   chr7  Bambu  transcript   84217259.0   84385650.0      +   gene.424   \n",
       "1480451   chr7  Bambu  transcript  129774130.0  129789047.0      -   gene.451   \n",
       "1486155   chr7  Bambu  transcript  140348628.0  140349858.0      +   gene.428   \n",
       "1491854   chr7  Bambu  transcript  150921784.0  150924213.0      -   gene.453   \n",
       "1495273   chr7  Bambu  transcript  153916931.0  153932929.0      -   gene.456   \n",
       "1503145   chr8  Bambu  transcript   12861724.0   12868233.0      -   gene.497   \n",
       "1514484   chr8  Bambu  transcript   37971115.0   37971975.0      -   gene.499   \n",
       "1518232   chr8  Bambu  transcript   42454297.0   42467949.0      +   gene.468   \n",
       "1523192   chr8  Bambu  transcript   54242425.0   54257253.0      -   gene.501   \n",
       "1539471   chr8  Bambu  transcript  102044592.0  102045602.0      +   gene.470   \n",
       "1544278   chr8  Bambu  transcript  120984563.0  120997822.0      +   gene.472   \n",
       "1551113   chr8  Bambu  transcript  133798933.0  133799364.0      +   gene.478   \n",
       "1553225   chr8  Bambu  transcript  140822598.0  140827244.0      +   gene.480   \n",
       "1555020   chr8  Bambu  transcript  143722621.0  143724487.0      +   gene.484   \n",
       "1558308   chr8  Bambu  transcript  145242092.0  145243382.0      -   gene.510   \n",
       "1571275   chr9  Bambu  transcript   34399400.0   34411946.0      +   gene.514   \n",
       "1578509   chr9  Bambu  transcript   78039622.0   78054932.0      +   gene.518   \n",
       "1579174   chr9  Bambu  transcript   80416469.0   80423125.0      +   gene.520   \n",
       "1579597   chr9  Bambu  transcript   81130550.0   81149406.0      -   gene.550   \n",
       "1581379   chr9  Bambu  transcript   84674326.0   84678574.0      -   gene.553   \n",
       "1581380   chr9  Bambu  transcript   84674326.0   84678619.0      -   gene.553   \n",
       "1602651   chr9  Bambu  transcript  136194008.0  136197159.0      -   gene.562   \n",
       "1617463   chr9  Bambu  transcript  148347192.0  148351240.0      +   gene.536   \n",
       "1617470   chr9  Bambu  transcript  148348403.0  148351240.0      +   gene.536   \n",
       "1639213   chrX  Bambu  transcript   48246162.0   48247878.0      +  gene.1013   \n",
       "1648700   chrX  Bambu  transcript   70162333.0   70186304.0      +  gene.1015   \n",
       "1648701   chrX  Bambu  transcript   70162333.0   70186304.0      -  gene.1029   \n",
       "1675176   chrX  Bambu  transcript  154250789.0  154252403.0      +  gene.1019   \n",
       "\n",
       "        transcript_id exon_number is_novel_transcript is_novel_gene  \n",
       "17202            tx.2         NaN                True          True  \n",
       "26029            tx.6         NaN                True          True  \n",
       "27377            tx.7         NaN                True          True  \n",
       "29806           tx.10         NaN                True          True  \n",
       "30273           tx.11         NaN                True          True  \n",
       "36935           tx.12         NaN                True          True  \n",
       "36936           tx.13         NaN                True          True  \n",
       "61708           tx.22         NaN                True          True  \n",
       "64998           tx.23         NaN                True          True  \n",
       "85157           tx.26         NaN                True          True  \n",
       "91413           tx.37         NaN                True          True  \n",
       "92695           tx.39         NaN                True          True  \n",
       "98730           tx.41         NaN                True          True  \n",
       "111178          tx.46         NaN                True          True  \n",
       "117401          tx.47         NaN                True          True  \n",
       "125445          tx.51         NaN                True          True  \n",
       "131400          tx.53         NaN                True          True  \n",
       "132960          tx.54         NaN                True          True  \n",
       "135005          tx.56         NaN                True          True  \n",
       "140388          tx.58         NaN                True          True  \n",
       "144983          tx.60         NaN                True          True  \n",
       "146378          tx.61         NaN                True          True  \n",
       "153290          tx.67         NaN                True          True  \n",
       "153292          tx.66         NaN                True          True  \n",
       "155249         tx.372         NaN                True          True  \n",
       "155700         tx.374         NaN                True          True  \n",
       "155702         tx.375         NaN                True          True  \n",
       "155705         tx.376         NaN                True          True  \n",
       "171270         tx.379         NaN                True          True  \n",
       "189091         tx.388         NaN                True          True  \n",
       "193716         tx.394         NaN                True          True  \n",
       "194879         tx.395         NaN                True          True  \n",
       "208300         tx.396         NaN                True          True  \n",
       "215889         tx.400         NaN                True          True  \n",
       "226187         tx.405         NaN                True          True  \n",
       "269596         tx.415         NaN                True          True  \n",
       "273498         tx.416         NaN                True          True  \n",
       "279345         tx.417         NaN                True          True  \n",
       "308884         tx.424         NaN                True          True  \n",
       "313300         tx.425         NaN                True          True  \n",
       "350302         tx.428         NaN                True          True  \n",
       "357784         tx.429         NaN                True          True  \n",
       "370756         tx.433         NaN                True          True  \n",
       "374900         tx.434         NaN                True          True  \n",
       "388906         tx.439         NaN                True          True  \n",
       "389681         tx.440         NaN                True          True  \n",
       "404578         tx.444         NaN                True          True  \n",
       "407118         tx.446         NaN                True          True  \n",
       "407120         tx.447         NaN                True          True  \n",
       "417849         tx.455         NaN                True          True  \n",
       "421677         tx.457         NaN                True          True  \n",
       "421693         tx.458         NaN                True          True  \n",
       "426221         tx.461         NaN                True          True  \n",
       "430190         tx.464         NaN                True          True  \n",
       "447602         tx.472         NaN                True          True  \n",
       "456102         tx.477         NaN                True          True  \n",
       "484306         tx.482         NaN                True          True  \n",
       "487013         tx.483         NaN                True          True  \n",
       "494251         tx.484         NaN                True          True  \n",
       "494253         tx.485         NaN                True          True  \n",
       "495833         tx.486         NaN                True          True  \n",
       "512179         tx.490         NaN                True          True  \n",
       "514644         tx.493         NaN                True          True  \n",
       "521539         tx.494         NaN                True          True  \n",
       "555183         tx.502         NaN                True          True  \n",
       "568741         tx.505         NaN                True          True  \n",
       "572961         tx.506         NaN                True          True  \n",
       "578543         tx.509         NaN                True          True  \n",
       "582226         tx.510         NaN                True          True  \n",
       "610001         tx.516         NaN                True          True  \n",
       "612575         tx.517         NaN                True          True  \n",
       "614745         tx.518         NaN                True          True  \n",
       "629577         tx.520         NaN                True          True  \n",
       "634074         tx.523         NaN                True          True  \n",
       "634544         tx.524         NaN                True          True  \n",
       "649181         tx.532         NaN                True          True  \n",
       "662261         tx.534         NaN                True          True  \n",
       "673585         tx.535         NaN                True          True  \n",
       "683404         tx.540         NaN                True          True  \n",
       "691105         tx.544         NaN                True          True  \n",
       "706520         tx.549         NaN                True          True  \n",
       "715546         tx.555         NaN                True          True  \n",
       "742119         tx.562         NaN                True          True  \n",
       "748338         tx.566         NaN                True          True  \n",
       "752121         tx.568         NaN                True          True  \n",
       "771656         tx.570         NaN                True          True  \n",
       "774109         tx.572         NaN                True          True  \n",
       "774111         tx.573         NaN                True          True  \n",
       "777360         tx.574         NaN                True          True  \n",
       "777362         tx.575         NaN                True          True  \n",
       "783220         tx.576         NaN                True          True  \n",
       "793397         tx.581         NaN                True          True  \n",
       "804559         tx.586         NaN                True          True  \n",
       "813325         tx.592         NaN                True          True  \n",
       "847914         tx.607         NaN                True          True  \n",
       "848477         tx.608         NaN                True          True  \n",
       "854521         tx.609         NaN                True          True  \n",
       "864944          tx.77         NaN                True          True  \n",
       "891191          tx.84         NaN                True          True  \n",
       "899117          tx.88         NaN                True          True  \n",
       "907354          tx.93         NaN                True          True  \n",
       "911514          tx.94         NaN                True          True  \n",
       "911515          tx.95         NaN                True          True  \n",
       "913734          tx.96         NaN                True          True  \n",
       "921797         tx.104         NaN                True          True  \n",
       "922150         tx.105         NaN                True          True  \n",
       "926396         tx.106         NaN                True          True  \n",
       "935270         tx.110         NaN                True          True  \n",
       "939774         tx.111         NaN                True          True  \n",
       "946553         tx.113         NaN                True          True  \n",
       "961734         tx.120         NaN                True          True  \n",
       "961753         tx.121         NaN                True          True  \n",
       "974730         tx.122         NaN                True          True  \n",
       "980978         tx.125         NaN                True          True  \n",
       "990124         tx.618         NaN                True          True  \n",
       "998062         tx.620         NaN                True          True  \n",
       "998923         tx.621         NaN                True          True  \n",
       "1013489        tx.626         NaN                True          True  \n",
       "1016063        tx.627         NaN                True          True  \n",
       "1023504        tx.631         NaN                True          True  \n",
       "1029111        tx.633         NaN                True          True  \n",
       "1031905        tx.637         NaN                True          True  \n",
       "1034200        tx.638         NaN                True          True  \n",
       "1045610        tx.643         NaN                True          True  \n",
       "1057067        tx.652         NaN                True          True  \n",
       "1057068        tx.653         NaN                True          True  \n",
       "1057069        tx.654         NaN                True          True  \n",
       "1057070        tx.655         NaN                True          True  \n",
       "1059717        tx.656         NaN                True          True  \n",
       "1060073        tx.657         NaN                True          True  \n",
       "1061738        tx.659         NaN                True          True  \n",
       "1067845        tx.661         NaN                True          True  \n",
       "1081056        tx.662         NaN                True          True  \n",
       "1084613        tx.126         NaN                True          True  \n",
       "1091894        tx.129         NaN                True          True  \n",
       "1100230        tx.133         NaN                True          True  \n",
       "1101471        tx.134         NaN                True          True  \n",
       "1110109        tx.142         NaN                True          True  \n",
       "1110111        tx.143         NaN                True          True  \n",
       "1132806        tx.149         NaN                True          True  \n",
       "1145454        tx.154         NaN                True          True  \n",
       "1153520        tx.160         NaN                True          True  \n",
       "1162885        tx.163         NaN                True          True  \n",
       "1186901        tx.168         NaN                True          True  \n",
       "1201674        tx.174         NaN                True          True  \n",
       "1204051        tx.175         NaN                True          True  \n",
       "1205196        tx.176         NaN                True          True  \n",
       "1205800        tx.177         NaN                True          True  \n",
       "1218292        tx.180         NaN                True          True  \n",
       "1221372        tx.183         NaN                True          True  \n",
       "1221373        tx.184         NaN                True          True  \n",
       "1251415        tx.194         NaN                True          True  \n",
       "1255614        tx.195         NaN                True          True  \n",
       "1257794        tx.196         NaN                True          True  \n",
       "1263598        tx.197         NaN                True          True  \n",
       "1271493        tx.200         NaN                True          True  \n",
       "1278222        tx.201         NaN                True          True  \n",
       "1287731        tx.205         NaN                True          True  \n",
       "1288326        tx.206         NaN                True          True  \n",
       "1302651        tx.212         NaN                True          True  \n",
       "1312281        tx.218         NaN                True          True  \n",
       "1328350        tx.220         NaN                True          True  \n",
       "1329641        tx.222         NaN                True          True  \n",
       "1330428        tx.225         NaN                True          True  \n",
       "1333368        tx.226         NaN                True          True  \n",
       "1335939        tx.228         NaN                True          True  \n",
       "1342763        tx.237         NaN                True          True  \n",
       "1346946        tx.238         NaN                True          True  \n",
       "1347642        tx.239         NaN                True          True  \n",
       "1348823        tx.241         NaN                True          True  \n",
       "1348825        tx.240         NaN                True          True  \n",
       "1362371        tx.254         NaN                True          True  \n",
       "1362373        tx.253         NaN                True          True  \n",
       "1365077        tx.255         NaN                True          True  \n",
       "1379257        tx.260         NaN                True          True  \n",
       "1387623        tx.262         NaN                True          True  \n",
       "1390318        tx.264         NaN                True          True  \n",
       "1393683        tx.265         NaN                True          True  \n",
       "1394528        tx.267         NaN                True          True  \n",
       "1397103        tx.268         NaN                True          True  \n",
       "1407155        tx.274         NaN                True          True  \n",
       "1426025        tx.279         NaN                True          True  \n",
       "1431399        tx.283         NaN                True          True  \n",
       "1451600        tx.291         NaN                True          True  \n",
       "1454634        tx.292         NaN                True          True  \n",
       "1480451        tx.299         NaN                True          True  \n",
       "1486155        tx.300         NaN                True          True  \n",
       "1491854        tx.304         NaN                True          True  \n",
       "1495273        tx.308         NaN                True          True  \n",
       "1503145        tx.313         NaN                True          True  \n",
       "1514484        tx.314         NaN                True          True  \n",
       "1518232        tx.315         NaN                True          True  \n",
       "1523192        tx.317         NaN                True          True  \n",
       "1539471        tx.319         NaN                True          True  \n",
       "1544278        tx.322         NaN                True          True  \n",
       "1551113        tx.325         NaN                True          True  \n",
       "1553225        tx.326         NaN                True          True  \n",
       "1555020        tx.329         NaN                True          True  \n",
       "1558308        tx.331         NaN                True          True  \n",
       "1571275        tx.336         NaN                True          True  \n",
       "1578509        tx.341         NaN                True          True  \n",
       "1579174        tx.344         NaN                True          True  \n",
       "1579597        tx.345         NaN                True          True  \n",
       "1581379        tx.347         NaN                True          True  \n",
       "1581380        tx.348         NaN                True          True  \n",
       "1602651        tx.360         NaN                True          True  \n",
       "1617463        tx.365         NaN                True          True  \n",
       "1617470        tx.366         NaN                True          True  \n",
       "1639213        tx.668         NaN                True          True  \n",
       "1648700        tx.669         NaN                True          True  \n",
       "1648701        tx.670         NaN                True          True  \n",
       "1675176        tx.675         NaN                True          True  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13_ref.loc[chm13_ref[\"is_novel_gene\"] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
