{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "73877976",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import libraries\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "96f60159",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_df_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True, is_transcript=False, is_exon=False):\n",
    "\n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "        \n",
    "        ## Get gene names\n",
    "        df[\"gene_name\"] = df[\"other\"].str.split(\"gene_name \\\"\", expand=True)[1].str.split('\\\";', expand=True)[0]\n",
    "        \n",
    "        ## Get get transcript biotype\n",
    "        df[\"gene_biotype\"] = df[\"other\"].str.split('gene_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "        \n",
    "        if is_transcript:\n",
    "            df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            df[\"transcript_biotype\"] = df[\"other\"].str.split('transcript_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "        \n",
    "        if is_exon:\n",
    "            df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            df[\"transcript_biotype\"] = df[\"other\"].str.split('transcript_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "            \n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in df.columns:\n",
    "        df.loc[df[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "074ab4fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open original reference\n",
    "original_ref = pd.read_csv(\"../../references/Homo_sapiens.GRCh38.107_ERCC.gtf\", header=None, delimiter=\"\\t\", low_memory=False, \n",
    "                       names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"], comment=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b8da85c",
   "metadata": {},
   "outputs": [],
   "source": [
    "original_ref_gene = original_ref.loc[original_ref[\"type\"]==\"gene\"].copy()\n",
    "orig_ref_gene =  parse_df_columns(original_ref_gene.copy(), is_ref=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "815388a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Drop genes without name\n",
    "orig_ref_gene.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7a49435",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create parkinson's dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6201690/\n",
    "## https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8048219/\n",
    "\n",
    "dict_parkinsons = {\"gene_name\": ['LRRK2', 'PARK2', 'PRNP', 'PARK7', 'PINK1', 'SNCA', 'MAPT', 'GBA', 'ACMSD', 'ASXL3', 'BCKDK', 'BRIP1', 'BST1', \n",
    "                                 'C5orf24', 'CAB39L', 'CCDC62', 'CD19', 'CHRNB1', 'CLCN3', 'CRLS1', 'DDRGK1', 'DGKQ', 'DNAH17', 'DYRK1A', \n",
    "                                 'FAM171A2', 'FAM47E', 'FAM49B', 'FBRSL1', 'FCGR2A', 'FGF20', 'FYN', 'GAK', 'GBA', 'GBF1', 'GPNMB', 'HIP1', 'HLA-DQB1',\n",
    "                                 'HLA-DRA', 'HLA-DRB5', 'INPP5F', 'KCNIP3', 'KCNS3', 'KPNA1', 'LAMP', 'LCORL', 'LINC00693', 'MBNL2', 'MCCC1/3',\n",
    "                                 'MED12L', 'MEX3C', 'MIPOL1', 'NOD2', 'NUCKS', 'PAM', 'RAB29', 'RAB7L1', 'RAI1', 'RIMS1', 'RIT2', 'RNF141', 'RPS12', 'RPS6KL1',\n",
    "                                 'SCAF11', 'SCARB2', 'SIPA1L2', 'SPTSSB', 'SREBF1', 'STBD1', 'STK39', 'STX1B', 'SYT11', 'TMEM163', 'TMEM175', \n",
    "                                 'TRIM40', 'UBAP2', 'UBTF', 'VAMP4', 'VPS13C']}\n",
    "\n",
    "\n",
    "df_parkinsons = pd.DataFrame(data=dict_parkinsons)\n",
    "\n",
    "df_parkinsons[\"disease\"] = \"PD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7ed56f13",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create schizophrenia dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.nature.com/articles/s41586-022-04556-w\n",
    "\n",
    "dict_schizophrenia = {\"gene_name\": ['TRIO', 'SP4', 'CUL1', 'XPO7', 'RB1CC1', 'SETD1A', 'HERC1', \n",
    "                                    'GRIN2A', 'CACNA1G', 'GRIA3']}\n",
    "\n",
    "\n",
    "df_schizophrenia = pd.DataFrame(data=dict_schizophrenia)\n",
    "\n",
    "df_schizophrenia[\"disease\"] = \"Schizophrenia\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f52f55fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create mood disorders dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.nature.com/articles/s41588-022-01034-x (Bipolar Disorder)\n",
    "## https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6065213/ (Major Depressive Disorder)\n",
    "\n",
    "\n",
    "dict_mood = {\"gene_name\": ['DRD4', 'HTR1A', 'MAOA', 'PCLO', 'SLC6A3', 'SLC6A4', 'TPH2', \"ACE\", \"APOE\",\n",
    "                          \"MTHFR\", \"CHST11\", \"PTPRR\", \"ADCY9\", \"ITPR1\", \"DNAJB2\", \"EHD3\", \"FREM3\", \n",
    "                          \"GNB3\", \"PHACTR3\", \"HS6ST3\", \"KLHL29\", \"LHFPL2\", \"SLC25A21\", \"UGT2A1\", \"VGLL4\"]}\n",
    "\n",
    "df_mood = pd.DataFrame(data=dict_mood)\n",
    "\n",
    "df_mood[\"disease\"] = \"MDD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6737eab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create LATE (Limbic Predominant TDP-43 Encephalopathy) dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://pubmed.ncbi.nlm.nih.gov/31039256/\n",
    "\n",
    "\n",
    "dict_late = {\"gene_name\": ['GRN', 'TMEM106B', 'ABCC9', 'KCNMB2', 'APOE']}\n",
    "\n",
    "df_late = pd.DataFrame(data=dict_late)\n",
    "\n",
    "df_late[\"disease\"] = \"LATE\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4c6e4cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create ALS & FTD dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8048219/\n",
    "\n",
    "\n",
    "dict_als_ftd = {\"gene_name\": ['GRN', 'MAPT', 'FUS', 'SOD1', 'TARDBP', 'C9orf72', 'TBK1', 'VCP', 'BTNL2', 'C4orf27', 'CTSC',\n",
    "                           'DPP6', 'HLA-DRA', 'HLA-DRB5', 'HLA-DQA2', 'IMMP2L', 'IRF2', 'MIR548AP', 'OLFM1', 'RAB38', 'RERG',\n",
    "                           'TMEM106B', 'UNC13A', 'C21orf2', 'DPP6', 'FGGY', 'ITPR2', 'KIF5A', 'MOBP', 'SARM', 'SCFD1', 'UNC13A']}\n",
    "\n",
    "df_als_ftd = pd.DataFrame(data=dict_als_ftd)\n",
    "\n",
    "df_als_ftd[\"disease\"] = \"ALS/FTD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cf9a5eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create Autism dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7250485/\n",
    "\n",
    "\n",
    "dict_autism = {\"gene_name\": ['ASH1L', 'POGZ', 'SCN2A', 'FOXP1', 'SLC6A1', 'CTNNB1', 'ANK2', 'ARID1B', 'SYNGAP1', \n",
    "                              'GIGYF1', 'PTEN', 'DEAF1', 'KMT5B', 'GRIN2B', 'MED13L', 'CHD8', 'CHD2', 'KDM6B', 'ANKRD11',\n",
    "                              'TLK2', 'DSCAM', 'SHANK3', 'DYRK1A', 'ADNP']}\n",
    "\n",
    "df_autism = pd.DataFrame(data=dict_autism)\n",
    "\n",
    "df_autism[\"disease\"] = \"ASD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "32379545",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create PTSD dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8917986/#SD1\n",
    "## https://www.nature.com/articles/s41588-020-00767-x\n",
    "\n",
    "\n",
    "#dict_ptsd = {\"gene_name\": ['DCAF5', 'EXD2', 'FAM120A', 'FOXP2', 'GALNT16', 'PHF2', 'METTL15', 'AUTS2',\n",
    "#                           'MAD1L1', 'CNTN6', 'BMP2', 'PACRG', 'PARK2']}\n",
    "\n",
    "#df_ptsd = pd.DataFrame(data=dict_ptsd)\n",
    "\n",
    "#df_ptsd[\"disease\"] = \"PTSD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e2d34f5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create substance use disorders dictionary\n",
    "## List of curated genes made from the following articles:\n",
    "## https://www.nature.com/articles/s44220-023-00034-y#MOESM1\n",
    "\n",
    "dict_sud = {\"gene_name\": ['PDE4B', 'GTF3C2', 'IFT172', 'GCKR', 'C2orf16', 'PLCL2', 'PRKAR2A', 'SLC25A20', 'ARIH2', 'P4HTM',\n",
    "                          'WDR6', 'QRICH1', 'QARS', 'USP19', 'C3orf84', 'CCDC36', 'RP11-3B7.1', 'USP4', 'RHOA', 'NICN1',\n",
    "                          'DAG1', 'ADD1', 'NOP14', 'BANK1', 'PPP6C', 'SOX6', 'BDNF', 'MTCH2', 'ZDHHC5', 'TMX2', 'TMX2-CTNND1',\n",
    "                          'C11orf31', 'RP11-691N7.6', 'CTNND1', 'ANKK1', 'DRD2', 'HS6ST3', 'ARID4A', 'PPP1R13B', 'SEMA6D',\n",
    "                          'FTO', 'C20orf112']}\n",
    "\n",
    "df_sud = pd.DataFrame(data=dict_sud)\n",
    "\n",
    "df_sud[\"disease\"] = \"SUD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ec3c0af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create dictionary with all brain diseases\n",
    "df_brain_disease = pd.concat([df_parkinsons, df_schizophrenia, df_mood, df_late, df_als_ftd, df_autism, df_sud])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8a3f7b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only keep genes with a gene name that matches the reference genome\n",
    "\n",
    "df_brain_disease = df_brain_disease.merge(orig_ref_gene[[\"gene_name\", \"gene_id\"]], how=\"inner\", on=[\"gene_name\"]).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1af934af",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load medically relevant genes\n",
    "med_relevant_genes = pd.read_csv(\"../../references/medically_relevant_genes.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "37534178",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load AD genes \n",
    "ad_relevant_genes = pd.read_csv(\"../../references/AD_gwas_genes.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d04fdaa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add annotation to AD genes\n",
    "ad_relevant_genes[\"disease\"] = \"AD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2a65eb97",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add AD genes to brain disease genes\n",
    "df_brain_disease = pd.concat([df_brain_disease, ad_relevant_genes])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8932f3c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MAPT</td>\n",
       "      <td>PD</td>\n",
       "      <td>ENSG00000186868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MAPT</td>\n",
       "      <td>ALS/FTD</td>\n",
       "      <td>ENSG00000186868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BCKDK</td>\n",
       "      <td>PD</td>\n",
       "      <td>ENSG00000103507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>DYRK1A</td>\n",
       "      <td>PD</td>\n",
       "      <td>ENSG00000157540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>DYRK1A</td>\n",
       "      <td>ASD</td>\n",
       "      <td>ENSG00000157540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>PD</td>\n",
       "      <td>ENSG00000204287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>ALS/FTD</td>\n",
       "      <td>ENSG00000204287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>PD</td>\n",
       "      <td>ENSG00000198502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>ALS/FTD</td>\n",
       "      <td>ENSG00000198502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ACE</td>\n",
       "      <td>MDD</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>APOE</td>\n",
       "      <td>MDD</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>APOE</td>\n",
       "      <td>LATE</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>HS6ST3</td>\n",
       "      <td>MDD</td>\n",
       "      <td>ENSG00000185352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>HS6ST3</td>\n",
       "      <td>SUD</td>\n",
       "      <td>ENSG00000185352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>GRN</td>\n",
       "      <td>LATE</td>\n",
       "      <td>ENSG00000030582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>GRN</td>\n",
       "      <td>ALS/FTD</td>\n",
       "      <td>ENSG00000030582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>TMEM106B</td>\n",
       "      <td>LATE</td>\n",
       "      <td>ENSG00000106460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>TMEM106B</td>\n",
       "      <td>ALS/FTD</td>\n",
       "      <td>ENSG00000106460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TMEM106B</td>\n",
       "      <td>AD</td>\n",
       "      <td>ENSG00000106460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>BCKDK</td>\n",
       "      <td>AD</td>\n",
       "      <td>ENSG00000103507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>GRN</td>\n",
       "      <td>AD</td>\n",
       "      <td>ENSG00000030582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>ACE</td>\n",
       "      <td>AD</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>APOE</td>\n",
       "      <td>AD</td>\n",
       "      <td>ENSG00000130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MAPT</td>\n",
       "      <td>AD</td>\n",
       "      <td>ENSG00000186868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name  disease          gene_id\n",
       "5        MAPT       PD  ENSG00000186868\n",
       "6        MAPT  ALS/FTD  ENSG00000186868\n",
       "11      BCKDK       PD  ENSG00000103507\n",
       "24     DYRK1A       PD  ENSG00000157540\n",
       "25     DYRK1A      ASD  ENSG00000157540\n",
       "37    HLA-DRA       PD  ENSG00000204287\n",
       "38    HLA-DRA  ALS/FTD  ENSG00000204287\n",
       "39   HLA-DRB5       PD  ENSG00000198502\n",
       "40   HLA-DRB5  ALS/FTD  ENSG00000198502\n",
       "92        ACE      MDD  ENSG00000159640\n",
       "93       APOE      MDD  ENSG00000130203\n",
       "94       APOE     LATE  ENSG00000130203\n",
       "105    HS6ST3      MDD  ENSG00000185352\n",
       "106    HS6ST3      SUD  ENSG00000185352\n",
       "112       GRN     LATE  ENSG00000030582\n",
       "113       GRN  ALS/FTD  ENSG00000030582\n",
       "114  TMEM106B     LATE  ENSG00000106460\n",
       "115  TMEM106B  ALS/FTD  ENSG00000106460\n",
       "24   TMEM106B       AD  ENSG00000106460\n",
       "53      BCKDK       AD  ENSG00000103507\n",
       "62        GRN       AD  ENSG00000030582\n",
       "66        ACE       AD  ENSG00000159640\n",
       "76       APOE       AD  ENSG00000130203\n",
       "79       MAPT       AD  ENSG00000186868"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Show all gene names involved in multiple diseases\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "33fc30e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Annotate gene names involved in multiple diseases\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"] == \"MAPT\", \"disease\"] = \"PD + ALS/FTD + AD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"BCKDK\"]), \"disease\"] = \"PD + AD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"DYRK1A\"]), \"disease\"] = \"PD + ASD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"HLA-DRA\", \"HLA-DRB5\"]), \"disease\"] = \"PD + ALS/FTD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"TMEM106B\", \"GRN\"]), \"disease\"] = \"LATE + ALS/FTD + AD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"APOE\"]), \"disease\"] = \"LATE + AD + MDD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"APOE\"]), \"disease\"] = \"LATE + AD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"ACE\"]), \"disease\"] = \"AD + MDD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"HS6ST3\"]), \"disease\"] = \"MDD + SUD\"\n",
    "df_brain_disease.loc[df_brain_disease[\"gene_name\"].isin([\"TARDBP\"]), \"disease\"] = \"ALS/FTD + LATE\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d44355fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_brain_disease.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "17ba7551",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create dataframe with brain disease IDs\n",
    "df_brain_disease_IDS = df_brain_disease[[\"gene_id\", \"gene_name\"]].copy().drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a8ee292a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merge brain diseases and med relevant genes so that med relevant genes include all brain disease genes\n",
    "med_relevant_genes_final = pd.concat([med_relevant_genes, df_brain_disease_IDS])\n",
    "\n",
    "med_relevant_genes_final.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "94f1539c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save new medically relvant genes list including the brain disease genes\n",
    "med_relevant_genes.to_csv(\"../../references/medically_relevant_genes_02-04-2023_UPDATED.tsv\", \n",
    "                          sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ed56bfe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save brain disease dataframes\n",
    "\n",
    "df_brain_disease.to_csv(\"../../references/brain_disease_genes_with_disease.tsv\", \n",
    "                        sep=\"\\t\", index=False)\n",
    "\n",
    "df_brain_disease_IDS.to_csv(\"../../references/brain_disease_genes_only_IDs.tsv\", \n",
    "                        sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
